
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>50. Glossary &#8212; Single-cell best practices</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/book.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/book.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="shortcut icon" href="_static/favicon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="49. Acknowledgements" href="acknowledgements.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Single-cell best practices</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Introduction
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="introduction/prior_art.html">
   1. Prior art
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="introduction/scrna_seq.html">
   2. Single-cell RNA sequencing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="introduction/raw_data_processing.html">
   3. Raw data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="introduction/analysis_tools.html">
   4. Analysis frameworks and tools
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="introduction/interoperability.html">
   5. Interoperability
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Preprocessing and visualization
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="preprocessing_visualization/quality_control.html">
   6. Quality Control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="preprocessing_visualization/normalization.html">
   7. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="preprocessing_visualization/feature_selection.html">
   8. Feature selection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="preprocessing_visualization/dimensionality_reduction.html">
   9. Dimensionality Reduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Identifying cellular structure
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="cellular_structure/clustering.html">
   10. Clustering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="cellular_structure/annotation.html">
   11. Annotation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="cellular_structure/integration.html">
   12. Data integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Inferring trajectories
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="trajectories/pseudotemporal.html">
   13. Pseudotemporal ordering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="trajectories/rna_velocity.html">
   14. RNA velocity
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="trajectories/lineage_tracing.html">
   15. Lineage tracing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Dealing with conditions
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="conditions/differential_gene_expression.html">
   16. Differential gene expression analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="conditions/compositional.html">
   17. Compositional analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="conditions/gsea_pathway.html">
   18. Gene set enrichment and pathway analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="conditions/perturbation_modeling.html">
   19. Perturbation modeling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Modeling mechanisms
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="mechanisms/gene_regulatory_networks.html">
   20. Gene regulatory networks
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="mechanisms/cell_cell_communication.html">
   21. Cell-cell communication
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Deconvolution
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="deconvolution/bulk_deconvolution.html">
   22. Bulk deconvolution
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Chromatin Accessibility
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="chromatin_accessibility/introduction.html">
   23. Single-cell ATAC sequencing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="chromatin_accessibility/quality_control.html">
   24. Quality Control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="chromatin_accessibility/gene_regulatory_networks_atac.html">
   25. Gene regulatory networks using chromatin accessibility
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Spatial omics
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="spatial/introduction.html">
   26. Single-cell data resolved in space
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="spatial/neighborhood.html">
   27. Neighborhood analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="spatial/domains.html">
   28. Spatial domains
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="spatial/spatially_variable_genes.html">
   29. Spatially variable genes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="spatial/deconvolution.html">
   30. Spatial deconvolution
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="spatial/imputation.html">
   31. Imputation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Surface protein
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="surface_protein/quality_control.html">
   32. Quality control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="surface_protein/normalization.html">
   33. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="surface_protein/doublet_detection.html">
   34. Doublet detection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="surface_protein/dimensionality_reduction.html">
   35. Dimensionality Reduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="surface_protein/batch_correction.html">
   36. Batch correction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="surface_protein/annotation.html">
   37. Annotation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Adaptive immune receptor repertoire
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="air_repertoire/ir_profiling.html">
   38. Immune Receptor Profiling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="air_repertoire/clonotype.html">
   39. Clonotype analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="air_repertoire/specificity.html">
   43. Specificity analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="air_repertoire/multimodal_integration.html">
   44. Integrating AIR and transcriptomics
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multimodal integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="multimodal_integration/paired_integration.html">
   45. Paired integration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="multimodal_integration/advanced_integration.html">
   46. Advanced integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reproducibility
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="reproducibility/introduction.html">
   47. Reproducibility
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Outlook
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="outlook.html">
   48. Outlook
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Acknowledgements
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="acknowledgements.html">
   49. Acknowledgements
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Glossary
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   50. Glossary
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fglossary.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/glossary.md"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Edit this page"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="headerbtn__text-container">suggest edit</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="_sources/glossary.md"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.md</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Glossary</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="glossary">
<h1><span class="section-number">50. </span>Glossary<a class="headerlink" href="#glossary" title="Permalink to this headline">#</a></h1>
<dl class="glossary simple">
<dt id="term-Algorithm">Algorithm<a class="headerlink" href="#term-Algorithm" title="Permalink to this term">#</a></dt><dt id="term-Algorithms">Algorithms<a class="headerlink" href="#term-Algorithms" title="Permalink to this term">#</a></dt><dd><p>A pre-defined set of instructions to solve a problem.</p>
</dd>
<dt id="term-AnnData">AnnData<a class="headerlink" href="#term-AnnData" title="Permalink to this term">#</a></dt><dt id="term-AnnDatas">AnnDatas<a class="headerlink" href="#term-AnnDatas" title="Permalink to this term">#</a></dt><dd><p>A Python package for annotated data matrices. The primary data structure used in the scverse ecosystem.</p>
</dd>
<dt id="term-Barcode">Barcode<a class="headerlink" href="#term-Barcode" title="Permalink to this term">#</a></dt><dt id="term-Barcodes">Barcodes<a class="headerlink" href="#term-Barcodes" title="Permalink to this term">#</a></dt><dt id="term-Bar-code">Bar code<a class="headerlink" href="#term-Bar-code" title="Permalink to this term">#</a></dt><dt id="term-Bar-codes">Bar codes<a class="headerlink" href="#term-Bar-codes" title="Permalink to this term">#</a></dt><dd><p>Short DNA barcode fragments (“tags”) that are used to identify reads originating from the same cell. Reads are later grouped by their barcode during raw data processing steps.</p>
</dd>
<dt id="term-Batch-effect">Batch effect<a class="headerlink" href="#term-Batch-effect" title="Permalink to this term">#</a></dt><dd><p>Technical confounding factors in an experiment that cause dataset distribution shifts. Usually lead to inaccurate conclusions if the causes of the batch effects are correlated with outcomes of interest in an experiment and should be accounted for (usually removed).</p>
</dd>
<dt id="term-Benchmark">Benchmark<a class="headerlink" href="#term-Benchmark" title="Permalink to this term">#</a></dt><dd><p>An (independent) comparison of performance of several tools with respect to pre-defined metrics.</p>
</dd>
<dt id="term-Bulk-RNA-sequencing">Bulk RNA sequencing<a class="headerlink" href="#term-Bulk-RNA-sequencing" title="Permalink to this term">#</a></dt><dd><p>Contrary to single-cell sequencing, bulk sequencing measures the average expression values of several cells. Therefore, resolution is lost, but bulk sequencing is usually cheaper, less laborious and faster to analyze.</p>
</dd>
<dt id="term-Cell">Cell<a class="headerlink" href="#term-Cell" title="Permalink to this term">#</a></dt><dd><p>The fundamental unit of life. Consists of cytoplasm enclosed within a membrane that contains many biomolecules such as proteins and nucleic acids. Cells acquire specific functions, transition to cell types, divide, communicate and keep the organism going. Learning about the structure, activity and communication of cells helps deciphering biology.</p>
</dd>
<dt id="term-Cell-barcode">Cell barcode<a class="headerlink" href="#term-Cell-barcode" title="Permalink to this term">#</a></dt><dd><p>See <a class="reference internal" href="#term-Barcode"><span class="xref std std-term">barcode</span></a></p>
</dd>
<dt id="term-Cluster">Cluster<a class="headerlink" href="#term-Cluster" title="Permalink to this term">#</a></dt><dt id="term-Clusters">Clusters<a class="headerlink" href="#term-Clusters" title="Permalink to this term">#</a></dt><dd><p>A group of a population or data points that share similarities. In single-cell, clusters usually share a common function or marker gene expression that is used for annotation (see <a class="reference internal" href="#term-Cell-type-annotation"><span class="xref std std-term">cell type annotation</span></a>).</p>
</dd>
<dt id="term-Cell-type-annotation">Cell type annotation<a class="headerlink" href="#term-Cell-type-annotation" title="Permalink to this term">#</a></dt><dd><p>The process of labeling groups of <a class="reference internal" href="#term-Clusters"><span class="xref std std-term">clusters</span></a> of cells by <a class="reference internal" href="#term-Cell-type"><span class="xref std std-term">cell type</span></a>. Commonly done based on <a class="reference internal" href="#term-Cell-type"><span class="xref std std-term">cell type</span></a> specific markers, automatically with classifiers or by mapping against a reference.</p>
</dd>
<dt id="term-Cell-type">Cell type<a class="headerlink" href="#term-Cell-type" title="Permalink to this term">#</a></dt><dd><p>Cells that share common morphological or phenotypic features.</p>
</dd>
<dt id="term-Cell-state">Cell state<a class="headerlink" href="#term-Cell-state" title="Permalink to this term">#</a></dt><dd><p>Cells can be annotated according to <a class="reference internal" href="#term-Cell-type"><span class="xref std std-term">cell type</span></a> or other cell states as defined by the cell-cycle, perturbational state or other features.</p>
</dd>
<dt id="term-Chromatin">Chromatin<a class="headerlink" href="#term-Chromatin" title="Permalink to this term">#</a></dt><dd><p>The complex of DNA and proteins efficiently packaging the DNA inside the nucleus and involved in regulating gene expression.</p>
</dd>
<dt id="term-Demultiplexing">Demultiplexing<a class="headerlink" href="#term-Demultiplexing" title="Permalink to this term">#</a></dt><dd><p>The process of determining which sequencing reads belong to which cell using <a class="reference internal" href="#term-Barcodes"><span class="xref std std-term">barcodes</span></a>.</p>
</dd>
<dt id="term-DNA">DNA<a class="headerlink" href="#term-DNA" title="Permalink to this term">#</a></dt><dd><p>DNA is the acronym of Deoxyribonucleic acid. It is the organic chemical storing hereditary information and instructions for protein synthesis. DNA gets transcribed into <a class="reference internal" href="#term-RNA"><span class="xref std std-term">RNA</span></a>.</p>
</dd>
<dt id="term-Doublets">Doublets<a class="headerlink" href="#term-Doublets" title="Permalink to this term">#</a></dt><dd><p>Reads obtained from droplet based assays might be mistakenly associated to a single cell while the RNA expression origins from two or more cells (a doublet).</p>
</dd>
<dt id="term-Dropout">Dropout<a class="headerlink" href="#term-Dropout" title="Permalink to this term">#</a></dt><dd><p>A gene with low expression that is observed in one cell, but not in other cells of the same <a class="reference internal" href="#term-Cell-type"><span class="xref std std-term">cell type</span></a>. The reason for dropouts are commonly low amounts of mRNA expression in cells and the general stochasticity of mRNA expression. Dropouts are one of the reasons why scRNA-seq data is sparse.</p>
</dd>
<dt id="term-Drop-seq">Drop-seq<a class="headerlink" href="#term-Drop-seq" title="Permalink to this term">#</a></dt><dd><p>A protocol for scRNA-seq that separates cells into nano-liter sized aqueous droplets enabling large-scale profiling.</p>
</dd>
<dt id="term-FASTQ-reads">FASTQ reads<a class="headerlink" href="#term-FASTQ-reads" title="Permalink to this term">#</a></dt><dd><p>Sequencing reads that are saved in the FASTQ format. FASTQ files are then used to map against the reference genome of interest to obtain gene counts for cells.</p>
</dd>
<dt id="term-Gene-expression-matrix">Gene expression matrix<a class="headerlink" href="#term-Gene-expression-matrix" title="Permalink to this term">#</a></dt><dd><p>A cell (barcode) by gene (scverse ecosystem) or gene by cell (barcode) matrix storing counts in the cell values.</p>
</dd>
<dt id="term-Imputation">Imputation<a class="headerlink" href="#term-Imputation" title="Permalink to this term">#</a></dt><dd><p>The replacement of missing values with usually artificial values.</p>
</dd>
<dt id="term-Indrop">Indrop<a class="headerlink" href="#term-Indrop" title="Permalink to this term">#</a></dt><dd><p>A Droplet based protocol for scRNA-seq.</p>
</dd>
<dt id="term-Library">Library<a class="headerlink" href="#term-Library" title="Permalink to this term">#</a></dt><dd><p>Also known as sequencing library. A pool of DNA fragments with attached sequencing adapters.</p>
</dd>
<dt id="term-MuData">MuData<a class="headerlink" href="#term-MuData" title="Permalink to this term">#</a></dt><dd><p>A Python package for multimodal annotated data matrices. The primary data structure in the scverse ecosystem for multimodal data.</p>
</dd>
<dt id="term-muon">muon<a class="headerlink" href="#term-muon" title="Permalink to this term">#</a></dt><dd><p>A Python package for multi-modal single-cell analysis in Python by scverse.</p>
</dd>
<dt id="term-Negative-binomial-distribution">Negative binomial distribution<a class="headerlink" href="#term-Negative-binomial-distribution" title="Permalink to this term">#</a></dt><dd><p>A discrete probability distribution that models the number of successes in a sequence of independent and identically distributed Bernoulli trials before a specified number of failures.</p>
</dd>
<dt id="term-Pipeline">Pipeline<a class="headerlink" href="#term-Pipeline" title="Permalink to this term">#</a></dt><dd><p>Also often times denoted as workflow. A pre-specified selection of steps that are commonly executed in order.</p>
</dd>
<dt id="term-RNA">RNA<a class="headerlink" href="#term-RNA" title="Permalink to this term">#</a></dt><dd><p>Ribonucleic acid. Single-stranded nucleic acid present in all living cells that encodes and regulates gene expression.</p>
</dd>
<dt id="term-RT-qPCR">RT-qPCR<a class="headerlink" href="#term-RT-qPCR" title="Permalink to this term">#</a></dt><dd><p>Quantitative reverse transcription PCR (RT-qPCR) monitors the amplification of a targeted <a class="reference internal" href="#term-DNA"><span class="xref std std-term">DNA</span></a> molecule during the PCR.</p>
</dd>
<dt id="term-PCR">PCR<a class="headerlink" href="#term-PCR" title="Permalink to this term">#</a></dt><dd><p>Polymercase chain reaction (PCR) is a method to amplify sequences to create billions of copies. PCR requires primers, which are short synthetic <a class="reference internal" href="#term-DNA"><span class="xref std std-term">DNA</span></a> fragments, to select the genome segments to be amplified and subsequently multiple rounds of <a class="reference internal" href="#term-DNA"><span class="xref std std-term">DNA</span></a> synthesis to amplify the targeted segments.</p>
</dd>
<dt id="term-Poisson-distribution">Poisson distribution<a class="headerlink" href="#term-Poisson-distribution" title="Permalink to this term">#</a></dt><dd><p>Discrete probability distribution denoting the probability of a specified number of events occurring in a fixed interval of time or space with the events occurring independently at a known constant mean rate.</p>
</dd>
<dt id="term-Promoter">Promoter<a class="headerlink" href="#term-Promoter" title="Permalink to this term">#</a></dt><dd><p>Sequence of DNA to which proteins bind to initiate and control transcription.</p>
</dd>
<dt id="term-Pseudotime">Pseudotime<a class="headerlink" href="#term-Pseudotime" title="Permalink to this term">#</a></dt><dd><p>Latent and therefore unobserved dimension reflecting cells’ progression through transitions. Pseudotime is usually related to real time events, but not necessarily the same.</p>
</dd>
<dt id="term-scanpy">scanpy<a class="headerlink" href="#term-scanpy" title="Permalink to this term">#</a></dt><dd><p>A Python package for single-cell analysis in Python by scverse.</p>
</dd>
<dt id="term-scverse">scverse<a class="headerlink" href="#term-scverse" title="Permalink to this term">#</a></dt><dd><p>A consortium for fundamental single-cell tools in the life sciences that are maintaining computational analysis tools like scanpy, muon and scvi-tools. See: <a class="reference external" href="https://scverse.org/">https://scverse.org/</a></p>
</dd>
<dt id="term-Spike-in-RNA">Spike-in RNA<a class="headerlink" href="#term-Spike-in-RNA" title="Permalink to this term">#</a></dt><dd><p>RNA transcripts of known sequence and quantity to calibrate measurements in RNA hybridization steps for RNA-seq.</p>
</dd>
<dt id="term-Trajectory-inference">Trajectory inference<a class="headerlink" href="#term-Trajectory-inference" title="Permalink to this term">#</a></dt><dd><p>Also known as pseudotemporal ordering. The computational recovery of dynamic processes by ordering cells by similarity or other means.</p>
</dd>
<dt id="term-Unique-Molecular-Identifier-UMI">Unique Molecular Identifier (UMI)<a class="headerlink" href="#term-Unique-Molecular-Identifier-UMI" title="Permalink to this term">#</a></dt><dd><p>Specific type of molecular barcodes aiding with error correction and increased accuracy during sequencing. UMIs unique tag molecules in sample libraries enabling estimation of PCR duplication rates.</p>
</dd>
</dl>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="acknowledgements.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title"><span class="section-number">49. </span>Acknowledgements</p>
        </div>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Lukas Heumos, Anna Schaar, single-cell best practices consortium<br/>
  
      &copy; Copyright 2022.<br/>
    <div class="extra_footer">
      <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>