
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>49. Outlook &#8212; Single-cell best practices</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="_static/book.css?v=148c11e2" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="_static/copybutton.js?v=f281be69"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'outlook';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/book.js?v=dcd3d8bf"></script>
    <link rel="icon" href="_static/favicon.ico"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="50. Acknowledgements" href="acknowledgements.html" />
    <link rel="prev" title="48. Reproducibility" href="reproducibility/introduction.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="preamble.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/logo.png" class="logo__image only-light" alt="Single-cell best practices - Home"/>
    <img src="_static/logo.png" class="logo__image only-dark pst-js-only" alt="Single-cell best practices - Home"/>
  
  
</a></div>
        <div class="sidebar-primary-item">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="introduction/prior_art.html">1. Prior art</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction/scrna_seq.html">2. Single-cell RNA sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction/raw_data_processing.html">3. Raw data processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction/analysis_tools.html">4. Analysis frameworks and tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction/interoperability.html">5. Interoperability</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Preprocessing and visualization</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="preprocessing_visualization/quality_control.html">6. Quality Control</a></li>

<li class="toctree-l1"><a class="reference internal" href="preprocessing_visualization/normalization.html">8. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="preprocessing_visualization/feature_selection.html">9. Feature selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="preprocessing_visualization/dimensionality_reduction.html">10. Dimensionality Reduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Identifying cellular structure</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="cellular_structure/clustering.html">11. Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="cellular_structure/annotation.html">12. Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="cellular_structure/integration.html">13. Data integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inferring trajectories</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="trajectories/pseudotemporal.html">14. Pseudotemporal ordering</a></li>
<li class="toctree-l1"><a class="reference internal" href="trajectories/rna_velocity.html">15. RNA velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="trajectories/lineage_tracing.html">16. Lineage tracing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Dealing with conditions</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="conditions/differential_gene_expression.html">17. Differential gene expression analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="conditions/compositional.html">18. Compositional analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="conditions/gsea_pathway.html">19. Gene set enrichment and pathway analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="conditions/perturbation_modeling.html">20. Perturbation modeling</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Modeling mechanisms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="mechanisms/gene_regulatory_networks.html">21. Gene regulatory networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="mechanisms/cell_cell_communication.html">22. Cell-cell communication</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Deconvolution</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="deconvolution/bulk_deconvolution.html">23. Bulk deconvolution</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Chromatin Accessibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="chromatin_accessibility/introduction.html">24. Single-cell ATAC sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="chromatin_accessibility/quality_control.html">25. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="chromatin_accessibility/gene_regulatory_networks_atac.html">26. Gene regulatory networks using chromatin accessibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Spatial omics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="spatial/introduction.html">27. Single-cell data resolved in space</a></li>
<li class="toctree-l1"><a class="reference internal" href="spatial/neighborhood.html">28. Neighborhood analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="spatial/domains.html">29. Spatial domains</a></li>
<li class="toctree-l1"><a class="reference internal" href="spatial/spatially_variable_genes.html">30. Spatially variable genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="spatial/deconvolution.html">31. Spatial deconvolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="spatial/imputation.html">32. Imputation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Surface protein</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="surface_protein/quality_control.html">33. Quality control</a></li>
<li class="toctree-l1"><a class="reference internal" href="surface_protein/normalization.html">34. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="surface_protein/doublet_detection.html">35. Doublet detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="surface_protein/dimensionality_reduction.html">36. Dimensionality Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="surface_protein/batch_correction.html">37. Batch correction</a></li>
<li class="toctree-l1"><a class="reference internal" href="surface_protein/annotation.html">38. Annotation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Adaptive immune receptor repertoire</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="air_repertoire/ir_profiling.html">39. Immune Receptor Profiling</a></li>
<li class="toctree-l1"><a class="reference internal" href="air_repertoire/clonotype.html">40. Clonotype analysis</a></li>



<li class="toctree-l1"><a class="reference internal" href="air_repertoire/specificity.html">44. Specificity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="air_repertoire/multimodal_integration.html">45. Integrating AIR and transcriptomics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multimodal integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="multimodal_integration/paired_integration.html">46. Paired integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="multimodal_integration/advanced_integration.html">47. Advanced integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reproducibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="reproducibility/introduction.html">48. Reproducibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Outlook</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">49. Outlook</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Acknowledgements</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="acknowledgements.html">50. Acknowledgements</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Glossary</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="glossary.html">51. Glossary</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/outlook.md" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Foutlook.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/outlook.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button>


<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Outlook</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#benchmarking">49.1. Benchmarking</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#atlas-building-and-reference-mapping">49.2. Atlas building and reference mapping</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#modality-prediction">49.3. Modality prediction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#end-to-end-pipelines">49.4. End-to-end pipelines</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#single-cell-proteomics">49.5. Single-cell proteomics</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">49.6. References</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="outlook">
<h1><span class="section-number">49. </span>Outlook<a class="headerlink" href="#outlook" title="Link to this heading">#</a></h1>
<p>The single-cell field is progressing at a rapid pace. This process is accompanied by a massive amount of newly developed tools, which do not allow for new discoveries, but also extend the pool of tools to select from.</p>
<section id="benchmarking">
<h2><span class="section-number">49.1. </span>Benchmarking<a class="headerlink" href="#benchmarking" title="Link to this heading">#</a></h2>
<p>Due to the vast amount of tools to choose from, we expect further and regular benchmarking to become even more important. Efforts such as <a class="reference external" href="https://openproblems.bio/">Open Problems in Single-Cell Analysis</a>, which aim to benchmark tools using well-defined metrics applied to standard datasets, will become essential.
However, for benchmarking to be successful, the single-cell community has to agree on a set of metrics per task.
We therefore expect discussions on what defines “good” or “bad” analysis steps to become increasingly important.
As soon as new benchmarking results and tool recommendations become available, we will update our chapters.</p>
</section>
<section id="atlas-building-and-reference-mapping">
<h2><span class="section-number">49.2. </span>Atlas building and reference mapping<a class="headerlink" href="#atlas-building-and-reference-mapping" title="Link to this heading">#</a></h2>
<p>With the advent of more single-cell atlases, we expect a slow transition from less manual annotation to more automated annotation <em>via</em> reference mapping.
For this to become a reality, the single-cell community needs not only healthy reference atlases, but also reference atlases for various conditions such as diseases.
We envision reference mapping to also allow for automated quality control through uncertainty estimation.
Cells that have a hard time getting mapped onto high quality references are either of low quality or highly biologically relevant.</p>
</section>
<section id="modality-prediction">
<h2><span class="section-number">49.3. </span>Modality prediction<a class="headerlink" href="#modality-prediction" title="Link to this heading">#</a></h2>
<p>Single-cell analysis is slowly starting to move on from unimodal to multimodal perspectives to obtain complete views of cells.
Though, it is not always possible to obtain all modalities. We therefore expect modality prediction and unpaired integration methods to become more important.</p>
</section>
<section id="end-to-end-pipelines">
<h2><span class="section-number">49.4. </span>End-to-end pipelines<a class="headerlink" href="#end-to-end-pipelines" title="Link to this heading">#</a></h2>
<p>The amount of single-cell data and analysis possibilities is growing on such a big scale that it is inevitable that as many processes as possible should be automated.
With the advent of reference mapping to atlases to automate cell type annotation and automated quality control tools, the development of highly automated pipelines is becoming a real possibility.
Especially, the reanalysis of “old”, existing datasets could uncover previously missed biology.
The first pipelines are starting to appear <span id="id1">[<a class="reference internal" href="#id5" title="Combiz Khozoie, Nurun Fancy, Mahdi M. Marjaneh, Alan E. Murphy, Paul M. Matthews, and Nathan Skene. Scflow: a scalable and reproducible analysis pipeline for single-cell term`rna` sequencing data. bioRxiv, 2021. URL: https://www.biorxiv.org/content/early/2021/08/19/2021.08.16.456499.1, arXiv:https://www.biorxiv.org/content/early/2021/08/19/2021.08.16.456499.1.full.pdf, doi:10.1101/2021.08.16.456499.">Khozoie <em>et al.</em>, 2021</a>]</span>, and we expect more pipelines with support for modalities beyond <a class="reference internal" href="glossary.html#term-RNA"><span class="xref std std-term">RNA</span></a> to be developed.</p>
</section>
<section id="single-cell-proteomics">
<h2><span class="section-number">49.5. </span>Single-cell proteomics<a class="headerlink" href="#single-cell-proteomics" title="Link to this heading">#</a></h2>
<p>Furthermore, the young and fast evolving field of single-cell proteomics <span id="id2">[<a class="reference internal" href="#id6" title="Andreas-David Brunner, Marvin Thielert, Catherine Vasilopoulou, Constantin Ammar, Fabian Coscia, Andreas Mund, Ole B Hoerning, Nicolai Bache, Amalia Apalategui, Markus Lubeck, Sabrina Richter, David S Fischer, Oliver Raether, Melvin A Park, Florian Meier, Fabian J Theis, and Matthias Mann. Ultra-high sensitivity mass spectrometry quantifies single-cell proteome changes upon perturbation. Molecular Systems Biology, 18(3):e10798, 2022. URL: https://www.embopress.org/doi/abs/10.15252/msb.202110798, arXiv:https://www.embopress.org/doi/pdf/10.15252/msb.202110798, doi:https://doi.org/10.15252/msb.202110798.">Brunner <em>et al.</em>, 2022</a>]</span> will bring an additional modality into the single-cell space with more integration challenges, but also great potential for new discoveries.
Currently, single-cell proteomics suffers from a lack of high quality datasets, making it challenging to develop tools and pipelines for sound statistical analysis.
As soon as high quality single-cell proteomics datasets are generated on a regular basis, we expect the single-cell community to develop new tools for the analysis.</p>
</section>
<section id="references">
<h2><span class="section-number">49.6. </span>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<div class="docutils container" id="id3">
<div role="list" class="citation-list">
<div class="citation" id="id6" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id2">olBTV+22</a><span class="fn-bracket">]</span></span>
<p>Andreas-David Brunner, Marvin Thielert, Catherine Vasilopoulou, Constantin Ammar, Fabian Coscia, Andreas Mund, Ole B Hoerning, Nicolai Bache, Amalia Apalategui, Markus Lubeck, Sabrina Richter, David S Fischer, Oliver Raether, Melvin A Park, Florian Meier, Fabian J Theis, and Matthias Mann. Ultra-high sensitivity mass spectrometry quantifies single-cell proteome changes upon perturbation. <em>Molecular Systems Biology</em>, 18(3):e10798, 2022. URL: <a class="reference external" href="https://www.embopress.org/doi/abs/10.15252/msb.202110798">https://www.embopress.org/doi/abs/10.15252/msb.202110798</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.embopress.org/doi/pdf/10.15252/msb.202110798">arXiv:https://www.embopress.org/doi/pdf/10.15252/msb.202110798</a>, <a class="reference external" href="https://doi.org/https://doi.org/10.15252/msb.202110798">doi:https://doi.org/10.15252/msb.202110798</a>.</p>
</div>
<div class="citation" id="id5" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">olKFM+21</a><span class="fn-bracket">]</span></span>
<p>Combiz Khozoie, Nurun Fancy, Mahdi M. Marjaneh, Alan E. Murphy, Paul M. Matthews, and Nathan Skene. Scflow: a scalable and reproducible analysis pipeline for single-cell term`rna` sequencing data. <em>bioRxiv</em>, 2021. URL: <a class="reference external" href="https://www.biorxiv.org/content/early/2021/08/19/2021.08.16.456499.1">https://www.biorxiv.org/content/early/2021/08/19/2021.08.16.456499.1</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.biorxiv.org/content/early/2021/08/19/2021.08.16.456499.1.full.pdf">arXiv:https://www.biorxiv.org/content/early/2021/08/19/2021.08.16.456499.1.full.pdf</a>, <a class="reference external" href="https://doi.org/10.1101/2021.08.16.456499">doi:10.1101/2021.08.16.456499</a>.</p>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="reproducibility/introduction.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">48. </span>Reproducibility</p>
      </div>
    </a>
    <a class="right-next"
       href="acknowledgements.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">50. </span>Acknowledgements</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#benchmarking">49.1. Benchmarking</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#atlas-building-and-reference-mapping">49.2. Atlas building and reference mapping</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#modality-prediction">49.3. Modality prediction</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#end-to-end-pipelines">49.4. End-to-end pipelines</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#single-cell-proteomics">49.5. Single-cell proteomics</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">49.6. References</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Lukas Heumos, Anna Schaar, single-cell best practices consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>