
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>3. Raw data processing &#8212; Single-cell best practices</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css?v=148c11e2" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'introduction/raw_data_processing';</script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js?v=dcd3d8bf"></script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="4. Analysis frameworks and tools" href="analysis_tools.html" />
    <link rel="prev" title="2. Single-cell RNA sequencing" href="scrna_seq.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../preamble.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/logo.png" class="logo__image only-light" alt="Single-cell best practices - Home"/>
    <img src="../_static/logo.png" class="logo__image only-dark pst-js-only" alt="Single-cell best practices - Home"/>
  
  
</a></div>
        <div class="sidebar-primary-item">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Introduction</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="prior_art.html">1. Prior art</a></li>
<li class="toctree-l1"><a class="reference internal" href="scrna_seq.html">2. Single-cell RNA sequencing</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">3. Raw data processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis_tools.html">4. Analysis frameworks and tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="interoperability.html">5. Interoperability</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Preprocessing and visualization</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/quality_control.html">6. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/normalization.html">7. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/feature_selection.html">8. Feature selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">9. Dimensionality Reduction</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Identifying cellular structure</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/clustering.html">10. Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/annotation.html">11. Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cellular_structure/integration.html">12. Data integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Inferring trajectories</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../trajectories/pseudotemporal.html">13. Pseudotemporal ordering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/rna_velocity.html">14. RNA velocity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../trajectories/lineage_tracing.html">15. Lineage tracing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Dealing with conditions</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../conditions/differential_gene_expression.html">16. Differential gene expression analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/compositional.html">17. Compositional analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/gsea_pathway.html">18. Gene set enrichment and pathway analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../conditions/perturbation_modeling.html">19. Perturbation modeling</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Modeling mechanisms</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/gene_regulatory_networks.html">20. Gene regulatory networks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mechanisms/cell_cell_communication.html">21. Cell-cell communication</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Deconvolution</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../deconvolution/bulk_deconvolution.html">22. Bulk deconvolution</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Chromatin Accessibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/introduction.html">23. Single-cell ATAC sequencing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/quality_control.html">24. Quality Control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../chromatin_accessibility/gene_regulatory_networks_atac.html">25. Gene regulatory networks using chromatin accessibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Spatial omics</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../spatial/introduction.html">26. Single-cell data resolved in space</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/neighborhood.html">27. Neighborhood analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/domains.html">28. Spatial domains</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/spatially_variable_genes.html">29. Spatially variable genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/deconvolution.html">30. Spatial deconvolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../spatial/imputation.html">31. Imputation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Surface protein</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/quality_control.html">32. Quality control</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/normalization.html">33. Normalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/doublet_detection.html">34. Doublet detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/dimensionality_reduction.html">35. Dimensionality Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/batch_correction.html">36. Batch correction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../surface_protein/annotation.html">37. Annotation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Adaptive immune receptor repertoire</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/ir_profiling.html">38. Immune Receptor Profiling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/clonotype.html">39. Clonotype analysis</a></li>



<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/specificity.html">43. Specificity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../air_repertoire/multimodal_integration.html">44. Integrating AIR and transcriptomics</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multimodal integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/paired_integration.html">45. Paired integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../multimodal_integration/advanced_integration.html">46. Advanced integration</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reproducibility</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../reproducibility/introduction.html">47. Reproducibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Outlook</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../outlook.html">48. Outlook</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Acknowledgements</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">49. Acknowledgements</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Glossary</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">50. Glossary</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/introduction/raw_data_processing.md" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fintroduction/raw_data_processing.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/introduction/raw_data_processing.md" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.md</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button>


<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Raw data processing</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#raw-data-quality-control">3.1. Raw data quality control</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#alignment-and-mapping">3.2. Alignment and mapping</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#types-of-mapping">3.2.1. Types of mapping</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-against-different-reference-sequences">3.2.2. Mapping against different reference sequences</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-to-the-full-genome">3.2.2.1. Mapping to the full genome</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-to-the-spliced-transcriptome">3.2.2.2. Mapping to the spliced transcriptome</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-to-an-augmented-transcriptome">3.2.2.3. Mapping to an augmented transcriptome</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cell-barcode-correction">3.3. Cell barcode correction</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#type-of-errors-in-barcoding">3.3.1. Type of errors in barcoding</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#future-challenges">3.3.2. Future challenges</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#umi-resolution">3.4. UMI resolution</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-need-for-umi-resolution">3.4.1. The need for UMI resolution</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#graph-based-umi-resolution">3.4.2. Graph-based UMI resolution</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#defining-nodes">3.4.2.1. Defining nodes</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#defining-the-adjacency-relationship">3.4.2.2. Defining the adjacency relationship</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#defining-the-graph-resolution-approach">3.4.2.3. Defining the graph resolution approach</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#quantification">3.4.2.4. Quantification</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#count-matrix-quality-control">3.5. Count matrix quality control</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#empty-droplet-detection">3.5.1. Empty droplet detection</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#doublet-detection">3.5.2. Doublet detection</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#count-data-representation">3.6. Count data representation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#brief-discussion">3.7. Brief discussion</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#a-real-world-example">3.8. A real-world example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#preparation">3.8.1. Preparation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#simplified-raw-data-processing-pipeline">3.8.2. Simplified raw data processing pipeline</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-complete-alevin-fry-pipeline">3.8.3. The complete alevin-fry pipeline</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#building-the-index">3.8.3.1. Building the index</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-and-quantification">3.8.3.2. Mapping and quantification</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#useful-links">3.9. Useful links</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">3.10. References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributors">3.11. Contributors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">3.11.1. Authors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">3.11.2. Reviewers</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="raw-data-processing">
<span id="raw-proc"></span><h1><span class="section-number">3. </span>Raw data processing<a class="headerlink" href="#raw-data-processing" title="Link to this heading">#</a></h1>
<p>In this section, we discuss some of the fundamental issues surrounding what is commonly called “preprocessing” of single-cell and single-nucleus RNA-sequencing (sc/snRNA-seq) data. Though this is common terminology, it seems a bit of a misnomer, as this process involves several
steps that make important decisions about how to deal with and represent the data that can enable or preclude subsequent analyses. Here, we will primarily refer to this phase of processing as “raw data processing”, and our focus will be on the phase of data analysis that begins with lane-demultiplexed FASTQ files, and ends with a count matrix representing the estimated number of distinct molecules arising from each gene within each quantified cell, potentially stratified by the inferred splicing status of each molecule (<a class="reference internal" href="#raw-proc-fig-overview"><span class="std std-numref">Fig. 3.1</span></a>).</p>
<figure class="align-default" id="raw-proc-fig-overview">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/overview_raw_data_processing.jpg"><img alt="Chapter Overview" class="bg-primary mb-1" src="../_images/overview_raw_data_processing.jpg" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3.1 </span><span class="caption-text">An overview of the topics discussed in this chapter. In the plot, “txome” stands for transcriptome.</span><a class="headerlink" href="#raw-proc-fig-overview" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>This count matrix then serves as the input for the myriad methods that have been developed for various analyses carried out with scRNA-seq data <span id="id1">[<a class="reference internal" href="../preamble.html#id10" title="Luke Zappia and Fabian J. Theis. Over 1000 tools reveal trends in the single-cell rna-seq analysis landscape. Genome Biology, 22(1):301, Oct 2021. URL: https://doi.org/10.1186/s13059-021-02519-4, doi:10.1186/s13059-021-02519-4.">Zappia and Theis, 2021</a>]</span>, from methods for normalization, integration, and filtering through methods for inferring cell types, developmental trajectories, and expression dynamics. Given that it serves as the starting point for all of these analyses, a robust and accurate estimation of this matrix is a foundational and critical step to support and enable accurate and reliable subsequent analyses. Fundamental misestimation in raw data processing can contribute to invalid inferences in higher-level analyses and can preclude discoveries based on the signal present in the raw data, which has been missed, diminished, or distorted by raw data processing. As we will see in this section, despite the intuitive nature of the input and output we seek from this step in the processing pipeline, several important and difficult challenges arise during raw data processing, and improved computational methodology for dealing with these challenges is still being actively developed. In particular, we will cover the fundamental steps in raw data processing, including read alignment/mapping, cell barcode (CB) identification and correction, and the estimation of molecule counts through the resolution of unique molecular identifiers (UMIs). We will also mention the choices and challenges that surface when performing these processing steps.</p>
<div class="admonition-a-note-on-what-precedes-raw-data-processing admonition">
<p class="admonition-title">A note on what precedes raw data processing</p>
<p>The decision of where to begin discussing raw data processing is somewhat arbitrary. We note that while, here, we consider starting with lane-demultiplexed FASTQ files as the <em>raw</em> input, even this already represents data that has been processed
and transformed from raw measurements. Further, some of the processing that precedes the generation of the FASTQ files is relevant to challenges that may arise in subsequent processing. For example, the process of base calling and base quality estimation affects the errors that arise in the FASTQ representation of the sequenced fragments, as well as the instrument’s estimation of the confidence of each called nucleotide. Further, issues can arise upstream of FASTQ generation, such as index hopping <span id="id2">[<a class="reference internal" href="#id109" title="Rick Farouni, Haig Djambazian, Lorenzo E Ferri, Jiannis Ragoussis, and Hamed S Najafabadi. Model-based analysis of sample index hopping reveals its widespread artifacts in multiplexed single-cell term`RNA`-sequencing. Nature communications, 11(1):1–8, 2020.">Farouni <em>et al.</em>, 2020</a>]</span>, though these issues can be largely mitigated both with <em>in silico</em> approaches <span id="id3">[<a class="reference internal" href="#id109" title="Rick Farouni, Haig Djambazian, Lorenzo E Ferri, Jiannis Ragoussis, and Hamed S Najafabadi. Model-based analysis of sample index hopping reveals its widespread artifacts in multiplexed single-cell term`RNA`-sequencing. Nature communications, 11(1):1–8, 2020.">Farouni <em>et al.</em>, 2020</a>]</span> and through enhanced protocols such as <a class="reference external" href="https://www.10xgenomics.com/blog/sequence-with-confidence-understand-index-hopping-and-how-to-resolve-it">dual indexing</a>. In this section, however, we will not consider upstream effects such as these, and instead will consider the FASTQ files, derived from, e.g., BCL files via <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/bcl2fastq-direct">appropriate tools</a>, as the raw input under consideration.</p>
</div>
<section id="raw-data-quality-control">
<h2><span class="section-number">3.1. </span>Raw data quality control<a class="headerlink" href="#raw-data-quality-control" title="Link to this heading">#</a></h2>
<p>Once raw FASTQ files have been obtained, the quality of the reads themselves can be quickly diagnosed by running a quality control (QC) tool, such as <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/"><code class="docutils literal notranslate"><span class="pre">FastQC</span></code></a>, to assess read quality, sequence content, etc. If run successfully, <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> generates a QC report for each input FASTQ file. Overall, this report summarizes the quality score, base content, and certain relevant statistics to spot potential problems originating from library preparation or sequencing.</p>
<p>Although nowadays, single-cell raw data processing tools internally evaluate some quality checks that are important for single-cell data, such as the N content of sequences and the fraction of mapped reads, it is still a good habit to run a quick quality check before processing the data, as it evaluates other useful QC metrics.
For readers who are interested in knowing what the <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> report for a FASTQ file from an RNA-seq sample looks like, in the following toggle content, we use the example <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> report of <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/good_sequence_short_fastqc.html">good</a> together with <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/bad_sequence_fastqc.html">bad</a> Illumina data provided by the <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/"><code class="docutils literal notranslate"><span class="pre">FastQC</span></code> manual webpage</a>, along with the tutorials and descriptions from <a class="reference external" href="https://rtsf.natsci.msu.edu/genomics/tech-notes/fastqc-tutorial-and-faq/">the RTSF at MSU</a>, <a class="reference external" href="https://hbctraining.github.io/Intro-to-rnaseq-hpc-salmon/lessons/qc_fastqc_assessment.html">the HBC training program</a>, and <a class="reference external" href="https://sequencing.qcfail.com/software/fastqc/">the QC Fail website</a> to demonstrate the modules in the <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> report. The detailed description of these modules can be found on the <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">manual webpage</a>. Although these tutorials are not explicitly made for single-cell data, many of the results are still relevant for single-cell data, with a few caveats described below. In the toggle section, all graphs, except specifically mentioned, are taken from the example reports on the <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">manual webpage</a>. Finally, most of the metrics reported in such quality control reports make sense only for the “biological” reads of a single-cell dataset (i.e. the read being drawn from gene transcripts). For example, for 10x Chromium v2 and v3 datasets, this would be read 2 (the files containing <code class="docutils literal notranslate"><span class="pre">R2</span></code> in their filename). This is because the technical reads are comprised primarily of barcode and UMI sequences which are not drawn from the underlying transcriptome and which we do not expect to have typical biologically plausible sequence or GC content, though certain metrics like the fraction of <code class="docutils literal notranslate"><span class="pre">N</span></code> base calls are still relevant.</p>
<div class="toggle docutils container">
<p><strong>0. Summary</strong></p>
<p>The summary panel on the left-hand side of the HTML report shows the module names included in the report, together with a sign used for quick evaluation of the module results. However, because <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> uses uniform thresholds for files generated from all kinds of sequencing platforms and underlying biological material, we sometimes see warnings (orange exclamation) and failures(red crosses) for good data and passes (green ticks) for questionable data. So, all modules should be carefully evaluated before concluding about the data quality.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-summary">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/summary.jpg"><img alt="Summary" class="bg-primary mb-1" src="../_images/summary.jpg" style="width: 300px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3.2 </span><span class="caption-text">The summary panel of a bad example.</span><a class="headerlink" href="#raw-proc-fig-fastqc-summary" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p><strong>1. Basic Statistics</strong></p>
<p>The basic statistics module contains the overview information and statistics of the reads in the input FASTQ file, such as the filename, total number of sequences, number of poor quality sequences, sequence length, and the overall %GC of all bases in all sequences. Good single-cell data usually have a very little number of poor quality sequences and most often have uniform sequence length. The GC content should match the overall GC content of the genome or transcriptome of the sequenced species.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-basic-statistics">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/basic_statistics.jpg"><img alt="Basic Statistics" class="bg-primary mb-1" src="../_images/basic_statistics.jpg" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3.3 </span><span class="caption-text">A good basic statistics report example.</span><a class="headerlink" href="#raw-proc-fig-fastqc-basic-statistics" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p><strong>2. Per Base Sequence Quality</strong></p>
<p>The per base sequence quality view contains a BoxWhisker type plot for each position in the read, which shows the range of quality scores across all bases at each position along all reads in the file. The x-axis represents the positions in the read, and the y-axis shows the quality scores. For single-cell data of good quality, all yellow boxes in the view, which represent the inter-quantile range of the quality scores of positions, should fall into the green (calls of good quality) area. So do all the whiskers, which represent the 10% and 90% of the distribution. It is not unexpected to see that the quality scores drop along the body of the reads, and some base calls of the last positions fall into the orange (calls of reasonable quality) area because of the decrease in the signal-to-noise ratio that is common in most sequencing-by-synthesis methods. Still, boxes should fall outside of the red (calls of poor quality) area. If poor quality calls are observed, one may consider performing quality trimming of their reads. <a class="reference external" href="https://hbctraining.github.io/Intro-to-rnaseq-hpc-salmon/lessons/qc_fastqc_assessment.html">A detailed explanation</a> of the sequencing error profiles is provided by the <a class="reference external" href="https://hbctraining.github.io/main/">HBC training program</a>.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-per-read-sequence-quality">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/per_read_sequence_quality.jpg"><img alt="per read sequence quality" class="bg-primary mb-1" src="../_images/per_read_sequence_quality.jpg" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3.4 </span><span class="caption-text">A good (left) and a bad (right) per-read sequence quality graph.</span><a class="headerlink" href="#raw-proc-fig-fastqc-per-read-sequence-quality" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p><strong>3. Per Tile Sequence Quality</strong></p>
<p>Using an Illumina library, the per tile sequence quality plot shows the deviation from the average quality for the reads in each flowcell tile that was sequenced. The “hotter” the color, the larger the deviation. For high-quality data, we should see blue over the entire plot, meaning that the flowcell has consistent quality in all tiles. If only part of a flowcell has poor quality, some hot colors will appear in the plot. In that case, the sequencing step might have encountered transient problems, such as bubbles going through the flowcell or smudges and debris inside the flowcell lane. For further diagnoses, please refer to <a class="reference external" href="https://sequencing.qcfail.com/articles/position-specific-failures-of-flowcells/">QC Fail</a> and <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/12%20Per%20Tile%20Sequence%20Quality.html">common reasons for warnings</a> from <code class="docutils literal notranslate"><span class="pre">FastQC</span></code>.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-per-tile-sequence-quality">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/per_tile_sequence_quality.jpg"><img alt="per tile sequence quality" class="bg-primary mb-1" src="../_images/per_tile_sequence_quality.jpg" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3.5 </span><span class="caption-text">A good (left) and a bad (right) per tile sequence quality view.</span><a class="headerlink" href="#raw-proc-fig-fastqc-per-tile-sequence-quality" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p><strong>4. Per Sequence Quality Scores</strong></p>
<p>The per sequence quality score plot shows the distribution of the average quality score of each read in the file. The x-axis shows the average quality scores, and the y-axis represents the frequency of each quality score. For good data, this plot should have only one peak at the tail. If some other peaks show up, a subset of reads might have some quality issue.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-per-sequence-quality-scores">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/per_sequence_quality_scores.jpg"><img alt="per sequence quality scores" class="bg-primary mb-1" src="../_images/per_sequence_quality_scores.jpg" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3.6 </span><span class="caption-text">A good (left) and a bad (right) per sequence quality score plot.</span><a class="headerlink" href="#raw-proc-fig-fastqc-per-sequence-quality-scores" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p><strong>5. Per Base Sequence Content</strong></p>
<p>The per base sequence content plot reports the percent of each base position of all reads in the file for which each of the four nucleotides has been called. For single-cell data, it is not uncommon to see fluctuations at the start of the lines because the first bases of reads represent the sequence of the priming sites, which may not be perfectly random, as explained on the <a class="reference external" href="https://sequencing.qcfail.com/articles/positional-sequence-bias-in-random-primed-libraries/">QC Fail website</a>. This happens frequently in RNA-seq libraries, even though <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> will call a warning or failure.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-per-base-sequence-content">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/per_base_sequence_content.jpg"><img alt="per base sequence content" class="bg-primary mb-1" src="../_images/per_base_sequence_content.jpg" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3.7 </span><span class="caption-text">A good (left) and bad (right) per base sequence content plot.</span><a class="headerlink" href="#raw-proc-fig-fastqc-per-base-sequence-content" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p><strong>6. Per Sequence GC Content</strong></p>
<p>The per sequence GC content plot shows the distribution of GC content over all of the reads in red and a theoretical (expected) distribution in blue. The central peak of the observed distribution should correspond to the overall GC content of the underlying transcriptome. Sometimes we see a wider or narrower distribution than the theoretical distribution because the GC content of the transcriptome might differ from the genome, which, in theory, should follow the distribution shown in blue. Such differences in the spread of the distributions are not uncommon, even though it may trigger a warning or failure. However, a complex distribution usually indicates a contaminated library. It is worth noting, however, that a GC content plot can be somewhat complicated to interpret in transcriptomics experiments, as the expected GC content distribution depends not only on the sequence content of the underlying transcriptome, but also on the expression of the genes in the sample, which are unknown.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-per-sequence-gc-content">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/per_sequence_gc_content.jpg"><img alt="Per Sequence GC Content" class="bg-primary mb-1" src="../_images/per_sequence_gc_content.jpg" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3.8 </span><span class="caption-text">A good (left) and a bad (right) per sequence GC content plot. The plot on the left is from <a class="reference external" href="https://rtsf.natsci.msu.edu/genomics/tech-notes/fastqc-tutorial-and-faq/">the RTSF at MSU</a>. The plot on the right is taken from <a class="reference external" href="https://hbctraining.github.io/Intro-to-rnaseq-hpc-salmon/lessons/qc_fastqc_assessment.html">the HBC training program</a>.</span><a class="headerlink" href="#raw-proc-fig-fastqc-per-sequence-gc-content" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p><strong>7. Per Base N Content</strong></p>
<p>The per base N content plot shows the percent of bases at each position for which an <code class="docutils literal notranslate"><span class="pre">N</span></code> was called, which will occur when the sequencer has insufficient confidence to make a base call. In a high-quality library, we should not expect any noticeable non-zero <code class="docutils literal notranslate"><span class="pre">N</span></code> content throughout the line.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-per-base-n-content">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/per_base_n_content.jpg"><img alt="Per Base N Content" class="bg-primary mb-1" src="../_images/per_base_n_content.jpg" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3.9 </span><span class="caption-text">A good (left) and a bad (right) per base N content plot.</span><a class="headerlink" href="#raw-proc-fig-fastqc-per-base-n-content" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p><strong>8. Sequence Length Distribution</strong></p>
<p>The Sequence length distribution graph shows the distribution of the read lengths. In most single-cell chemistries, all reads will be of the same length. If trimming was performed before quality assessment, there may be some small variation in read lengths.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-sequence-length-distribution">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/sequence_length_distribution.jpg"><img alt="Sequence Length Distribution" class="bg-primary mb-1" src="../_images/sequence_length_distribution.jpg" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3.10 </span><span class="caption-text">A good (left) and a bad (right) sequence length distribution plot.</span><a class="headerlink" href="#raw-proc-fig-fastqc-sequence-length-distribution" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p><strong>9. Sequence Duplication Levels</strong></p>
<p>The sequence duplication level plot shows the distribution of the degree of duplication for read sequences (the blue line) and those after deduplication. As most single-cell platform requires multiple rounds of PCR, highly expressed genes usually express a large number of transcripts, and FastQC itself is not UMI aware, it is common that a small subset of sequences may have a large number of duplications. This may trigger a warning or failure for this module, but it does not necessarily represent a quality problem with the data. Still, the majority of sequences should have a low duplication level.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-sequence-duplication-levels">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/sequence_duplication_levels.jpg"><img alt="Sequence Duplication Levels" class="bg-primary mb-1" src="../_images/sequence_duplication_levels.jpg" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3.11 </span><span class="caption-text">A good (left) and a bad (right) per sequence duplication levels plot.</span><a class="headerlink" href="#raw-proc-fig-fastqc-sequence-duplication-levels" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p><strong>10. Overrepresented Sequences</strong></p>
<p>The overrepresented sequences module lists all read sequences that make up more than <span class="math notranslate nohighlight">\(0.1\%\)</span> of the total number of sequences. We might see some overrepresented sequences from the highly expressed genes after PCR amplification, but most sequences should not be overrepresented. Note that if the possible source of the overrepresented sequences is not “No Hit”, it might indicate that the library is contaminated by the corresponding type of source.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-overrepresented-sequences">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/overrepresented_sequences.jpg"><img alt="Overrepresented Sequences" class="bg-primary mb-1" src="../_images/overrepresented_sequences.jpg" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3.12 </span><span class="caption-text">An overrepresented sequence table.</span><a class="headerlink" href="#raw-proc-fig-fastqc-overrepresented-sequences" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p><strong>11. Adapter Content</strong></p>
<p>The adapter content module shows the cumulative percentage of the fraction of reads in which each of the adapter sequences has been observed at each base position. Ideally, we should not see any abundant adapter sequences in the data.</p>
<figure class="align-default" id="raw-proc-fig-fastqc-adapter-content">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/adapter_content.jpg"><img alt="Adapter Content" class="bg-primary mb-1" src="../_images/adapter_content.jpg" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3.13 </span><span class="caption-text">A good (left) and a bad (right) per sequence quality score plot. The plot on the right is from <a class="reference external" href="https://sequencing.qcfail.com/articles/read-through-adapters-can-appear-at-the-ends-of-sequencing-reads/">the QC Fail website</a>.</span><a class="headerlink" href="#raw-proc-fig-fastqc-adapter-content" title="Link to this image">#</a></p>
</figcaption>
</figure>
</div>
<p>Multiple FastQC reports can be combined into a single report using the tool <a class="reference external" href="https://multiqc.info"><code class="docutils literal notranslate"><span class="pre">MultiQC</span></code></a>.</p>
</section>
<section id="alignment-and-mapping">
<span id="raw-proc-aln-map"></span><h2><span class="section-number">3.2. </span>Alignment and mapping<a class="headerlink" href="#alignment-and-mapping" title="Link to this heading">#</a></h2>
<p>Mapping or alignment is a fundamental step in single-cell raw data processing. It refers to the process of determining the potential loci of origin of each sequenced fragment (e.g., the set of genomic or transcriptomic loci that are similar to the read). Depending on the sequencing protocol, the resulting raw sequence file contains the cell-level information, commonly known as cell barcodes (CB), the unique molecule identifier (UMI), and the raw cDNA sequence (read sequence) generated from the molecule. As the first step in the raw data processing of a single-cell sample (<a class="reference internal" href="#raw-proc-fig-overview"><span class="std std-numref">Fig. 3.1</span></a>), executing mapping or alignment accurately is instrumental for all downstream analyses, since incorrect read-to-transcript/gene mapping can lead to wrong or inaccurate matrices.</p>
<p>While mapping read sequences to reference sequences <em>far</em> predates the advent of scRNA-seq, the current and quickly growing scale of scRNA-seq samples (typically hundreds of millions to billions of reads) makes this stage particularly computationally intensive. Additionally, using pre-existing RNA-seq aligners that are agnostic to any specific scRNA-seq protocol – like the length and position of cell barcodes, UMI, etc. – requires making use of separate tools for performing other steps such as demultiplexing and UMI resolutions <span id="id4">[<a class="reference internal" href="#id110" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>]</span>. This added overhead can be computationally cumbersome. Further, it typically carries a substantial disk space requirement for storing intermediate files, and the extra input and output operations required for processing such intermediate files further increase runtime requirements.</p>
<p>To this end, several dedicated tools have been built specifically for aligning or mapping scRNA-seq data, which handle these additional processing requirements automatically and/or internally. Tools such as <code class="docutils literal notranslate"><span class="pre">Cell</span> <span class="pre">Ranger</span></code> (commercial software from 10x Genomics) <span id="id5">[<a class="reference internal" href="#id151" title="Grace X. Y. Zheng, Jessica M. Terry, Phillip Belgrader, Paul Ryvkin, Zachary W. Bent, Ryan Wilson, Solongo B. Ziraldo, Tobias D. Wheeler, Geoff P. McDermott, Junjie Zhu, Mark T. Gregory, Joe Shuga, Luz Montesclaros, Jason G. Underwood, Donald A. Masquelier, Stefanie Y. Nishimura, Michael Schnall-Levin, Paul W. Wyatt, Christopher M. Hindson, Rajiv Bharadwaj, Alexander Wong, Kevin D. Ness, Lan W. Beppu, H. Joachim Deeg, Christopher McFarland, Keith R. Loeb, William J. Valente, Nolan G. Ericson, Emily A. Stevens, Jerald P. Radich, Tarjei S. Mikkelsen, Benjamin J. Hindson, and Jason H. Bielas. Massively parallel digital transcriptional profiling of single cells. Nature Communications, 8(1):14049, Jan 2017. URL: https://doi.org/10.1038/ncomms14049, doi:10.1038/ncomms14049.">Zheng <em>et al.</em>, 2017</a>]</span>, <code class="docutils literal notranslate"><span class="pre">zUMIs</span></code> <span id="id6">[<a class="reference internal" href="#id121" title="Swati Parekh, Christoph Ziegenhain, Beate Vieth, Wolfgang Enard, and Ines Hellmann. zUMIs - a fast and flexible pipeline to process term`RNA` sequencing data with UMIs. GigaScience, May 2018. URL: https://doi.org/10.1093/gigascience/giy059, doi:10.1093/gigascience/giy059.">Parekh <em>et al.</em>, 2018</a>]</span>, <code class="docutils literal notranslate"><span class="pre">alevin</span></code> <span id="id7">[<a class="reference internal" href="#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>, <code class="docutils literal notranslate"><span class="pre">RainDrop</span></code> <span id="id8">[<a class="reference internal" href="#id98" title="Stefan Niebler, André Müller, Thomas Hankeln, and Bertil Schmidt. RainDrop: Rapid activation matrix computation for droplet-based single-cell term`RNA`-seq reads. BMC bioinformatics, 21(1):1–14, 2020.">Niebler <em>et al.</em>, 2020</a>]</span>, <code class="docutils literal notranslate"><span class="pre">kallisto|bustools</span></code> <span id="id9">[<a class="reference internal" href="#id116" title="Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. Nature Biotechnology, 39(7):813–818, April 2021. URL: https://doi.org/10.1038/s41587-021-00870-2, doi:10.1038/s41587-021-00870-2.">Melsted <em>et al.</em>, 2021</a>]</span>, <code class="docutils literal notranslate"><span class="pre">STARsolo</span></code> <span id="id10">[<a class="reference internal" href="#id115" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>]</span> and <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> <span id="id11">[<a class="reference internal" href="#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>]</span> provide dedicated treatment for aligning scRNA-seq reads along with parsing of technical read content (CBs and UMIs), as well as methods for demultiplexing and UMI resolution.
While all provide relatively simplified interfaces to the user, these tools use a variety of different approaches internally, with some generating traditional intermediate files (e.g., BAM files) and subsequently processing them, and others either working entirely in memory or by making use of reduced intermediate representations to reduce the input/output burden.</p>
<p>While the specific algorithms, data structures, and time and space trade-offs made by different alignment and mapping approaches can vary greatly, we can roughly categorize existing approaches along two axes:</p>
<ul class="simple">
<li><p>The type of mapping they perform and</p></li>
<li><p>The type of reference sequence against which they map reads.</p></li>
</ul>
<section id="types-of-mapping">
<span id="raw-proc-types-of-mapping"></span><h3><span class="section-number">3.2.1. </span>Types of mapping<a class="headerlink" href="#types-of-mapping" title="Link to this heading">#</a></h3>
<p>Here we consider three main types of mapping algorithms that are most commonly applied in the context of mapping sc/snRNA-seq data: spliced alignment, contiguous alignment, and variations of lightweight mapping.</p>
<p>First, let us draw a distinction here between alignment-based approaches and lightweight mapping-based approaches (<a class="reference internal" href="#raw-proc-fig-alignment-mapping"><span class="std std-numref">Fig. 3.14</span></a>). Alignment approaches apply a range of different heuristics to determine the potential loci from which reads may arise and then subsequently score, at each putative locus, the best nucleotide-level alignment between the read and reference, typically using dynamic programming-based approaches. Using dynamic programming algorithms to solve the alignment problem has a long and rich history. The exact type of dynamic programming algorithm used depends on the type of alignment being sought. <a class="reference external" href="https://en.wikipedia.org/wiki/Needleman%E2%80%93Wunsch_algorithm">Global alignment</a> is applicable for the case where the entirety of the query and reference sequence are to be aligned. On the other hand, <a class="reference external" href="https://en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm">local alignment</a> is applicable when, possibly, only a subsequence of the query is expected to match a subsequence of the reference. Typically, the models most applicable for short-read alignment are neither fully global nor fully local, but fall into a category of semi-global alignment where the majority of the query is expected to align to some substring of the reference (this is often called “fitting” alignment). Moreover, it is sometimes common to allow soft-clipping of the alignment so that the penalties for mismatches, insertions, or deletions appearing at the very start or end of the read are ignored or down-weighted. This is commonly done using <a class="reference external" href="https://github.com/smarco/WFA2-lib#-33-alignment-span">“extension” alignment</a>. Though these modifications change the specific rules used in the dynamic programming recurrence and traceback, they do not fundamentally change its overall complexity.</p>
<p>Apart from these general alignment techniques, a number of more sophisticated modifications and heuristics have been designed to make the alignment process more practical and efficient in the context of aligning genomic sequencing reads. For example, <code class="docutils literal notranslate"><span class="pre">banded</span> <span class="pre">alignment</span></code> <span id="id12">[<a class="reference internal" href="#id96" title="Kun-Mao Chao, William R. Pearson, and Webb Miller. Aligning two sequences within a specified diagonal band. Bioinformatics, 8(5):481–487, 1992. URL: https://doi.org/10.1093/bioinformatics/8.5.481, doi:10.1093/bioinformatics/8.5.481.">Chao <em>et al.</em>, 1992</a>]</span> is a popular heuristic included in the alignment implementation of many popular tools, which avoids computing large parts of the dynamic programming table if the user is uninterested in alignment scores below a certain threshold. Likewise, other heuristics like X-drop <span id="id13">[<a class="reference internal" href="#id108" title="Zheng Zhang, Scott Schwartz, Lukas Wagner, and Webb Miller. A greedy algorithm for aligning term`DNA` sequences. Journal of Computational Biology, 7(1-2):203–214, February 2000. URL: https://doi.org/10.1089/10665270050081478, doi:10.1089/10665270050081478.">Zhang <em>et al.</em>, 2000</a>]</span> and Z-drop <span id="id14">[<a class="reference internal" href="#id97" title="Heng Li. Minimap2: pairwise alignment for nucleotide sequences. Bioinformatics, 34(18):3094–3100, May 2018. URL: https://doi.org/10.1093/bioinformatics/bty191, doi:10.1093/bioinformatics/bty191.">Li, 2018</a>]</span> are popular for pruning un-promising alignments early. Recent algorithmic progress, such as wavefront alignment <span id="id15">[<a class="reference internal" href="#id107" title="Santiago Marco-Sola, Jordan M Eizenga, Andrea Guarracino, Benedict Paten, Erik Garrison, and Miquel Moreto. Optimal gap-affine alignment in o(s) space. bioRxiv, April 2022. URL: https://doi.org/10.1101/2022.04.14.488380, doi:10.1101/2022.04.14.488380.">Marco-Sola <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id105" title="Santiago Marco-Sola, Juan Carlos Moure, Miquel Moreto, and Antonio Espinosa. Fast gap-affine pairwise alignment using the wavefront algorithm. Bioinformatics, September 2020. URL: https://doi.org/10.1093/bioinformatics/btaa777, doi:10.1093/bioinformatics/btaa777.">Marco-Sola <em>et al.</em>, 2020</a>]</span>, allows for determining optimal alignments in asymptotically (and practically) shorter time and smaller space if high-scoring (low divergence) alignments exist. Finally, considerable effort has also been devoted to optimizing data layout and computation in a way that takes advantage of instruction-level parallelism <span id="id16">[<a class="reference internal" href="#id100" title="Michael Farrar. Striped Smith–Waterman speeds database searches six times over other SIMD implementations. Bioinformatics, 23(2):156–161, 2007.">Farrar, 2007</a>, <a class="reference internal" href="#id101" title="Torbjørn Rognes and Erling Seeberg. Six-fold speed-up of Smith–Waterman sequence database searches using parallel processing on common microprocessors. Bioinformatics, 16(8):699–706, 2000.">Rognes and Seeberg, 2000</a>, <a class="reference internal" href="#id99" title="Andrzej Wozniak. Using video-oriented instructions to speed up sequence comparison. Bioinformatics, 13(2):145–150, 1997.">Wozniak, 1997</a>]</span>,and to expressing the alignment dynamic programming recurrences in a manner that is highly amenable to data parallelism and vectorization (e.g., as in the difference encoding of <span id="id17">Suzuki and Kasahara [<a class="reference internal" href="#id106" title="Hajime Suzuki and Masahiro Kasahara. Introducing difference recurrence relations for faster semi-global alignment of long sequences. BMC Bioinformatics, February 2018. URL: https://doi.org/10.1186/s12859-018-2014-8, doi:10.1186/s12859-018-2014-8.">2018</a>]</span>). Most widely-used alignment tools make use of highly-optimized and vectorized alignment implementations.</p>
<p>In addition to the alignment score, a backtrace of the actual alignment that yields this score may be obtained. Such information is typically encoded as a <code class="docutils literal notranslate"><span class="pre">CIGAR</span></code> string (short for “Concise Idiosyncratic Gapped Alignment Report”), a compressed alphanumeric representation of the alignment, within the SAM or BAM file that is output. An example of a <code class="docutils literal notranslate"><span class="pre">CIGAR</span></code> string is <code class="docutils literal notranslate"><span class="pre">3M2D4M</span></code>, which represents that the alignment has three matches or mismatches, followed by a deletion of length two from the read (bases present in the reference but not the read), followed by four more matches or mismatches. Other variants of the <code class="docutils literal notranslate"><span class="pre">CIGAR</span></code> string can also delineate between matches or mismatches. For example, <code class="docutils literal notranslate"><span class="pre">3=2D2=2X</span></code> is a valid extended <code class="docutils literal notranslate"><span class="pre">CIGAR</span></code> string encoding the same alignment as just described, except that it makes clear that the three bases before the deletion constitute matches and that after the deletion, there are two matched bases followed by two mismatched bases. A detailed description of the <code class="docutils literal notranslate"><span class="pre">CIGAR</span></code> string format can be found in <a class="reference external" href="https://samtools.github.io/hts-specs/SAMv1.pdf">the SAMtools manual</a> or <a class="reference external" href="https://genome.sph.umich.edu/wiki/SAM#What_is_a_CIGAR.3F">the SAM wiki page of UMICH</a>.</p>
<p>At the cost of computing such scores, alignment-based approaches know the quality of each potential mapping of a read, which they can then use to filter reads that align well to the reference from mappings that arise as the result of low complexity or “spurious” matches between the read and reference. Alignment-based approaches include both traditional “full-alignment” approaches, as implemented in tools such as <code class="docutils literal notranslate"><span class="pre">STAR</span></code><span id="id18">[<a class="reference internal" href="#id149" title="Alexander Dobin, Carrie A Davis, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas R Gingeras. Star: ultrafast universal rna-seq aligner. Bioinformatics, 29(1):15–21, 2013.">Dobin <em>et al.</em>, 2013</a>]</span> and <code class="docutils literal notranslate"><span class="pre">STARsolo</span></code><span id="id19">[<a class="reference internal" href="#id115" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>]</span> as well as approaches like <em>selective-alignment</em> implemented in <code class="docutils literal notranslate"><span class="pre">salmon</span></code><span id="id20">[<a class="reference internal" href="#id139" title="Avi Srivastava, Laraib Malik, Hirak Sarkar, Mohsen Zakeri, Fatemeh Almodaresi, Charlotte Soneson, Michael I Love, Carl Kingsford, and Rob Patro. Alignment and mapping methodology influence transcript abundance estimation. Genome Biology, 21(1):1–29, 2020.">Srivastava <em>et al.</em>, 2020</a>]</span> and <code class="docutils literal notranslate"><span class="pre">alevin</span></code><span id="id21">[<a class="reference internal" href="#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>, which score mappings but omit the computation of the optimal alignment’s backtrace.</p>
<figure class="align-default" id="raw-proc-fig-alignment-mapping">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/alignment_vs_mapping.png"><img alt="Alignment vs Mapping" class="bg-primary mb-1" src="../_images/alignment_vs_mapping.png" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3.14 </span><span class="caption-text">An abstract overview of the alignment-based method and lightweight mapping-based method.</span><a class="headerlink" href="#raw-proc-fig-alignment-mapping" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>Alignment-based approaches can be further categorized into spliced-alignment and contiguous-alignment approaches (currently, there are no lightweight-mapping approaches that perform spliced mapping). Spliced-alignment approaches are capable of aligning a sequence read over several distinct segments of a reference, allowing potentially large gaps between the regions of the reference that align well with the corresponding parts of the read. These alignment approaches are typically applied when aligning RNA-seq reads to the genome, since the alignment procedure must be able to align reads that span across one or more splice junctions of the transcript, where a sequence that is contiguous in the read may be separated by intron and exon subsequences (potentially spanning kilobases of sequence) in the reference. Spliced alignment is a challenging problem, particularly in cases where only a small region of the read spans a splicing junction, since very little informative sequence may be available to place the segment of the read that overhangs the splice junction by only a small amount. On the other hand, contiguous alignment seeks a contiguous substring of the reference that aligns well against the read. In such alignment problems, though small insertions and deletions may be allowed, large gaps such as those observed in spliced alignments are typically not tolerated.</p>
<p>Finally, we can draw a distinction between alignment-based methods such as those described above and lightweight-mapping methods, which include approaches such as pseudoalignment <span id="id22">[<a class="reference internal" href="#id135" title="Nicolas L Bray, Harold Pimentel, Páll Melsted, and Lior Pachter. Near-optimal probabilistic term`RNA`-seq quantification. Nature biotechnology, 34(5):525–527, 2016.">Bray <em>et al.</em>, 2016</a>]</span>, quasi-mapping <span id="id23">[<a class="reference internal" href="#id134" title="Avi Srivastava, Hirak Sarkar, Nitish Gupta, and Rob Patro. Rapmap: a rapid, sensitive and accurate tool for mapping term`rna`-seq reads to transcriptomes. Bioinformatics, 32(12):i192–i200, 2016.">Srivastava <em>et al.</em>, 2016</a>]</span>, and pseudoalignment with structural constraints <span id="id24">[<a class="reference internal" href="#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>]</span>. Such approaches generally achieve superior speed by avoiding nucleotide-level alignment between the read and reference sequences. Instead, they base the determination of the reported mapping loci of a read on a separate set of rules and heuristics that may look only at the set of matching k-mers or other types of exact matches (e.g., via a consensus rule) and, potentially, their orientations and relative positions with respect to each other on both the read and reference (e.g., chaining). This can lead to substantially increased speed and mapping throughput, but also precludes easily-interpretable score-based assessments of whether or not the matches between the read and reference constitute a good match capable of supporting a high-quality alignment.</p>
</section>
<section id="mapping-against-different-reference-sequences">
<span id="raw-proc-mapping-references"></span><h3><span class="section-number">3.2.2. </span>Mapping against different reference sequences<a class="headerlink" href="#mapping-against-different-reference-sequences" title="Link to this heading">#</a></h3>
<p>While different choices can be made among the mapping algorithms themselves, different choices can <em>also</em> be made about the reference against which the read is mapped. We consider three main categories of reference sequence against which a method might map:</p>
<ul class="simple">
<li><p>The full (likely annotated) reference genome</p></li>
<li><p>The annotated transcriptome</p></li>
<li><p>An augmented transcriptome</p></li>
</ul>
<p>It is also worth noting that, currently, not all combinations of mapping algorithms and reference sequences are possible. For example, lightweight mapping-based algorithms do not currently exist that can perform spliced mapping of reads against a reference genome.</p>
<section id="mapping-to-the-full-genome">
<span id="raw-proc-genome-mapping"></span><h4><span class="section-number">3.2.2.1. </span>Mapping to the full genome<a class="headerlink" href="#mapping-to-the-full-genome" title="Link to this heading">#</a></h4>
<p>The first type of reference, against which a method may map, consists of the entire genome of the target organism, usually with the annotated transcripts considered during mapping. Tools like <code class="docutils literal notranslate"><span class="pre">zUMIs</span></code> <span id="id25">[<a class="reference internal" href="#id121" title="Swati Parekh, Christoph Ziegenhain, Beate Vieth, Wolfgang Enard, and Ines Hellmann. zUMIs - a fast and flexible pipeline to process term`RNA` sequencing data with UMIs. GigaScience, May 2018. URL: https://doi.org/10.1093/gigascience/giy059, doi:10.1093/gigascience/giy059.">Parekh <em>et al.</em>, 2018</a>]</span>, <code class="docutils literal notranslate"><span class="pre">Cell</span> <span class="pre">Ranger</span></code> <span id="id26">[<a class="reference internal" href="#id151" title="Grace X. Y. Zheng, Jessica M. Terry, Phillip Belgrader, Paul Ryvkin, Zachary W. Bent, Ryan Wilson, Solongo B. Ziraldo, Tobias D. Wheeler, Geoff P. McDermott, Junjie Zhu, Mark T. Gregory, Joe Shuga, Luz Montesclaros, Jason G. Underwood, Donald A. Masquelier, Stefanie Y. Nishimura, Michael Schnall-Levin, Paul W. Wyatt, Christopher M. Hindson, Rajiv Bharadwaj, Alexander Wong, Kevin D. Ness, Lan W. Beppu, H. Joachim Deeg, Christopher McFarland, Keith R. Loeb, William J. Valente, Nolan G. Ericson, Emily A. Stevens, Jerald P. Radich, Tarjei S. Mikkelsen, Benjamin J. Hindson, and Jason H. Bielas. Massively parallel digital transcriptional profiling of single cells. Nature Communications, 8(1):14049, Jan 2017. URL: https://doi.org/10.1038/ncomms14049, doi:10.1038/ncomms14049.">Zheng <em>et al.</em>, 2017</a>]</span> and <code class="docutils literal notranslate"><span class="pre">STARsolo</span></code> <span id="id27">[<a class="reference internal" href="#id115" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>]</span> take this approach. Since many of the reads arise from spliced transcripts, such an approach also requires the use of a splice-aware alignment algorithm where the alignment for a read can be split across one or more splicing junctions. The main benefits of this approach are that reads arising from any location in the genome, not just from annotated transcripts, can be accounted for. Since these approaches require constructing a genome-wide index, there is little marginal cost for reporting not only the reads that map to known spliced transcripts but also reads that overlap or align within introns, making the alignment cost when using such approaches very similar for both single-cell and single-nucleus data. A final benefit is that even reads residing outside of the annotated transcripts, exons, and introns can be accounted for by such methods, which can enable <em>post hoc</em> augmentation of the set of quantified loci (e.g., as is done by <span id="id28">Pool <em>et al.</em> [<a class="reference internal" href="#id142" title="Allan-Hermann Pool, Helen Poldsam, Sisi Chen, Matt Thomson, and Yuki Oka. Enhanced recovery of single-cell term`RNA`-sequencing reads for missing gene expression data. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449, arXiv:https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449.full.pdf, doi:10.1101/2022.04.26.489449.">2022</a>]</span> by adding expressed UTR extensions to transcript annotations in a sample-specific and data-driven manner) and potentially increase gene detection and quantification sensitivity.</p>
<p>On the other hand, the versatility of the strategy of performing spliced alignment against the full genome comes with certain trade-offs. First, the most commonly-used alignment tools that adopt this strategy in the single-cell space have substantial memory requirements. Due to its speed and versatility, most of these tools are based upon the STAR <span id="id29">[<a class="reference internal" href="#id149" title="Alexander Dobin, Carrie A Davis, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas R Gingeras. Star: ultrafast universal rna-seq aligner. Bioinformatics, 29(1):15–21, 2013.">Dobin <em>et al.</em>, 2013</a>]</span> aligner. Yet, for a human-scale genome, constructing and storing the index can require upwards of <span class="math notranslate nohighlight">\(32\)</span> GB of memory. The use of a sparse <a class="reference external" href="https://en.wikipedia.org/wiki/Suffix_array">suffix array</a> can reduce the final index size by close to a factor of <span class="math notranslate nohighlight">\(2\)</span>, but this comes at a reduction in alignment speed, and it still requires larger memory for the initial construction. Second, the increased difficulty of spliced alignment compared to contiguous alignment and the fact that current spliced-alignment tools must explicitly compute a score for each read alignment, means that this approach comes at an increased computational cost compared to the alternatives. Finally, such an approach requires the genome of the organism under study is available. While this is not problematic for the most commonly-studied reference organisms and is rarely an issue, it can make using such tools difficult for non-model organisms where only a transcriptome assembly may be available.</p>
</section>
<section id="mapping-to-the-spliced-transcriptome">
<span id="raw-proc-txome-mapping"></span><h4><span class="section-number">3.2.2.2. </span>Mapping to the spliced transcriptome<a class="headerlink" href="#mapping-to-the-spliced-transcriptome" title="Link to this heading">#</a></h4>
<p>To reduce the computational overhead of spliced alignment to a genome, the widely-adopted alternative is to use just the sequences of the annotated transcripts themselves as a reference. Since the majority of single-cell experiments are generated from model organisms (such as mouse or human), which have well-annotated transcriptomes, such transcriptome-based quantification methods may provide similar read coverage to their genome-based counterparts. Compared to the genome, transcriptome sequences are much smaller in size, minimizing the computational resources required for running the mapping pipeline. Additionally, since the relevant splicing patterns are already represented in the transcript sequences themselves, this approach dispenses with the need to solve the difficult spliced-alignment problem. Instead, one can simply search for contiguous alignments or mappings for the read. Since only contiguous mappings need to be discovered, both alignment and lightweight mapping techniques are amenable to transcriptome references, and both approaches have been used in popular tools that adopt the spliced transcriptome as the target of reference mapping.</p>
<p>However, while such approaches can greatly reduce the memory and time required for alignment and mapping, they will fail to capture reads that arise from outside of the spliced transcriptome. Such an approach is therefore not adequate when processing single-nucleus data. Even in single-cell experiments, reads arising from outside of the spliced transcriptome can constitute a substantial fraction of all data, and there is growing evidence that such reads should be incorporated into subsequent analysis <span id="id30">[<a class="reference internal" href="#id142" title="Allan-Hermann Pool, Helen Poldsam, Sisi Chen, Matt Thomson, and Yuki Oka. Enhanced recovery of single-cell term`RNA`-sequencing reads for missing gene expression data. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449, arXiv:https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449.full.pdf, doi:10.1101/2022.04.26.489449.">Pool <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id112" title="10x Genomics. Technical note - interpreting intronic and antisense reads in 10x genomics single cell gene expression data. https://www.10xgenomics.com/support/single-cell-gene-expression/documentation/steps/sequencing/interpreting-intronic-and-antisense-reads-in-10-x-genomics-single-cell-gene-expression-data, August 2021.">10x Genomics, 2021</a>]</span>. Further, when paired with lightweight-mapping approaches, short matches shared between the spliced transcriptome and the reference sequences that truly give rise to a read may lead to spurious read mappings, which, in turn, can lead to spurious and even biologically implausible estimates of the expression of certain genes <span id="id31">[<a class="reference internal" href="#id141" title="Ralf Schulze Brüning, Lukas Tombor, Marcel H Schulz, Stefanie Dimmeler, and David John. Comparative analysis of common alignment tools for single-cell term`RNA` sequencing. GigaScience, 2022.">Brüning <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id115" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>]</span>.</p>
</section>
<section id="mapping-to-an-augmented-transcriptome">
<span id="raw-proc-aug-txome-mapping"></span><h4><span class="section-number">3.2.2.3. </span>Mapping to an augmented transcriptome<a class="headerlink" href="#mapping-to-an-augmented-transcriptome" title="Link to this heading">#</a></h4>
<p>To deal with the fact that sequenced reads may arise from outside of spliced transcripts, it is possible to augment the spliced transcript sequences with other reference sequences that may be expected to give rise to reads (e.g., full-length unspliced transcripts or excised intronic sequences). Transcriptome references, when augmented with further sequences such as introns, allow reference indices typically smaller than those required for the full genome while simultaneously retaining the ability to search only for contiguous read alignments. This means they can potentially enable both faster and less memory-hungry alignment than when mapping against the full genome while still accounting for many of the reads that arise from outside of the spliced transcriptome. Finally, by mapping to an expanded collection of reference sequences, not only are the mapping locations of more reads recovered compared to mapping against just the spliced transcriptome, but when using lightweight mapping-based approaches, spurious mappings can be greatly reduced <span id="id32">[<a class="reference internal" href="#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>]</span>. Such an expanded or augmented transcriptome is commonly used among approaches (those that do not map to the full genome) when they need to quantify single-nucleus data or prepare data for RNA-velocity analysis <span id="id33">[<a class="reference internal" href="#id117" title="Charlotte Soneson, Avi Srivastava, Rob Patro, and Michael B Stadler. Preprocessing choices affect RNA velocity results for droplet scRNA-seq data. PLoS computational biology, 17(1):e1008585, 2021.">Soneson <em>et al.</em>, 2021</a>]</span>. Therefore, such augmented references can be constructed for all of the common methods that don’t make use of spliced alignment to the full genome <span id="id34">[<a class="reference internal" href="#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id116" title="Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. Nature Biotechnology, 39(7):813–818, April 2021. URL: https://doi.org/10.1038/s41587-021-00870-2, doi:10.1038/s41587-021-00870-2.">Melsted <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>.</p>
<p><span id="id35">He <em>et al.</em> [<a class="reference internal" href="#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">2022</a>]</span> argue that such an approach is valuable even when processing standard single-cell RNA-seq data and recommend constructing an augmented <em>splici</em> (meaning spliced + intronic) reference for mapping and quantification. The <em>splici</em> reference is constructed using the spliced transcriptome sequence along with the sequences containing the merged intervals corresponding to the introns of the annotated genes. Each reference is then labeled with its annotated splicing status, and the mapping to this reference is subsequently paired with splicing status-aware methods for <a class="reference internal" href="#raw-proc-umi-resolution"><span class="std std-ref">UMI resolution</span></a>. The main benefits of this approach are that it admits the use of lightweight-mapping methods while substantially reducing the prevalence of spurious mappings. It enables the detection of reads of both spliced and unspliced origin, which can increase the sensitivity of subsequent analysis <span id="id36">[<a class="reference internal" href="#id142" title="Allan-Hermann Pool, Helen Poldsam, Sisi Chen, Matt Thomson, and Yuki Oka. Enhanced recovery of single-cell term`RNA`-sequencing reads for missing gene expression data. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449, arXiv:https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449.full.pdf, doi:10.1101/2022.04.26.489449.">Pool <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id112" title="10x Genomics. Technical note - interpreting intronic and antisense reads in 10x genomics single cell gene expression data. https://www.10xgenomics.com/support/single-cell-gene-expression/documentation/steps/sequencing/interpreting-intronic-and-antisense-reads-in-10-x-genomics-single-cell-gene-expression-data, August 2021.">10x Genomics, 2021</a>]</span>, and, since splicing status is tracked during quantification and reported separately in the output, it unifies the processing pipeline for single-cell, single-nucleus, and RNA-velocity preprocessing.</p>
</section>
</section>
</section>
<section id="cell-barcode-correction">
<span id="raw-proc-cb-correction"></span><h2><span class="section-number">3.3. </span>Cell barcode correction<a class="headerlink" href="#cell-barcode-correction" title="Link to this heading">#</a></h2>
<p>Droplet-based single-cell segregation systems, such as those provided by 10x Genomics, have become an important tool for studying the cause and consequences of cellular heterogeneity. In this segregation system, the RNA material of each captured cell is extracted within a water-based droplet encapsulation along with a barcoded bead. These beads tag the RNA content of individual cells with unique oligonucleotides, called cell barcodes (CBs), that are later sequenced along with the fragments of the cDNAs that are reversely transcribed from the RNA content. The beads contain high-diversity DNA barcodes enabling parallel barcoding of the cell’s molecular content and <em>in silico</em> demultiplexing of the sequencing reads into individual cellular bins.</p>
<div class="admonition-a-note-on-alignment-orientation admonition">
<p class="admonition-title">A note on alignment orientation</p>
<p>Depending on the chemistry of the sample being analyzed and the processing options provided by the user, it is not necessarily the case that all sequenced fragments aligning to the reference will be considered for quantification and barcode correction.
One commonly-applied criterion for filtering is alignment orientation. Specifically, certain chemistries specify protocols such that the aligned reads should only derive from (i.e. map back to) the underlying transcripts in a specific orientation.
For example, in 10x Genomics 3’ Chromium chemistries, we expect the biological read to align to the underlying transcript’s forward strand, though anti-sense reads do exist <span id="id37">[<a class="reference internal" href="#id112" title="10x Genomics. Technical note - interpreting intronic and antisense reads in 10x genomics single cell gene expression data. https://www.10xgenomics.com/support/single-cell-gene-expression/documentation/steps/sequencing/interpreting-intronic-and-antisense-reads-in-10-x-genomics-single-cell-gene-expression-data, August 2021.">10x Genomics, 2021</a>]</span>. Thus, mappings of the reads in the reverse-complement orientation to the underlying sequences may be ignored or filtered out at the user’s discretion. If a chemistry follows such a so-called “stranded” protocol, this should be documented.</p>
</div>
<section id="type-of-errors-in-barcoding">
<h3><span class="section-number">3.3.1. </span>Type of errors in barcoding<a class="headerlink" href="#type-of-errors-in-barcoding" title="Link to this heading">#</a></h3>
<p>The tag, sequence, and demultiplex method for single-cell profiling generally works well. However, the number of observed cell barcodes (CBs) in a droplet-based library can significantly differ from the number of originally encapsulated cells by several fold. A few main sources of error can lead to such observation:</p>
<ul class="simple">
<li><p>Doublet / Multiplet: A single barcode can be associated with two or more cells and lead to undercounting of cells.</p></li>
<li><p>Empty Droplet: A droplet can be captured with no encapsulated cell, where the ambient RNA is tagged with the barcode and can be sequenced; this leads to overcounting of cells.</p></li>
<li><p>Sequence error: Errors can arise during the PCR amplification or sequencing process and can contribute to both under and over-counting of the cells.</p></li>
</ul>
<p>Computational tools for demultiplexing the <a class="reference internal" href="../glossary.html#term-RNA"><span class="xref std std-term">RNA</span></a>-seq reads into cell-specific bins use a wide range of diagnostic indicators to filter out artefactual or low-quality data. For example, numerous methods exist for the removal of ambient <a class="reference internal" href="../glossary.html#term-RNA"><span class="xref std std-term">RNA</span></a> contamination <span id="id38">[<a class="reference internal" href="#id123" title="Aaron TL Lun, Samantha Riesenfeld, Tallulah Andrews, Tomas Gomes, John C Marioni, and others. EmptyDrops: distinguishing cells from empty droplets in droplet-based single-cell term`RNA` sequencing data. Genome biology, 20(1):1–9, 2019.">Lun <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id126" title="Walter Muskovic and Joseph E Powell. DropletQC: improved identification of empty droplets and damaged cells in single-cell term`RNA`-seq data. Genome Biology, 22(1):1–9, 2021.">Muskovic and Powell, 2021</a>, <a class="reference internal" href="#id128" title="Matthew D Young and Sam Behjati. SoupX removes ambient term`RNA` contamination from droplet-based single-cell term`RNA` sequencing data. GigaScience, December 2020. URL: https://doi.org/10.1093/gigascience/giaa151, doi:10.1093/gigascience/giaa151.">Young and Behjati, 2020</a>]</span>, doublet detection <span id="id39">[<a class="reference internal" href="#id132" title="Abha S Bais and Dennis Kostka. Scds: computational annotation of doublets in single-cell term`RNA` sequencing data. Bioinformatics, 36(4):1150–1158, September 2019. URL: https://doi.org/10.1093/bioinformatics/btz698, doi:10.1093/bioinformatics/btz698.">Bais and Kostka, 2019</a>, <a class="reference internal" href="#id129" title="Erica A.K. DePasquale, Daniel J. Schnell, Pieter-Jan Van Camp, Íñigo Valiente-Aland\'ı, Burns C. Blaxall, H. Leighton Grimes, Harinder Singh, and Nathan Salomonis. DoubletDecon: deconvoluting doublets from single-cell term`RNA`-sequencing data. Cell Reports, 29(6):1718–1727.e8, November 2019. URL: https://doi.org/10.1016/j.celrep.2019.09.082, doi:10.1016/j.celrep.2019.09.082.">DePasquale <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id130" title="Christopher S. McGinnis, Lyndsay M. Murrow, and Zev J. Gartner. DoubletFinder: doublet detection in single-cell term`RNA` sequencing data using artificial nearest neighbors. Cell Systems, 8(4):329–337.e4, April 2019. URL: https://doi.org/10.1016/j.cels.2019.03.003, doi:10.1016/j.cels.2019.03.003.">McGinnis <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id131" title="Samuel L. Wolock, Romain Lopez, and Allon M. Klein. Scrublet: Computational Identification of Cell Doublets in Single-Cell Transcriptomic Data. Cell Systems, 8(4):281–291.e9, April 2019. URL: https://doi.org/10.1016/j.cels.2018.11.005, doi:10.1016/j.cels.2018.11.005.">Wolock <em>et al.</em>, 2019</a>]</span> and cell barcodes correction for sequence errors based on nucleotide sequence similarity.</p>
<p>Several common strategies are used for cell barcode identification and correction.</p>
<ul class="simple">
<li><p>Correction against a known list of <em>potential</em> barcodes: Certain chemistries, such as 10x Chromium, draw CBs from a known pool of potential barcode sequences. Thus, the set of barcodes observed in any sample is expected to be a subset of this known list, often called a “whitelist”, “permit list”, or “pass list”. In this case, a common strategy is to assume each barcode that exactly matches some element from the known list is correct and for all other barcodes to be correct against the known list of barcodes (i.e., to find barcodes from the permit list that are some small Hamming distance or edit distance away from the observed barcodes). This approach leverages the known permit list to allow efficient correction of many barcodes that have been potentially corrupted. However, one difficulty with this approach is that a given corrupted barcode may have multiple possible corrections in the permit list (i.e., its correction may be ambiguous). In fact, if one considers a barcode that is taken from the <a class="reference external" href="https://teichlab.github.io/scg_lib_structs/data/10X-Genomics/3M-february-2018.txt.gz">10x Chromium v3 permit list</a> and mutated at a single position to a barcode not in the list, there is a <span class="math notranslate nohighlight">\(\sim 81\%\)</span> chance that it sits at Hamming distance <span class="math notranslate nohighlight">\(1\)</span> from two or more barcodes in the permit list. The probability of such collisions can be reduced by only considering correcting against barcodes from the known permit list, which, themselves, occur exactly in the given sample (or even only those that occur exactly in the given sample above some nominal frequency threshold). Also, information such as the base quality at the “corrected” position can be used to potentially break ties in the case of ambiguous corrections. Yet, as the number of assayed cells increases, insufficient sequence diversity in the set of potential cell barcodes increases the frequency of ambiguous corrections, and reads tagged with barcodes having ambiguous corrections are most commonly discarded.</p></li>
<li><p>Knee or elbow-based methods: If a set of potential barcodes is unknown — or even if it is known, but one wishes to correct directly from the observed data itself without consulting an external list — one can adopt a method based on the observation that the list of “true” or high-quality barcodes in a sample is likely those associated with the greatest number of reads.
To do this, one can construct the cumulative frequency plot of the barcodes, in which barcodes are sorted in descending order of the number of distinct reads or UMIs with which they are associated. Often, this ranked cumulative frequency plot will contain a “knee” or “elbow” – an inflection point that can be used to characterize frequently occurring barcodes from infrequent (and therefore likely erroneous) barcodes. Many methods exist for attempting to identify such an inflection point <span id="id40">[<a class="reference internal" href="#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id123" title="Aaron TL Lun, Samantha Riesenfeld, Tallulah Andrews, Tomas Gomes, John C Marioni, and others. EmptyDrops: distinguishing cells from empty droplets in droplet-based single-cell term`RNA` sequencing data. Genome biology, 20(1):1–9, 2019.">Lun <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id110" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>]</span> as a likely point of discrimination between properly captured cells and empty droplets. Subsequently, the set of barcodes that appear “above” the knee can be treated as a permit list against which the rest of the barcodes may be corrected, as in the first method list above. Such an approach is flexible as it can be applied in chemistries that have an external permit list and those that don’t. Further parameters of the knee-finding algorithms can be altered to yield more or less restrictive selected barcode sets. Yet, such an approach can have certain drawbacks, like a tendency to be overly conservative and sometimes failing to work robustly in samples where no clear knee is present.</p></li>
<li><p>Filtering and correction based on an expected cell count provided by the user: These approaches seek to estimate a robust list of high-quality or present barcodes in the cases where the CB frequency distribution may not have a clear knee or exhibit bimodality due to technical artifacts. In such an approach, the user provides an estimate of the expected number of assayed cells. Then, the barcodes are ordered by descending frequency, the frequency <span class="math notranslate nohighlight">\(f\)</span> at a robust quantile index near the expected cell count is obtained, and all cells having a frequency within a small constant fraction of <span class="math notranslate nohighlight">\(f\)</span> (e.g., <span class="math notranslate nohighlight">\(\ge \frac{f}{10}\)</span>) are considered as valid barcodes. Again, the remaining barcodes are corrected against this valid list by attempting to correct uniquely to one of these valid barcodes based on sequence similarity.</p></li>
<li><p>Filtering based on a forced number of valid cells: Perhaps the simplest approach, although potentially problematic, is for the user to directly provide the index in the sorted frequency plot above which barcodes will be considered valid. All barcodes with a frequency greater than or equal to the frequency at the selected index are considered valid and treated as constituting the permit list. The remaining set of barcodes is then corrected against this list using the same approach described in the other methods above. If there are at least as many distinct barcodes as the number of cells the user requests, then this many cells will always be selected. Of course, such an approach is only reasonable when the user has a good reason to believe that the threshold frequency should be set around the provided index.</p></li>
</ul>
</section>
<section id="future-challenges">
<h3><span class="section-number">3.3.2. </span>Future challenges<a class="headerlink" href="#future-challenges" title="Link to this heading">#</a></h3>
<p>While cellular barcoding of high-throughput single-cell profiling has been a tremendously successful approach, some challenges still remain, especially as the scale of experiments continues to grow. For example, the design of a robust method for selecting high-quality cell barcodes from the set of all the observations is still an active area of research, with distinct challenges arising, e.g., between single-cell and single-nucleus experiments. Also, as single-cell technologies have advanced to profile increasing numbers of cells, insufficient sequence diversity in the CB sequence can result in sequence corrections leading to CB collision. Addressing this latter problem may require more intelligent barcode design methods and continuing increases in the lengths of oligonucleotides used for cell barcoding.</p>
</section>
</section>
<section id="umi-resolution">
<span id="raw-proc-umi-resolution"></span><h2><span class="section-number">3.4. </span>UMI resolution<a class="headerlink" href="#umi-resolution" title="Link to this heading">#</a></h2>
<p>After cell barcode (CB) correction, reads have either been discarded or assigned to a corrected CB. Subsequently, we wish to quantify the abundance of each gene within each corrected CB.</p>
<p>Because of the amplification bias as discussed in <a class="reference internal" href="scrna_seq.html#exp-data-transcript-quantification"><span class="std std-ref">Transcript quantification</span></a>, reads must be deduplicated, based upon their UMI, to assess the true count of sampled molecules. Additionally, several other complicating factors present challenges when attempting to perform this estimation.</p>
<p>The UMI deduplication step aims to identify the set of reads and UMIs derived from each original, pre-PCR molecule in each cell captured and sequenced in the experiment. The result of this process is to allocate a molecule count to each gene in each cell, which is subsequently used in the downstream analysis as the raw expression estimate for this gene. We refer to this process of looking at the collection of observed UMIs and their associated mapped reads and attempting to infer the original number of observed molecules arising from each gene as the process of <em>UMI resolution</em>.</p>
<p>To simplify the explanation, in the following text, the reads that map to a reference, for example, a genomic locus of the gene, are called the reads of that reference, and their UMI tags are called the UMIs of that reference; the set of reads that are tagged by a UMI is called the reads of that UMI. A read can only be tagged by one UMI but can belong to multiple references if it maps to multiple references. Furthermore, as the molecule barcoding process for each cell in scRNA-seq is usually isolated and independent (apart from the issues related to accurately resolving cell barcodes raised earlier), without loss of generality, <em>UMI resolution</em> will be explained for a specific cell. The same procedure will usually be applied to all cells independently.</p>
<section id="the-need-for-umi-resolution">
<span id="raw-proc-need-for-umi-resolution"></span><h3><span class="section-number">3.4.1. </span>The need for UMI resolution<a class="headerlink" href="#the-need-for-umi-resolution" title="Link to this heading">#</a></h3>
<p>In the ideal case, where the correct (unaltered) UMIs tag reads, the reads of each UMI uniquely map to a common reference gene, and there is a bijection between UMIs and pre-PCR molecules. Consequently, the UMI deduplication procedure is conceptually straightforward: the reads of a UMI are the PCR duplicates from a single pre-PCR molecule. The number of captured and sequenced molecules of each gene is the number of distinct UMIs observed for this gene.</p>
<p>However, the problems encountered in practice make the simple rules described above insufficient for identifying the gene origin of UMIs in general and necessitate the development of more sophisticated models. Here, we concern ourselves primarily with two challenges.</p>
<ul class="simple">
<li><p>The first problem we discuss is errors in UMIs. These occur when the sequenced UMI tag of reads contains errors introduced during PCR or the sequencing process. Common UMI errors include nucleotide substitutions during PCR and read errors during sequencing. Failing to address such UMI errors can inflate the estimated number of molecules <span id="id41">[<a class="reference internal" href="#id110" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>, <a class="reference internal" href="#id143" title="Christoph Ziegenhain, Gert-Jan Hendriks, Michael Hagemann-Jensen, and Rickard Sandberg. Molecular spikes: a gold standard for single-cell term`RNA` counting. Nature Methods, 19(5):560–566, 2022.">Ziegenhain <em>et al.</em>, 2022</a>]</span>.</p></li>
<li><p>The second issue we discuss is multimapping, which arises in cases where a read or UMI belongs to multiple references, for example, multi-gene reads/UMIs. This happens if different reads of a UMI map to different genes, a read maps to multiple genes, or both. The consequence of this issue is that the gene origin of the multi-gene reads/UMIs is ambiguous, which results in uncertainty about the sampled pre-PCR molecule count of those genes. Simply discarding multi-gene reads/UMIs can lead to a loss of data or a biased estimate among genes that tend to produce multimapping reads, such as sequence-similar gene families<span id="id42">[<a class="reference internal" href="#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>.</p></li>
</ul>
<div class="admonition-a-note-on-umi-errors admonition">
<p class="admonition-title">A note on UMI errors</p>
<p>UMI errors, especially those due to nucleotide substitutions and miscallings, are prevalent in single-cell experiments. <span id="id43">Smith <em>et al.</em> [<a class="reference internal" href="#id110" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">2017</a>]</span> establish that the average number of bases different (edit distance) between the observed UMI sequences in the tested single-cell experiments is lower than randomly sampled UMI sequences, and the enrichment of low edit distances is well correlated with the degree of PCR amplification. Multimapping also exists in single-cell data and, depending upon the gene being considered, can occur at a non-trivial rate. <span id="id44">Srivastava <em>et al.</em> [<a class="reference internal" href="#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">2019</a>]</span> show that discarding the multimapping reads can negatively bias the predicted molecule counts.</p>
</div>
<p>There exist other challenges that we do not focus upon here, such as “convergent” and “divergent” UMI collisions. We consider the case where the same UMI is used to tag two different pre-PCR molecules arising from the same gene, in the same cell, as a convergent collision. When two or more distinct UMIs arise from the same pre-PCR molecule, e.g., due to the sampling of multiple priming sites from this molecule, we consider this a divergent collision. We expect convergent UMI collisions to be rare and, therefore, their effect typically small. Further, transcript-level mapping information can sometimes be used to resolve such collisions<span id="id45">[<a class="reference internal" href="#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>. Divergent UMI collisions occur primarily among introns of unspliced transcripts<span id="id46">[<a class="reference internal" href="#id112" title="10x Genomics. Technical note - interpreting intronic and antisense reads in 10x genomics single cell gene expression data. https://www.10xgenomics.com/support/single-cell-gene-expression/documentation/steps/sequencing/interpreting-intronic-and-antisense-reads-in-10-x-genomics-single-cell-gene-expression-data, August 2021.">10x Genomics, 2021</a>]</span>, and approaches to addressing the issues they raise are an area of active research<span id="id47">[<a class="reference internal" href="#id144" title="Gennady Gorin and Lior Pachter. Length Biases in Single-Cell term`RNA` Sequencing of pre-mRNA. bioRxiv, 2021. URL: https://www.biorxiv.org/content/early/2021/07/31/2021.07.30.454514, arXiv:https://www.biorxiv.org/content/early/2021/07/31/2021.07.30.454514.full.pdf, doi:10.1101/2021.07.30.454514.">Gorin and Pachter, 2021</a>, <a class="reference internal" href="#id112" title="10x Genomics. Technical note - interpreting intronic and antisense reads in 10x genomics single cell gene expression data. https://www.10xgenomics.com/support/single-cell-gene-expression/documentation/steps/sequencing/interpreting-intronic-and-antisense-reads-in-10-x-genomics-single-cell-gene-expression-data, August 2021.">10x Genomics, 2021</a>]</span>.</p>
<p>Given that the use of UMIs is near ubiquitous in high-throughput scRNA-seq protocols and the fact that addressing these errors improves the estimation of gene abundances, there has been much attention paid to the problem of UMI resolution in recent literature <span id="id48">[<a class="reference internal" href="#id114" title="Sayantan Bose, Zhenmao Wan, Ambrose Carr, Abbas H. Rizvi, Gregory Vieira, Dana Pe'er, and Peter A. Sims. Scalable microfluidics for single-cell term`RNA` printing and sequencing. Genome Biology, June 2015. URL: https://doi.org/10.1186/s13059-015-0684-3, doi:10.1186/s13059-015-0684-3.">Bose <em>et al.</em>, 2015</a>, <a class="reference internal" href="#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id113" title="Saiful Islam, Amit Zeisel, Simon Joost, Gioele La Manno, Pawel Zajac, Maria Kasper, Peter Lönnerberg, and Sten Linnarsson. Quantitative single-cell term`RNA`-seq with unique molecular identifiers. Nature Methods, 11(2):163–166, December 2013. URL: https://doi.org/10.1038/nmeth.2772, doi:10.1038/nmeth.2772.">Islam <em>et al.</em>, 2013</a>, <a class="reference internal" href="#id115" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id150" title="Evan Z. Macosko, Anindita Basu, Rahul Satija, James Nemesh, Karthik Shekhar, Melissa Goldman, Itay Tirosh, Allison R. Bialas, Nolan Kamitaki, Emily M. Martersteck, John J. Trombetta, David A. Weitz, Joshua R. Sanes, Alex K. Shalek, Aviv Regev, and Steven A. McCarroll. Highly parallel genome-wide expression profiling of individual cells using nanoliter droplets. Cell, 161(5):1202-1214, May 2015. URL: https://doi.org/10.1016/j.cell.2015.05.002, doi:10.1016/j.cell.2015.05.002.">Macosko <em>et al.</em>, 2015</a>, <a class="reference internal" href="#id116" title="Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. Nature Biotechnology, 39(7):813–818, April 2021. URL: https://doi.org/10.1038/s41587-021-00870-2, doi:10.1038/s41587-021-00870-2.">Melsted <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id119" title="Baraa Orabi, Emre Erhan, Brian McConeghy, Stanislav V Volik, Stephane Le Bihan, Robert Bell, Colin C Collins, Cedric Chauve, and Faraz Hach. Alignment-free clustering of UMI tagged term`DNA` molecules. Bioinformatics, 35(11):1829–1836, October 2018. URL: https://doi.org/10.1093/bioinformatics/bty888, doi:10.1093/bioinformatics/bty888.">Orabi <em>et al.</em>, 2018</a>, <a class="reference internal" href="#id121" title="Swati Parekh, Christoph Ziegenhain, Beate Vieth, Wolfgang Enard, and Ines Hellmann. zUMIs - a fast and flexible pipeline to process term`RNA` sequencing data with UMIs. GigaScience, May 2018. URL: https://doi.org/10.1093/gigascience/giy059, doi:10.1093/gigascience/giy059.">Parekh <em>et al.</em>, 2018</a>, <a class="reference internal" href="#id110" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>, <a class="reference internal" href="#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id120" title="Maria Tsagiopoulou, Maria Christina Maniou, Nikolaos Pechlivanis, Anastasis Togkousidis, Michaela Kotrová, Tobias Hutzenlaub, Ilias Kappas, Anastasia Chatzidimitriou, and Fotis Psomopoulos. UMIc: a preprocessing method for UMI deduplication and reads correction. Frontiers in Genetics, May 2021. URL: https://doi.org/10.3389/fgene.2021.660366, doi:10.3389/fgene.2021.660366.">Tsagiopoulou <em>et al.</em>, 2021</a>]</span>.</p>
</section>
<section id="graph-based-umi-resolution">
<span id="raw-proc-graph-based-umi-resolution"></span><h3><span class="section-number">3.4.2. </span>Graph-based UMI resolution<a class="headerlink" href="#graph-based-umi-resolution" title="Link to this heading">#</a></h3>
<p>As a result of the problems that arise when attempting to resolve UMIs, many methods have been developed to address the problem of UMI resolution. While there are a host of different approaches for UMI resolution, we will focus on a framework for representing problem instances, modified from a framework initially proposed by <span id="id49">Smith <em>et al.</em> [<a class="reference internal" href="#id110" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">2017</a>]</span>, that relies upon the notion of a <em>UMI graph</em>. Each connected component of this graph represents a sub-problem wherein certain subsets of UMIs are collapsed (i.e., resolved as evidence of the same pre-PCR molecule). Many popular UMI resolution approaches can be interpreted in this framework by simply modifying precisely how the graph is refined and how the collapse or resolution procedure carried out over this graph works.</p>
<p>In the context of single-cell data, a UMI graph <span class="math notranslate nohighlight">\(G(V,E)\)</span> is a directed graph with a node set <span class="math notranslate nohighlight">\(V\)</span> and an edge set <span class="math notranslate nohighlight">\(E\)</span>. Each node <span class="math notranslate nohighlight">\(v_i \in V\)</span> represents an equivalence class (EC) of reads, and the edge set <span class="math notranslate nohighlight">\(E\)</span> encodes the relationship between the ECs. The equivalence relation <span class="math notranslate nohighlight">\(\sim_r\)</span> defined on reads is based on their UMI and mapping information. We say reads <span class="math notranslate nohighlight">\(r_x\)</span> and <span class="math notranslate nohighlight">\(r_y\)</span> are equivalent, <span class="math notranslate nohighlight">\(r_x \sim_r r_y\)</span>, if and only if they have identical UMI tags and map to the same set of references. UMI resolution approaches may define a “reference” as a genomic locus<span id="id50">[<a class="reference internal" href="#id110" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>]</span>, transcript<span id="id51">[<a class="reference internal" href="#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span> or gene<span id="id52">[<a class="reference internal" href="#id115" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id151" title="Grace X. Y. Zheng, Jessica M. Terry, Phillip Belgrader, Paul Ryvkin, Zachary W. Bent, Ryan Wilson, Solongo B. Ziraldo, Tobias D. Wheeler, Geoff P. McDermott, Junjie Zhu, Mark T. Gregory, Joe Shuga, Luz Montesclaros, Jason G. Underwood, Donald A. Masquelier, Stefanie Y. Nishimura, Michael Schnall-Levin, Paul W. Wyatt, Christopher M. Hindson, Rajiv Bharadwaj, Alexander Wong, Kevin D. Ness, Lan W. Beppu, H. Joachim Deeg, Christopher McFarland, Keith R. Loeb, William J. Valente, Nolan G. Ericson, Emily A. Stevens, Jerald P. Radich, Tarjei S. Mikkelsen, Benjamin J. Hindson, and Jason H. Bielas. Massively parallel digital transcriptional profiling of single cells. Nature Communications, 8(1):14049, Jan 2017. URL: https://doi.org/10.1038/ncomms14049, doi:10.1038/ncomms14049.">Zheng <em>et al.</em>, 2017</a>]</span>. Other UMI resolution approaches exist, for example, the reference-free model<span id="id53">[<a class="reference internal" href="#id120" title="Maria Tsagiopoulou, Maria Christina Maniou, Nikolaos Pechlivanis, Anastasis Togkousidis, Michaela Kotrová, Tobias Hutzenlaub, Ilias Kappas, Anastasia Chatzidimitriou, and Fotis Psomopoulos. UMIc: a preprocessing method for UMI deduplication and reads correction. Frontiers in Genetics, May 2021. URL: https://doi.org/10.3389/fgene.2021.660366, doi:10.3389/fgene.2021.660366.">Tsagiopoulou <em>et al.</em>, 2021</a>]</span> and the method of moments<span id="id54">[<a class="reference internal" href="#id116" title="Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. Nature Biotechnology, 39(7):813–818, April 2021. URL: https://doi.org/10.1038/s41587-021-00870-2, doi:10.1038/s41587-021-00870-2.">Melsted <em>et al.</em>, 2021</a>]</span>, but they may not be easily represented in this framework and are not discussed in further detail here.</p>
<p>In the UMI graph framework, a UMI resolution approach can be divided into three major steps, each of which has different options that can be modularly composed by different approaches. The three steps are defining nodes, defining adjacency relationships, and resolving components. Additionally, these steps may sometimes be preceded (and/or followed) by filtering steps designed to discard or heuristically assign (by modifying the set of reference mappings reported) reads and UMIs exhibiting certain types of mapping ambiguity.</p>
<section id="defining-nodes">
<span id="raw-proc-umi-graph-node-def"></span><h4><span class="section-number">3.4.2.1. </span>Defining nodes<a class="headerlink" href="#defining-nodes" title="Link to this heading">#</a></h4>
<p>As described above, a node <span class="math notranslate nohighlight">\(v_i \in V\)</span> is an equivalence class of reads. Therefore, <span class="math notranslate nohighlight">\(V\)</span> can be defined based on the full or filtered set of mapped reads and their associated <em>uncorrected</em> UMIs. All reads that satisfy the equivalence relation <span class="math notranslate nohighlight">\(\sim_r\)</span> based on their reference set and UMI tag are associated with the same vertex <span class="math notranslate nohighlight">\(v \in V\)</span>. An EC is a multi-gene EC if its UMI is a multi-gene UMI. Some approaches will avoid the creation of such ECs by filtering or heuristically assigning reads prior to node creation, while other approaches will retain and process these ambiguous vertices and attempt and resolve their gene origin via parsimony, probabilistic assignment, or based on a related rule or model <span id="id55">[<a class="reference internal" href="#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id115" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>.</p>
</section>
<section id="defining-the-adjacency-relationship">
<span id="raw-proc-umi-graph-edge-def"></span><h4><span class="section-number">3.4.2.2. </span>Defining the adjacency relationship<a class="headerlink" href="#defining-the-adjacency-relationship" title="Link to this heading">#</a></h4>
<p>After creating the node set <span class="math notranslate nohighlight">\(V\)</span> of a UMI graph, the adjacency of nodes in <span class="math notranslate nohighlight">\(V\)</span> is defined based on the distance, typically the Hamming or edit distance, between their UMI sequences and, optionally, the content of their associated reference sets.</p>
<p>Here we define the following functions on the node <span class="math notranslate nohighlight">\(v_i \in V\)</span>:</p>
<ul class="simple">
<li><p><span class="math notranslate nohighlight">\(u(v_i)\)</span> is the UMI tag of <span class="math notranslate nohighlight">\(v_i\)</span>.</p></li>
<li><p><span class="math notranslate nohighlight">\(c(v_i) = |v_i|\)</span> is the cardinality of <span class="math notranslate nohighlight">\(v_i\)</span>, i.e., the number of reads associated with <span class="math notranslate nohighlight">\(v_i\)</span> that are equivalent under <span class="math notranslate nohighlight">\(\sim_r\)</span>.</p></li>
<li><p><span class="math notranslate nohighlight">\(m(v_i)\)</span> is the reference set encoded in the mapping information, for <span class="math notranslate nohighlight">\(v_i\)</span>.</p></li>
<li><p><span class="math notranslate nohighlight">\(D(v_i, v_j)\)</span> is the distance between <span class="math notranslate nohighlight">\(u(v_i)\)</span> and <span class="math notranslate nohighlight">\(u(v_j)\)</span>, where <span class="math notranslate nohighlight">\(v_j \in V\)</span>.</p></li>
</ul>
<p>Given these function definitions, any two nodes <span class="math notranslate nohighlight">\(v_i, v_j \in V\)</span> will be incident with a bi-directed edge if and only if <span class="math notranslate nohighlight">\(m(v_i) \cap m(v_j) \ne \emptyset\)</span> and <span class="math notranslate nohighlight">\(D(v_i,v_j) \le \theta\)</span>, where <span class="math notranslate nohighlight">\(\theta\)</span> is a distance threshold and is often set as <span class="math notranslate nohighlight">\(\theta=1\)</span><span id="id56">[<a class="reference internal" href="#id115" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id110" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>, <a class="reference internal" href="#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>. Additionally, the bi-directed edge might be replaced by a directed edge incident from <span class="math notranslate nohighlight">\(v_i\)</span> to <span class="math notranslate nohighlight">\(v_j\)</span> if <span class="math notranslate nohighlight">\(c(v_i) \ge 2c(v_j) -1\)</span> or vice versa<span id="id57">[<a class="reference internal" href="#id110" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>, <a class="reference internal" href="#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>. Though these edge definitions are among the most common, others are possible, so long as they are completely defined by the <span class="math notranslate nohighlight">\(u\)</span>, <span class="math notranslate nohighlight">\(c\)</span>, <span class="math notranslate nohighlight">\(m\)</span>, and <span class="math notranslate nohighlight">\(D\)</span> functions. With <span class="math notranslate nohighlight">\(V\)</span> and <span class="math notranslate nohighlight">\(E\)</span> in hand, the UMI graph <span class="math notranslate nohighlight">\(G = (V,E)\)</span> is now defined.</p>
</section>
<section id="defining-the-graph-resolution-approach">
<span id="raw-proc-umi-graph-resolution-def"></span><h4><span class="section-number">3.4.2.3. </span>Defining the graph resolution approach<a class="headerlink" href="#defining-the-graph-resolution-approach" title="Link to this heading">#</a></h4>
<p>Given the defined UMI graph, many different resolution approaches may be applied. A resolution method may be as simple as finding the set of connected components, clustering the graph, greedily collapsing nodes or contracting edges<span id="id58">[<a class="reference internal" href="#id110" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>]</span>, or searching for a cover of the graph by structures following certain rules (e.g., monochromatic arboresences<span id="id59">[<a class="reference internal" href="#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>) to reduce the graph. As a result, each node in the reduced UMI graph, or each element in the cover in the case that the graph is not modified dynamically, represents a pre-PCR molecule. The collapsed nodes or covering sets are regarded as the PCR duplicates of that molecule.</p>
<p>Different rules for defining the adjacency relationship and different approaches for graph resolution itself can seek to preserve different properties and can define a wide variety of distinct overall UMI resolution approaches. Note that for the approaches that aim to probabilistically resolve the ambiguity caused by multimapping, the resolved UMI graph might contain multi-gene ECs, and their gene origin will be resolved in the following step.</p>
</section>
<section id="quantification">
<span id="raw-proc-umi-graph-quantification"></span><h4><span class="section-number">3.4.2.4. </span>Quantification<a class="headerlink" href="#quantification" title="Link to this heading">#</a></h4>
<p>The last step in UMI resolution is quantifying the abundance of each gene using the resolved UMI graph. For approaches that discard multi-gene ECs, the molecule count vector for the genes in the current cell being processed (or count vector for short) is generated by counting the number of ECs labeled with each gene. On the other hand, approaches that process, rather than discard, multi-gene ECs usually resolve the ambiguity by applying some statistical inference procedure. For example, <span id="id60">Srivastava <em>et al.</em> [<a class="reference internal" href="#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">2019</a>]</span> introduce an expectation-maximization (EM) approach for probabilistically assigning multi-gene UMIs, and related EM algorithms have also been introduced as optional steps in subsequent tools<span id="id61">[<a class="reference internal" href="#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id115" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id116" title="Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. Nature Biotechnology, 39(7):813–818, April 2021. URL: https://doi.org/10.1038/s41587-021-00870-2, doi:10.1038/s41587-021-00870-2.">Melsted <em>et al.</em>, 2021</a>]</span>. In this model, the collapsed-EC-to-gene assignments are latent variables, and the deduplicated molecule count of genes are the main parameters. Intuitively, evidence from gene-unique ECs will be used to help probabilistically apportion the multi-gene ECs. The EM algorithm seeks the parameters that together have the (locally) highest likelihood of generating the observed ECs.</p>
<p>Usually, the UMI resolution and quantification process described above will be performed separately for each cell, represented by a corrected CB, to create a complete count matrix for all genes in all cells. However, the relative paucity of per-cell information in high-throughput single-cell samples limits the evidence available when performing UMI resolution, which in turn limits the potential efficacy of model-based solutions like the statistical inference procedure described above.
Thus, further research here is certainly warranted. For example, <span id="id62">Srivastava <em>et al.</em> [<a class="reference internal" href="#id122" title="Avi Srivastava, Laraib Malik, Hirak Sarkar, and Rob Patro. A Bayesian framework for inter-cellular information sharing improves dscRNA-seq quantification. Bioinformatics, 36(Supplement_1):i292–i299, July 2020. URL: https://doi.org/10.1093/bioinformatics/btaa450, doi:10.1093/bioinformatics/btaa450.">2020</a>]</span> introduced an approach that allows sharing information among transcriptionally similar cells to improve the quantification result further.</p>
</section>
</section>
</section>
<section id="count-matrix-quality-control">
<span id="raw-proc-count-qc"></span><h2><span class="section-number">3.5. </span>Count matrix quality control<a class="headerlink" href="#count-matrix-quality-control" title="Link to this heading">#</a></h2>
<p>Once a count matrix has been generated, it is important to perform a quality control (QC) assessment.
There are several distinct assessments that generally fall under the rubric of quality control.
Basic global metrics are often recorded and reported to help assess the overall quality of the sequencing measurement itself. These metrics consist of quantities such as the total fraction of mapped reads, the distribution of distinct UMIs observed per cell, the distribution of UMI deduplication rates, the distribution of detected genes per cell, etc. These and similar metrics are often recorded by the quantification tools themselves<span id="id63">[<a class="reference internal" href="#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id115" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id116" title="Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. Nature Biotechnology, 39(7):813–818, April 2021. URL: https://doi.org/10.1038/s41587-021-00870-2, doi:10.1038/s41587-021-00870-2.">Melsted <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id151" title="Grace X. Y. Zheng, Jessica M. Terry, Phillip Belgrader, Paul Ryvkin, Zachary W. Bent, Ryan Wilson, Solongo B. Ziraldo, Tobias D. Wheeler, Geoff P. McDermott, Junjie Zhu, Mark T. Gregory, Joe Shuga, Luz Montesclaros, Jason G. Underwood, Donald A. Masquelier, Stefanie Y. Nishimura, Michael Schnall-Levin, Paul W. Wyatt, Christopher M. Hindson, Rajiv Bharadwaj, Alexander Wong, Kevin D. Ness, Lan W. Beppu, H. Joachim Deeg, Christopher McFarland, Keith R. Loeb, William J. Valente, Nolan G. Ericson, Emily A. Stevens, Jerald P. Radich, Tarjei S. Mikkelsen, Benjamin J. Hindson, and Jason H. Bielas. Massively parallel digital transcriptional profiling of single cells. Nature Communications, 8(1):14049, Jan 2017. URL: https://doi.org/10.1038/ncomms14049, doi:10.1038/ncomms14049.">Zheng <em>et al.</em>, 2017</a>]</span> since they arise naturally and can be computed during the process of read mapping, cell barcode correction, and UMI resolution. Likewise, there exist several tools to help organize and visualize these basic metrics, such as the <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/visualization/latest/what-is-loupe-cell-browser">Loupe browser</a>, <a class="reference external" href="https://github.com/csoneson/alevinQC">alevinQC</a>, or a <a class="reference external" href="https://github.com/pachterlab/kb_python">kb_python report</a>, depending upon the quantification pipeline being used. Beyond these basic global metrics, at this stage of analysis, QC metrics are designed primarily to help determine which cells (CBs) have been sequenced “successfully”, and which exhibit artifacts that warrant filtering or correction.</p>
<p>In the following toggle section, we discuss an example alevinQC report taken from the <code class="docutils literal notranslate"><span class="pre">alevinQC</span></code> <a class="reference external" href="https://github.com/csoneson/alevinQC">manual webpage</a>.</p>
<div class="toggle docutils container">
<p>Once <code class="docutils literal notranslate"><span class="pre">alevin</span></code> or <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> quantifies the single-cell data, the quality of the data can be assessed through the R package <a class="reference external" href="https://github.com/csoneson/alevinQC"><code class="docutils literal notranslate"><span class="pre">alevinQC</span></code></a>. The alevinQC report can be generated in PDF format or as R/Shiny applications, which summarizes various components of the single-cell library, such as reads, CBs, and UMIs.</p>
<p><strong>1. Metadata and summary tables</strong></p>
<figure class="align-default" id="raw-proc-fig-alevinqc-summary">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/alevinQC_summary.png"><img alt="AlevinQC Summary" class="bg-primary mb-1" src="../_images/alevinQC_summary.png" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3.15 </span><span class="caption-text">An example of the summary section of an alevinQC report.</span><a class="headerlink" href="#raw-proc-fig-alevinqc-summary" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>The first section of an alevinQC report shows a summary of the input files and the processing result, among which, the top left table displays the metadata provided by <code class="docutils literal notranslate"><span class="pre">alevin</span></code> (or <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code>) for the quantification results. For example, this includes the time of the run, the version of the tool, and the path to the input FASTQ and index files. The top right summary table provides the summary statistics for various components of the single-cell library, for example, the number of sequencing reads, the number of selected cell barcodes at various levels of filtering, and the total number of deduplicated UMIs.</p>
<p><strong>2. Knee plot, initial whitelist determination</strong></p>
<figure class="align-default" id="raw-proc-fig-alevinqc-plots">
<a class="bg-primary mb-1 reference internal image-reference" href="../_images/alevinQC_plots.png"><img alt="AlevinQC Plots" class="bg-primary mb-1" src="../_images/alevinQC_plots.png" style="width: 800px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 3.16 </span><span class="caption-text">The figure shows the plots in the alevinQC report of an example single-cell dataset, of which the cells are filtered using the “knee” finding method. Each dot represents a corrected cell barcode with its corrected profile.</span><a class="headerlink" href="#raw-proc-fig-alevinqc-plots" title="Link to this image">#</a></p>
</figcaption>
</figure>
<p>The first（top left) view in <a class="reference internal" href="#raw-proc-fig-alevinqc-plots"><span class="std std-numref">Fig. 3.16</span></a> shows the distribution of cell barcode frequency in decreasing order. In all plots shown above, each point represents a corrected cell barcode, with its x-coordinate corresponding to its cell barcode frequency rank. In the top left plot, the y-coordinate corresponds to the observed frequency of the corrected barcode. Generally, this plot shows a “knee”-like pattern, which can be used to identify the initial list of high-quality barcodes. The red dots in the plot represent the cell barcodes selected as the high-quality cell barcodes in the case that “knee”-based filtering was applied. In other words, these cell barcodes contain a sufficient number of reads to be deemed high-quality and likely derived from truly present cells. Suppose an external permit list is passed in the CB correction step, which implies no internal algorithm was used to distinguish high-quality cell barcodes. In that case, all dots in the plot will be colored red, as all these corrected cell barcodes are processed throughout the raw data processing pipeline and reported in the gene count matrix. One should be skeptical of the data quality if the frequency is consistently low across all cell barcodes.</p>
<p><strong>3. Barcode collapsing</strong></p>
<p>After identification of the barcodes that will be processed, either through an internal threshold (e.g., from the “knee”-based method) or through external whitelisting, <code class="docutils literal notranslate"><span class="pre">alevin</span></code> (or <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code>) performs cell barcode sequence correction. The barcode collapsing plot, the upper middle plot in the <a class="reference internal" href="#raw-proc-fig-alevinqc-plots"><span class="std std-numref">Fig. 3.16</span></a>, shows the number of reads assigned to a cell barcode after sequence correction of the cell barcodes versus prior to correction. Generally, we would see that all points fall close to the line representing <span class="math notranslate nohighlight">\(x = y\)</span>, which means that the reassignments in CB correction usually do not drastically change the profile of the cell barcodes.</p>
<p><strong>4. Knee Plot, number of genes per cell</strong></p>
<p>The upper right plot in <a class="reference internal" href="#raw-proc-fig-alevinqc-plots"><span class="std std-numref">Fig. 3.16</span></a> shows the distribution of the number of observed genes of all processed cell barcodes. Generally, a mean of <span class="math notranslate nohighlight">\(2,000\)</span> genes per cell is considered modest but reasonable for the downstream analyses. One should double-check the quality of the data if all cells have a low number of observed genes.</p>
<p><strong>5. Quantification summary</strong></p>
<p>Finally, a series of quantification summary plots, the bottom plots in <a class="reference internal" href="#raw-proc-fig-alevinqc-plots"><span class="std std-numref">Fig. 3.16</span></a>, compare the cell barcode frequency, the total number of UMIs after deduplication and the total number of non-zero genes using scatter plots. In general, in each plot, the plotted data should demonstrate a positive correlation, and, if high-quality filtering (e.g., knee filtering) has been performed, the high-quality cell barcodes should be well separated from the rest. Moreover, one should expect all three plots to convey similar trends. If using an external permit list, all the dots in the plots will be colored red, as all these cell barcodes are processed and reported in the gene count matrix. Still, we should see the correlation between the plots and the separation of the dots representing high-quality cells to others. If all of these metrics are consistently low across cells or if these plots convey substantially different trends, then one should be concerned about the data quality.</p>
</div>
<section id="empty-droplet-detection">
<h3><span class="section-number">3.5.1. </span>Empty droplet detection<a class="headerlink" href="#empty-droplet-detection" title="Link to this heading">#</a></h3>
<p>One of the first QC steps is determining which cell barcodes correspond to “high-confidence” sequenced cells. It is common in droplet-based protocols<span id="id64">[<a class="reference internal" href="#id150" title="Evan Z. Macosko, Anindita Basu, Rahul Satija, James Nemesh, Karthik Shekhar, Melissa Goldman, Itay Tirosh, Allison R. Bialas, Nolan Kamitaki, Emily M. Martersteck, John J. Trombetta, David A. Weitz, Joshua R. Sanes, Alex K. Shalek, Aviv Regev, and Steven A. McCarroll. Highly parallel genome-wide expression profiling of individual cells using nanoliter droplets. Cell, 161(5):1202-1214, May 2015. URL: https://doi.org/10.1016/j.cell.2015.05.002, doi:10.1016/j.cell.2015.05.002.">Macosko <em>et al.</em>, 2015</a>]</span> that certain barcodes are associated with ambient <a class="reference internal" href="../glossary.html#term-RNA"><span class="xref std std-term">RNA</span></a> instead of the <a class="reference internal" href="../glossary.html#term-RNA"><span class="xref std std-term">RNA</span></a> of a captured cell. This happens when droplets fail to capture a cell. These empty droplets still tend to produce sequenced reads, though the characteristics of these reads look markedly different from those associated with barcodes corresponding to properly captured cells. Many approaches exist to assess whether a barcode likely corresponds to an empty droplet or not. One simple method is to examine the cumulative frequency plot of the barcodes, in which barcodes are sorted in descending order of the number of distinct UMIs with which they are associated. This plot often contains a “knee” that can be identified as a likely point of discrimination between properly captured cells and empty droplets<span id="id65">[<a class="reference internal" href="#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id110" title="Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome research, 27(3):491–499, 2017.">Smith <em>et al.</em>, 2017</a>]</span>.
While this “knee” method is intuitive and can often estimate a reasonable threshold, it has several drawbacks. For example, not all cumulative histograms display an obvious knee, and it is notoriously difficult to design algorithms that can robustly and automatically detect such knees. Finally, the total UMI count associated with a barcode may not, alone, be the best signal to determine if the barcode was associated with an empty or damaged cell.</p>
<p>This led to the development of several tools specifically designed to detect empty or damaged droplets, or cells generally deemed to be of “low quality” <span id="id66">[<a class="reference internal" href="#id127" title="Marcus Alvarez, Elior Rahmani, Brandon Jew, Kristina M. Garske, Zong Miao, Jihane N. Benhammou, Chun Jimmie Ye, Joseph R. Pisegna, Kirsi H. Pietiläinen, Eran Halperin, and Päivi Pajukanta. Enhancing droplet-based single-nucleus term`RNA`-seq resolution using the semi-supervised machine learning classifier DIEM. Scientific Reports, July 2020. URL: https://doi.org/10.1038/s41598-020-67513-5, doi:10.1038/s41598-020-67513-5.">Alvarez <em>et al.</em>, 2020</a>, <a class="reference internal" href="#id124" title="Cody N. Heiser, Victoria M. Wang, Bob Chen, Jacob J. Hughey, and Ken S. Lau. Automated quality control and cell identification of droplet-based single-cell data using dropkick. Genome Research, 31(10):1742–1752, April 2021. URL: https://doi.org/10.1101/gr.271908.120, doi:10.1101/gr.271908.120.">Heiser <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id125" title="Ariel A Hippen, Matias M Falco, Lukas M Weber, Erdogan Pekcan Erkan, Kaiyang Zhang, Jennifer Anne Doherty, Anna Vähärautio, Casey S Greene, and Stephanie C Hicks. miQC: An adaptive probabilistic framework for quality control of single-cell term`RNA`-sequencing data. PLoS computational biology, 17(8):e1009290, 2021.">Hippen <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id123" title="Aaron TL Lun, Samantha Riesenfeld, Tallulah Andrews, Tomas Gomes, John C Marioni, and others. EmptyDrops: distinguishing cells from empty droplets in droplet-based single-cell term`RNA` sequencing data. Genome biology, 20(1):1–9, 2019.">Lun <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id126" title="Walter Muskovic and Joseph E Powell. DropletQC: improved identification of empty droplets and damaged cells in single-cell term`RNA`-seq data. Genome Biology, 22(1):1–9, 2021.">Muskovic and Powell, 2021</a>, <a class="reference internal" href="#id128" title="Matthew D Young and Sam Behjati. SoupX removes ambient term`RNA` contamination from droplet-based single-cell term`RNA` sequencing data. GigaScience, December 2020. URL: https://doi.org/10.1093/gigascience/giaa151, doi:10.1093/gigascience/giaa151.">Young and Behjati, 2020</a>]</span>. These tools incorporate a variety of different measures of cell quality, including the frequencies of distinct UMIs, the number of detected genes, and the fraction of mitochondrial <a class="reference internal" href="../glossary.html#term-RNA"><span class="xref std std-term">RNA</span></a>, and typically work by applying a statistical model to these features to classify high-quality cells from putative empty droplets or damaged cells. This means that cells can typically be scored, and a final filtering can be selected based on an estimated posterior probability that cells are not empty or compromised. While these models generally work well for single-cell <a class="reference internal" href="../glossary.html#term-RNA"><span class="xref std std-term">RNA</span></a>-seq data, one may have to apply several additional filters or heuristics to obtain robust filtering in single-nucleus <a class="reference internal" href="../glossary.html#term-RNA"><span class="xref std std-term">RNA</span></a>-seq data<span id="id67">[<a class="reference internal" href="#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id115" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>]</span>, like those exposed in the <a class="reference external" href="https://github.com/MarioniLab/DropletUtils/blob/master/R/emptyDropsCellRanger.R"><code class="docutils literal notranslate"><span class="pre">emptyDropsCellRanger</span></code></a> function of <code class="docutils literal notranslate"><span class="pre">DropletUtils</span></code><span id="id68">[<a class="reference internal" href="#id123" title="Aaron TL Lun, Samantha Riesenfeld, Tallulah Andrews, Tomas Gomes, John C Marioni, and others. EmptyDrops: distinguishing cells from empty droplets in droplet-based single-cell term`RNA` sequencing data. Genome biology, 20(1):1–9, 2019.">Lun <em>et al.</em>, 2019</a>]</span>.</p>
</section>
<section id="doublet-detection">
<h3><span class="section-number">3.5.2. </span>Doublet detection<a class="headerlink" href="#doublet-detection" title="Link to this heading">#</a></h3>
<p>In addition to determining which cell barcodes correspond to empty droplets or damaged cells, one may also wish to identify those cell barcodes that correspond to doublets or multiplets. When a given droplet captures two (doublets) or more (multiplets) cells, this can result in a skewed distribution for these cell barcodes in terms of quantities like the number of reads and UMIs they represent, as well as gene expression profiles they display. Many tools have also been developed to predict the doublet status of cell barcodes<span id="id69">[<a class="reference internal" href="#id132" title="Abha S Bais and Dennis Kostka. Scds: computational annotation of doublets in single-cell term`RNA` sequencing data. Bioinformatics, 36(4):1150–1158, September 2019. URL: https://doi.org/10.1093/bioinformatics/btz698, doi:10.1093/bioinformatics/btz698.">Bais and Kostka, 2019</a>, <a class="reference internal" href="#id145" title="Nicholas J. Bernstein, Nicole L. Fong, Irene Lam, Margaret A. Roy, David G. Hendrickson, and David R. Kelley. Solo: Doublet Identification in Single-Cell term`RNA`-Seq via Semi-Supervised Deep Learning. Cell Systems, 11(1):95–101.e5, July 2020. URL: https://doi.org/10.1016/j.cels.2020.05.010, doi:10.1016/j.cels.2020.05.010.">Bernstein <em>et al.</em>, 2020</a>, <a class="reference internal" href="#id129" title="Erica A.K. DePasquale, Daniel J. Schnell, Pieter-Jan Van Camp, Íñigo Valiente-Aland\'ı, Burns C. Blaxall, H. Leighton Grimes, Harinder Singh, and Nathan Salomonis. DoubletDecon: deconvoluting doublets from single-cell term`RNA`-sequencing data. Cell Reports, 29(6):1718–1727.e8, November 2019. URL: https://doi.org/10.1016/j.celrep.2019.09.082, doi:10.1016/j.celrep.2019.09.082.">DePasquale <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id130" title="Christopher S. McGinnis, Lyndsay M. Murrow, and Zev J. Gartner. DoubletFinder: doublet detection in single-cell term`RNA` sequencing data using artificial nearest neighbors. Cell Systems, 8(4):329–337.e4, April 2019. URL: https://doi.org/10.1016/j.cels.2019.03.003, doi:10.1016/j.cels.2019.03.003.">McGinnis <em>et al.</em>, 2019</a>, <a class="reference internal" href="#id131" title="Samuel L. Wolock, Romain Lopez, and Allon M. Klein. Scrublet: Computational Identification of Cell Doublets in Single-Cell Transcriptomic Data. Cell Systems, 8(4):281–291.e9, April 2019. URL: https://doi.org/10.1016/j.cels.2018.11.005, doi:10.1016/j.cels.2018.11.005.">Wolock <em>et al.</em>, 2019</a>]</span>. Once detected, cells determined to likely be doublets and multiplets can be removed or otherwise adjusted for in the subsequent analysis.</p>
</section>
</section>
<section id="count-data-representation">
<span id="raw-proc-output-representation"></span><h2><span class="section-number">3.6. </span>Count data representation<a class="headerlink" href="#count-data-representation" title="Link to this heading">#</a></h2>
<p>As one completes initial raw data processing and quality control and moves on to subsequent analyses, it is important to acknowledge and remember that the cell-by-gene count matrix is, at best, an approximation of the sequenced molecules in the original sample. At several stages of the raw data processing pipeline, heuristics are applied, and simplifications are made to enable the generation of this count matrix. For example, read mapping is imperfect, as is cell barcode correction. The accurate resolution of UMIs is particularly challenging, and issues related to UMIs attached to multimapping reads are often ignored, as is the fact that multiple priming sites, particularly among unspliced molecules, can violate the one molecule-to-one UMI relationship that is often assumed.</p>
<p>In light of these challenges and the simplifications adopted to address them, there remains active research as to how best to represent the preprocessed data to downstream tools. For example, several quantification tools<span id="id70">[<a class="reference internal" href="#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id115" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id116" title="Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. Nature Biotechnology, 39(7):813–818, April 2021. URL: https://doi.org/10.1038/s41587-021-00870-2, doi:10.1038/s41587-021-00870-2.">Melsted <em>et al.</em>, 2021</a>, <a class="reference internal" href="#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span> implement an <em>optional</em> EM algorithm, initially introduced in this context by <span id="id71">Srivastava <em>et al.</em> [<a class="reference internal" href="#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">2019</a>]</span>, that probabilistically apportions UMIs associated with reads that map to more than one gene. This, however, can result in non-integer count matrices that may be unexpected by certain downstream tools. Alternatively, one may choose to resolve UMIs to <em>gene groups</em> instead of individual genes, retaining the multimapping information in the preprocessed output (it is worth noting that a similar representation, but at the transcript level, has been used for over a decade as a succinct internal representation in bulk RNA-seq transcript quantification tools<span id="id72">[<a class="reference internal" href="#id135" title="Nicolas L Bray, Harold Pimentel, Páll Melsted, and Lior Pachter. Near-optimal probabilistic term`RNA`-seq quantification. Nature biotechnology, 34(5):525–527, 2016.">Bray <em>et al.</em>, 2016</a>, <a class="reference internal" href="#id133" title="Chelsea J-T Ju, Ruirui Li, Zhengliang Wu, Jyun-Yu Jiang, Zhao Yang, and Wei Wang. Fleximer: accurate quantification of term`RNA`-Seq via variable-length k-mers. In Proceedings of the 8th ACM International Conference on Bioinformatics, Computational Biology, and Health Informatics, 263–272. 2017.">Ju <em>et al.</em>, 2017</a>, <a class="reference internal" href="#id140" title="Marius Nicolae, Serghei Mangul, Ion I Măndoiu, and Alex Zelikovsky. Estimation of alternative splicing isoform frequencies from term`RNA`-Seq data. Algorithms for molecular biology, 6(1):1–13, 2011.">Nicolae <em>et al.</em>, 2011</a>, <a class="reference internal" href="#id137" title="Rob Patro, Geet Duggal, Michael I Love, Rafael A Irizarry, and Carl Kingsford. Salmon provides fast and bias-aware quantification of transcript expression. Nature methods, 14(4):417–419, 2017.">Patro <em>et al.</em>, 2017</a>, <a class="reference internal" href="#id136" title="Rob Patro, Stephen M Mount, and Carl Kingsford. Sailfish enables alignment-free isoform quantification from term`RNA`-seq reads using lightweight algorithms. Nature biotechnology, 32(5):462–464, 2014.">Patro <em>et al.</em>, 2014</a>, <a class="reference internal" href="#id138" title="Ernest Turro, Shu-Yi Su, Ângela Gonçalves, Lachlan JM Coin, Sylvia Richardson, and Alex Lewin. Haplotype and isoform specific expression estimation using multi-mapping term`RNA`-seq reads. Genome biology, 12(2):1–15, 2011.">Turro <em>et al.</em>, 2011</a>]</span>, and such a transcript-level representation has even been proposed for use in the clustering and dimensionality reduction of full-length single-cell RNA-seq data<span id="id73">[<a class="reference internal" href="#id147" title="Vasilis Ntranos, Govinda M Kamath, Jesse M Zhang, Lior Pachter, and David N Tse. Fast and accurate single-cell term`RNA`-seq analysis by clustering of transcript-compatibility counts. Genome biology, 17(1):1–14, 2016.">Ntranos <em>et al.</em>, 2016</a>]</span> and differential expression analysis of single-cell RNA-seq data<span id="id74">[<a class="reference internal" href="#id146" title="Vasilis Ntranos, Lynn Yi, Páll Melsted, and Lior Pachter. A discriminative learning approach to differential expression analysis for single-cell term`rna`-seq. Nature methods, 16(2):163–166, 2019.">Ntranos <em>et al.</em>, 2019</a>]</span>). In this case, instead of the resulting count matrix having dimensions <span class="math notranslate nohighlight">\(C \times G\)</span>, where <span class="math notranslate nohighlight">\(G\)</span> is the number of genes in the quantified annotation, it will have dimension <span class="math notranslate nohighlight">\(C \times E\)</span>, where <span class="math notranslate nohighlight">\(E\)</span> is the number of distinct <em>gene groups</em> (commonly called equivalence class labels) across all cells in the given sample. By propagating this information to the output count matrix, one can avoid the application of heuristic or probabilistic UMI resolution methods that can result in loss of data, or bias, in the counts used in downstream analyses. Of course, to make use of this information, downstream analysis methods must expect the information in this format. Further, those downstream methods must typically have a way to resolve these counts, eventually, to the level of genes, as the abundance of <em>gene groups</em> lacks the intuitive biological interpretability of gene abundance. Nonetheless, the benefits provided by such representations, in terms of conveying more complete and accurate information to downstream analysis tools, can be substantial, and tools taking advantage of such representations are being developed (e.g. <a class="reference external" href="https://github.com/SimoneTiberi/DifferentialRegulation">DifferentialRegulation</a>); this is still an active area of research.</p>
</section>
<section id="brief-discussion">
<h2><span class="section-number">3.7. </span>Brief discussion<a class="headerlink" href="#brief-discussion" title="Link to this heading">#</a></h2>
<p>To close this chapter, we convey some observations and suggestions that have arisen from recent benchmarking and review studies surrounding some of the common preprocessing tools described above <span id="id75">[<a class="reference internal" href="#id103" title="Ralf Schulze Brüning, Lukas Tombor, Marcel H Schulz, Stefanie Dimmeler, and David John. Comparative analysis of common alignment tools for single-cell term`RNA` sequencing. GigaScience, 2022. URL: https://doi.org/10.1093%2Fgigascience%2Fgiac001, doi:10.1093/gigascience/giac001.">Brüning <em>et al.</em>, 2022</a>, <a class="reference internal" href="#id104" title="Yue You, Luyi Tian, Shian Su, Xueyi Dong, Jafar S. Jabbari, Peter F. Hickey, and Matthew E. Ritchie. Benchmarking UMI-based single-cell term`RNA`-seq preprocessing workflows. Genome Biology, dec 2021. URL: https://doi.org/10.1186%2Fs13059-021-02552-3, doi:10.1186/s13059-021-02552-3.">You <em>et al.</em>, 2021</a>]</span>. It is, of course, important to note that the development of methods and tools for single-cell and single-nucleus RNA-seq raw data processing, as well as the continual evaluation of such methods, is an ongoing community effort. It is therefore often useful and reasonable, when performing your own analyses, to experiment with several different tools.</p>
<p>At the coarsest level, the most common tools can process data robustly and accurately. It has been suggested that with many common downstream analyses like clustering, and the methods used to perform them, the choice of preprocessing tool typically makes less difference than other steps in the analysis process <span id="id76">[<a class="reference internal" href="#id104" title="Yue You, Luyi Tian, Shian Su, Xueyi Dong, Jafar S. Jabbari, Peter F. Hickey, and Matthew E. Ritchie. Benchmarking UMI-based single-cell term`RNA`-seq preprocessing workflows. Genome Biology, dec 2021. URL: https://doi.org/10.1186%2Fs13059-021-02552-3, doi:10.1186/s13059-021-02552-3.">You <em>et al.</em>, 2021</a>]</span>. Nonetheless, it has also been observed that applying lightweight mapping restricted to the spliced transcriptome can increase the probability of spurious mapping and gene expression <span id="id77">[<a class="reference internal" href="#id103" title="Ralf Schulze Brüning, Lukas Tombor, Marcel H Schulz, Stefanie Dimmeler, and David John. Comparative analysis of common alignment tools for single-cell term`RNA` sequencing. GigaScience, 2022. URL: https://doi.org/10.1093%2Fgigascience%2Fgiac001, doi:10.1093/gigascience/giac001.">Brüning <em>et al.</em>, 2022</a>]</span>.</p>
<p>Ultimately, the choice of a specific tool largely depends on the task at hand, and the constraints on available computational resources. If performing a standard single-cell analysis, lightweight mapping-based methods are a good choice since they are faster (often considerably so) and more memory frugal than existing alignment-based tools. If performing single-nucleus RNA-seq analysis, <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> is an attractive option in particular, as it remains memory frugal and its index remains relatively small even as the transcriptome reference is expanded to include unspliced reference sequence. On the other hand, alignment-based methods are recommended if it is important to recover reads that map outside of the (extended) transcriptome or if the genomic mapping sites of reads are necessary for use in the relevant downstream tools or analyses (e.g., such as for differential transcript usage analysis with a tool like <code class="docutils literal notranslate"><span class="pre">sierra</span></code> <span id="id78">[<a class="reference internal" href="#id102" title="Ralph Patrick, David T. Humphreys, Vaibhao Janbandhu, Alicia Oshlack, Joshua W.K. Ho, Richard P. Harvey, and Kitty K. Lo. Sierra: discovery of differential transcript usage from polyA-captured single-cell term`RNA`-seq data. Genome Biology, jul 2020. URL: https://doi.org/10.1186%2Fs13059-020-02071-7, doi:10.1186/s13059-020-02071-7.">Patrick <em>et al.</em>, 2020</a>]</span>). Among the alignment-based pipelines, according to <span id="id79">Brüning <em>et al.</em> [<a class="reference internal" href="#id103" title="Ralf Schulze Brüning, Lukas Tombor, Marcel H Schulz, Stefanie Dimmeler, and David John. Comparative analysis of common alignment tools for single-cell term`RNA` sequencing. GigaScience, 2022. URL: https://doi.org/10.1093%2Fgigascience%2Fgiac001, doi:10.1093/gigascience/giac001.">2022</a>]</span>, <code class="docutils literal notranslate"><span class="pre">STARsolo</span></code> should be favored over <code class="docutils literal notranslate"><span class="pre">Cell</span> <span class="pre">Ranger</span></code> because the former is much faster than the latter, and requires less memory, while it is also capable of producing almost identical results.</p>
</section>
<section id="a-real-world-example">
<span id="raw-proc-example-workflow"></span><h2><span class="section-number">3.8. </span>A real-world example<a class="headerlink" href="#a-real-world-example" title="Link to this heading">#</a></h2>
<p>Given that we have covered the concepts underlying various approaches for raw data processing, we now turn our attention to demonstrating how a specific tool (in this case, <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code>) can be used to process a small example dataset. To start, we need the sequenced reads from a single-cell experiment in <a class="reference external" href="https://en.wikipedia.org/wiki/FASTQ_format">FASTQ format</a> and the reference (e.g., transcriptome) against which the reads will be mapped. Usually, a reference includes the genome sequences and the corresponding gene annotations of the sequenced species in the <a class="reference external" href="https://en.wikipedia.org/wiki/FASTA_format">FASTA</a> and <a class="reference external" href="https://useast.ensembl.org/info/website/upload/gff.html">GTF</a> format, respectively.</p>
<p>In this example, we will use <em>chromosome 5</em> of the human genome and its related gene annotations as the reference, which is a subset of the Human reference, <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/release-notes/build#GRCh38_2020A">GRCh38 (GENCODE v32/Ensembl 98) reference</a> from the 10x Genomics reference build. Correspondingly, we extract the subset of reads that map to the generated reference from a <a class="reference external" href="https://www.10xgenomics.com/resources/datasets/200-sorted-cells-from-human-glioblastoma-multiforme-3-lt-v-3-1-3-1-low-6-0-0">human brain tumor dataset</a> from 10x Genomics.</p>
<p><a class="reference external" href="https://alevin-fry.readthedocs.io/en/latest/"><code class="docutils literal notranslate"><span class="pre">Alevin-fry</span></code></a><span id="id80">[<a class="reference internal" href="#id118" title="Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. Nature Methods, 19(3):316–322, 2022.">He <em>et al.</em>, 2022</a>]</span> is a fast, accurate, and memory-frugal single-cell and single-nucleus data processing tool. <a class="reference external" href="https://github.com/COMBINE-lab/simpleaf">Simpleaf</a> is a program, written in <a class="reference external" href="https://www.rust-lang.org/">rust</a>, that exposes a unified and simplified interface for processing some of the most common protocols and data types using the <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> pipeline. A nextflow-based <a class="reference external" href="https://github.com/COMBINE-lab/quantaf">workflow</a> tool also exists to process extensive collections of single-cell data. Here we will first show how to process single-cell raw data using two <code class="docutils literal notranslate"><span class="pre">simpleaf</span></code> commands. Then, we describe the complete set of <code class="docutils literal notranslate"><span class="pre">salmon</span> <span class="pre">alevin</span></code> and <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> commands to which these <code class="docutils literal notranslate"><span class="pre">simpleaf</span></code> commands correspond, to outline where the steps described in this section occur and to convey the possible different processing options. These commands will be run from the command line, and <a class="reference external" href="https://docs.conda.io/en/latest/"><code class="docutils literal notranslate"><span class="pre">conda</span></code></a> will be used for installing all of the software required for running this example.</p>
<section id="preparation">
<span id="raw-proc-example-prep"></span><h3><span class="section-number">3.8.1. </span>Preparation<a class="headerlink" href="#preparation" title="Link to this heading">#</a></h3>
<p>Before we start, we create a conda environment in the terminal and install the required package. <code class="docutils literal notranslate"><span class="pre">Simpleaf</span></code> depends on <a class="reference external" href="https://alevin-fry.readthedocs.io/en/latest/"><code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code></a>, <a class="reference external" href="https://salmon.readthedocs.io/en/latest/"><code class="docutils literal notranslate"><span class="pre">salmon</span></code></a> and <a class="reference external" href="https://github.com/COMBINE-lab/pyroe"><code class="docutils literal notranslate"><span class="pre">pyroe</span></code></a>. They are all available on <code class="docutils literal notranslate"><span class="pre">bioconda</span></code> and will be automatically installed when installing <code class="docutils literal notranslate"><span class="pre">simpleaf</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>create<span class="w"> </span>-n<span class="w"> </span>af<span class="w"> </span>-y<span class="w"> </span>-c<span class="w"> </span>bioconda<span class="w"> </span>simpleaf
conda<span class="w"> </span>activate<span class="w"> </span>af
</pre></div>
</div>
<div class="admonition-note-on-using-an-apple-silicon-based-device admonition">
<p class="admonition-title">Note on using an Apple silicon-based device</p>
<p>Conda does not currently build most packages natively for Apple silicon. Therefore, if you
are using a non-Intel-based Apple computer (e.g., with an M1(Pro/Max/Ultra) or M2 chip), you
should make sure to specify that your environment uses the Rosetta2 translation layer.
To do this, you can replace the above commands with the following (instructions adopted
from <a class="reference external" href="https://github.com/Haydnspass/miniforge#rosetta-on-mac-with-apple-silicon-hardware">here</a>):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">CONDA_SUBDIR</span><span class="o">=</span>osx-64<span class="w"> </span>conda<span class="w"> </span>create<span class="w"> </span>-n<span class="w"> </span>af<span class="w"> </span>-y<span class="w"> </span>-c<span class="w"> </span>bioconda<span class="w"> </span>simpleaf<span class="w">   </span><span class="c1"># create a new environment</span>
conda<span class="w"> </span>activate<span class="w"> </span>af
conda<span class="w"> </span>env<span class="w"> </span>config<span class="w"> </span>vars<span class="w"> </span><span class="nb">set</span><span class="w"> </span><span class="nv">CONDA_SUBDIR</span><span class="o">=</span>osx-64<span class="w">  </span><span class="c1"># subsequent commands use intel packages</span>
</pre></div>
</div>
</div>
<p>Next, we create a working directory, <code class="docutils literal notranslate"><span class="pre">af_xmpl_run</span></code>, and download and uncompress the example dataset from a remote host.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a working dir and go to the working directory</span>
<span class="c1">## The &amp;&amp; operator helps execute two commands using a single line of code.</span>
mkdir<span class="w"> </span>af_xmpl_run<span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span><span class="nb">cd</span><span class="w"> </span>af_xmpl_run

<span class="c1"># Fetch the example dataset and CB permit list and decompress them</span>
<span class="c1">## The pipe operator (|) passes the output of the wget command to the tar command.</span>
<span class="c1">## The dash operator (-) after `tar xzf` captures the output of the first command.</span>
<span class="c1">## - example dataset</span>
wget<span class="w"> </span>-qO-<span class="w"> </span>https://umd.box.com/shared/static/lx2xownlrhz3us8496tyu9c4dgade814.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>tar<span class="w"> </span>xzf<span class="w"> </span>-<span class="w"> </span>--strip-components<span class="o">=</span><span class="m">1</span><span class="w"> </span>-C<span class="w"> </span>.
<span class="c1">## The fetched folder containing the fastq files is called toy_read_fastq.</span>
<span class="nv">fastq_dir</span><span class="o">=</span><span class="s2">&quot;toy_read_fastq&quot;</span>
<span class="c1">## The fetched folder containing the human ref files is called toy_human_ref.</span>
<span class="nv">ref_dir</span><span class="o">=</span><span class="s2">&quot;toy_human_ref&quot;</span>

<span class="c1"># Fetch CB permit list</span>
<span class="c1">## the right chevron (&gt;) redirects the STDOUT to a file.</span>
wget<span class="w"> </span>-qO-<span class="w"> </span>https://raw.githubusercontent.com/10XGenomics/cellranger/master/lib/python/cellranger/barcodes/3M-february-2018.txt.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>gunzip<span class="w"> </span>-<span class="w"> </span>&gt;<span class="w"> </span>3M-february-2018.txt
</pre></div>
</div>
<p>With the reference files (the genome FASTA file and the gene annotation GTF file) and read records (the FASTQ files) ready, we can now apply the raw data processing pipeline discussed above to generate the gene count matrix.</p>
</section>
<section id="simplified-raw-data-processing-pipeline">
<span id="raw-proc-example-simpleaf"></span><h3><span class="section-number">3.8.2. </span>Simplified raw data processing pipeline<a class="headerlink" href="#simplified-raw-data-processing-pipeline" title="Link to this heading">#</a></h3>
<p><a class="reference external" href="https://github.com/COMBINE-lab/simpleaf">Simpleaf</a> is designed to simplify the <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> interface for single-cell and nucleus raw data processing. It encapsulates the whole processing pipeline into two steps:</p>
<ol class="arabic simple">
<li><p><a class="reference external" href="https://simpleaf.readthedocs.io/en/latest/index-command.html"><code class="docutils literal notranslate"><span class="pre">simpleaf</span> <span class="pre">index</span></code></a> indexes the provided reference or makes a <em>splici</em> reference (<u>splic</u>ed transcripts + <u>i</u>ntrons) and index it.</p></li>
<li><p><a class="reference external" href="https://simpleaf.readthedocs.io/en/latest/quant-command.html"><code class="docutils literal notranslate"><span class="pre">simpleaf</span> <span class="pre">quant</span></code></a> maps the sequencing reads against the indexed reference and quantifies the mapping records to generate a gene count matrix.</p></li>
</ol>
<p>More advanced usages and options for mapping with <code class="docutils literal notranslate"><span class="pre">simpleaf</span></code> can be found <a class="reference external" href="https://simpleaf.readthedocs.io/en/latest/">here</a>.</p>
<p>When running <code class="docutils literal notranslate"><span class="pre">simpleaf</span> <span class="pre">index</span></code>, if a genome FASTA file (<code class="docutils literal notranslate"><span class="pre">-f</span></code>) and a gene annotation GTF file(<code class="docutils literal notranslate"><span class="pre">-g</span></code>) are provided, it will generate a <em>splici</em> reference and index it; if only a transcriptome FASTA file is provided (<code class="docutils literal notranslate"><span class="pre">--refseq</span></code>), it will directly index it. Currently, we recommend the <em>splici</em> index.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># simpleaf needs the environment variable ALEVIN_FRY_HOME to store configuration and data.</span>
<span class="c1"># For example, the paths to the underlying programs it uses and the CB permit list</span>
mkdir<span class="w"> </span>alevin_fry_home<span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span><span class="nb">export</span><span class="w"> </span><span class="nv">ALEVIN_FRY_HOME</span><span class="o">=</span><span class="s1">&#39;alevin_fry_home&#39;</span>

<span class="c1"># the simpleaf set-paths command finds the path to the required tools and writes a configuration JSON file in the ALEVIN_FRY_HOME folder.</span>
simpleaf<span class="w"> </span>set-paths

<span class="c1"># simpleaf index</span>
<span class="c1"># Usage: simpleaf index -o out_dir [-f genome_fasta -g gene_annotation_GTF|--refseq transcriptome_fasta] -r read_length -t number_of_threads</span>
<span class="c1">## The -r read_lengh is the number of sequencing cycles performed by the sequencer to generate biological reads (read2 in Illumina).</span>
<span class="c1">## Publicly available datasets usually have the read length in the description. Sometimes they are called the number of cycles.</span>
simpleaf<span class="w"> </span>index<span class="w"> </span><span class="se">\</span>
-o<span class="w"> </span>simpleaf_index<span class="w"> </span><span class="se">\</span>
-f<span class="w"> </span>toy_human_ref/fasta/genome.fa<span class="w"> </span><span class="se">\</span>
-g<span class="w"> </span>toy_human_ref/genes/genes.gtf<span class="w"> </span><span class="se">\</span>
-r<span class="w"> </span><span class="m">90</span><span class="w"> </span><span class="se">\</span>
-t<span class="w"> </span><span class="m">8</span>
</pre></div>
</div>
<p>In the output directory <code class="docutils literal notranslate"><span class="pre">simpleaf_index</span></code>, the <code class="docutils literal notranslate"><span class="pre">ref</span></code> folder contains the <em>splici</em> reference; The <code class="docutils literal notranslate"><span class="pre">index</span></code> folder contains the salmon index built upon the <em>splici</em> reference.</p>
<p>The next step, <code class="docutils literal notranslate"><span class="pre">simpleaf</span> <span class="pre">quant</span></code>, consumes an index directory and the mapping record FASTQ files to generate a gene count matrix. This command encapsulates all the major steps discussed in this section, including mapping, cell barcode correction, and UMI resolution.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Collecting sequencing read files</span>
<span class="c1">## The reads1 and reads2 variables are defined by finding the filenames with the pattern &quot;_R1_&quot; and &quot;_R2_&quot; from the toy_read_fastq directory.</span>
<span class="nv">reads1_pat</span><span class="o">=</span><span class="s2">&quot;_R1_&quot;</span>
<span class="nv">reads2_pat</span><span class="o">=</span><span class="s2">&quot;_R2_&quot;</span>

<span class="c1">## The read files must be sorted and separated by a comma.</span>
<span class="c1">### The find command finds the files in the fastq_dir with the name pattern</span>
<span class="c1">### The sort command sorts the file names</span>
<span class="c1">### The awk command and the paste command together convert the file names into a comma-separated string.</span>
<span class="nv">reads1</span><span class="o">=</span><span class="s2">&quot;</span><span class="k">$(</span>find<span class="w"> </span>-L<span class="w"> </span><span class="si">${</span><span class="nv">fastq_dir</span><span class="si">}</span><span class="w"> </span>-name<span class="w"> </span><span class="s2">&quot;*</span><span class="nv">$reads1_pat</span><span class="s2">*&quot;</span><span class="w"> </span>-type<span class="w"> </span>f<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span>-v<span class="w"> </span><span class="nv">OFS</span><span class="o">=</span>,<span class="w"> </span><span class="s1">&#39;{$1=$1;print}&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>paste<span class="w"> </span>-sd,<span class="k">)</span><span class="s2">&quot;</span>
<span class="nv">reads2</span><span class="o">=</span><span class="s2">&quot;</span><span class="k">$(</span>find<span class="w"> </span>-L<span class="w"> </span><span class="si">${</span><span class="nv">fastq_dir</span><span class="si">}</span><span class="w"> </span>-name<span class="w"> </span><span class="s2">&quot;*</span><span class="nv">$reads2_pat</span><span class="s2">*&quot;</span><span class="w"> </span>-type<span class="w"> </span>f<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span>-v<span class="w"> </span><span class="nv">OFS</span><span class="o">=</span>,<span class="w"> </span><span class="s1">&#39;{$1=$1;print}&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>paste<span class="w"> </span>-sd,<span class="k">)</span><span class="s2">&quot;</span>

<span class="c1"># simpleaf quant</span>
<span class="c1">## Usage: simpleaf quant -c chemistry -t threads -1 reads1 -2 reads2 -i index -u [unspliced permit list] -r resolution -m t2g_3col -o output_dir</span>
simpleaf<span class="w"> </span>quant<span class="w"> </span><span class="se">\</span>
-c<span class="w"> </span>10xv3<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span><span class="w"> </span><span class="se">\</span>
-1<span class="w"> </span><span class="nv">$reads1</span><span class="w"> </span>-2<span class="w"> </span><span class="nv">$reads2</span><span class="w"> </span><span class="se">\</span>
-i<span class="w"> </span>simpleaf_index/index<span class="w"> </span><span class="se">\</span>
-u<span class="w"> </span>-r<span class="w"> </span>cr-like<span class="w"> </span><span class="se">\</span>
-m<span class="w"> </span>simpleaf_index/index/t2g_3col.tsv<span class="w"> </span><span class="se">\</span>
-o<span class="w"> </span>simpleaf_quant
</pre></div>
</div>
<p>After running these commands, the resulting quantification information can be found in the <code class="docutils literal notranslate"><span class="pre">simpleaf_quant/af_quant/alevin</span></code> folder. Within this directory, there are three files: <code class="docutils literal notranslate"><span class="pre">quants_mat.mtx</span></code>, <code class="docutils literal notranslate"><span class="pre">quants_mat_cols.txt</span></code>, and <code class="docutils literal notranslate"><span class="pre">quants_mat_rows.txt</span></code>, which correspond, respectively, to the count matrix, the gene names for each column of this matrix, and the corrected, filtered cell barcodes for each row of this matrix. The tail lines of these files are shown below. Of note here is the fact that <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> was run in the USA-mode (<u>u</u>nspliced, <u>s</u>pliced, and <u>a</u>mbiguous mode), and so quantification was performed for both the spliced and unspliced status of each gene — the resulting <code class="docutils literal notranslate"><span class="pre">quants_mat_cols.txt</span></code> file will then have a number of rows equal to 3 times the number of annotated genes which correspond, to the names used for the spliced (S), unspliced (U), and splicing-ambiguous variants (A) of each gene.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Each line in `quants_mat.mtx` represents</span>
<span class="c1"># a non-zero entry in the format row column entry</span>
$<span class="w"> </span>tail<span class="w"> </span>-3<span class="w"> </span>simpleaf_quant/af_quant/alevin/quants_mat.mtx
<span class="m">138</span><span class="w"> </span><span class="m">58</span><span class="w"> </span><span class="m">1</span>
<span class="m">139</span><span class="w"> </span><span class="m">9</span><span class="w"> </span><span class="m">1</span>
<span class="m">139</span><span class="w"> </span><span class="m">37</span><span class="w"> </span><span class="m">1</span>

<span class="c1"># Each line in `quants_mat_cols.txt` is a splice status</span>
<span class="c1"># of a gene in the format (gene name)-(splice status)</span>
$<span class="w"> </span>tail<span class="w"> </span>-3<span class="w"> </span>simpleaf_quant/af_quant/alevin/quants_mat_cols.txt
ENSG00000120705-A
ENSG00000198961-A
ENSG00000245526-A

<span class="c1"># Each line in `quants_mat_rows.txt` is a corrected</span>
<span class="c1"># (and, potentially, filtered) cell barcode</span>
$<span class="w"> </span>tail<span class="w"> </span>-3<span class="w"> </span>simpleaf_quant/af_quant/alevin/quants_mat_rows.txt
TTCGATTTCTGAATCG
TGCTCGTGTTCGAAGG
ACTGTGAAGAAATTGC
</pre></div>
</div>
<p>We can load the count matrix into Python as an <a class="reference external" href="https://anndata.readthedocs.io/en/latest/"><code class="docutils literal notranslate"><span class="pre">AnnData</span></code></a> object using the <code class="docutils literal notranslate"><span class="pre">load_fry</span></code> function from <a class="reference external" href="https://github.com/COMBINE-lab/pyroe"><code class="docutils literal notranslate"><span class="pre">pyroe</span></code></a>. A similar function, <a class="reference external" href="https://rdrr.io/github/mikelove/fishpond/man/loadFry.html">loadFry</a>, has been implemented in the <a class="reference external" href="https://github.com/mikelove/fishpond"><code class="docutils literal notranslate"><span class="pre">fishpond</span></code></a> R package.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pyroe</span>

<span class="n">quant_dir</span> <span class="o">=</span> <span class="s1">&#39;simpleaf_quant/af_quant&#39;</span>
<span class="n">adata_sa</span> <span class="o">=</span> <span class="n">pyroe</span><span class="o">.</span><span class="n">load_fry</span><span class="p">(</span><span class="n">quant_dir</span><span class="p">)</span>
</pre></div>
</div>
<p>The default behavior loads the <code class="docutils literal notranslate"><span class="pre">X</span></code> layer of the <code class="docutils literal notranslate"><span class="pre">Anndata</span></code> object as the sum of the spliced and ambiguous counts for each gene. However, recent work<span id="id81">[<a class="reference internal" href="#id142" title="Allan-Hermann Pool, Helen Poldsam, Sisi Chen, Matt Thomson, and Yuki Oka. Enhanced recovery of single-cell term`RNA`-sequencing reads for missing gene expression data. bioRxiv, 2022. URL: https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449, arXiv:https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449.full.pdf, doi:10.1101/2022.04.26.489449.">Pool <em>et al.</em>, 2022</a>]</span> and <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/release-notes">updated practices</a> suggest that the inclusion of intronic counts, even in single-cell RNA-seq data, may increase sensitivity and benefit downstream analyses. While the best way to make use of this information is the subject of ongoing research, since <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> automatically quantifies spliced, unspliced, and ambiguous reads in each sample, the count matrix containing the total counts for each gene can be simply obtained as follows:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pyroe</span>

<span class="n">quant_dir</span> <span class="o">=</span> <span class="s1">&#39;simpleaf_quant/af_quant&#39;</span>
<span class="n">adata_usa</span> <span class="o">=</span> <span class="n">pyroe</span><span class="o">.</span><span class="n">load_fry</span><span class="p">(</span><span class="n">quant_dir</span><span class="p">,</span> <span class="n">output_format</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;X&#39;</span> <span class="p">:</span> <span class="p">[</span><span class="s1">&#39;U&#39;</span><span class="p">,</span><span class="s1">&#39;S&#39;</span><span class="p">,</span><span class="s1">&#39;A&#39;</span><span class="p">]})</span>
</pre></div>
</div>
</section>
<section id="the-complete-alevin-fry-pipeline">
<span id="raw-proc-example-map"></span><h3><span class="section-number">3.8.3. </span>The complete alevin-fry pipeline<a class="headerlink" href="#the-complete-alevin-fry-pipeline" title="Link to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Simpleaf</span></code> makes it possible to process single-cell raw data in the “standard” way with a few commands. Next, we will show how to generate the identical quantification result by explicitly calling the <code class="docutils literal notranslate"><span class="pre">pyroe</span></code>, <code class="docutils literal notranslate"><span class="pre">salmon</span></code>, and <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> commands. On top of the pedagogical value, knowing the exact command of each step will be helpful if only a part of the pipeline needs to be rerun or if some parameters not currently exposed by <code class="docutils literal notranslate"><span class="pre">simpleaf</span></code> need to be specified.</p>
<p>Please note that the commands in the <a class="reference internal" href="#raw-proc-example-prep"><span class="std std-ref">Preparation</span></a> section should be executed in advance. All the tools called in the following commands, <code class="docutils literal notranslate"><span class="pre">pyroe</span></code>, <code class="docutils literal notranslate"><span class="pre">salmon</span></code>, and <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code>, have already been installed when installing <code class="docutils literal notranslate"><span class="pre">simpleaf</span></code>.</p>
<section id="building-the-index">
<h4><span class="section-number">3.8.3.1. </span>Building the index<a class="headerlink" href="#building-the-index" title="Link to this heading">#</a></h4>
<p>First, we process the genome FASTA file and gene annotation GTF file to obtain the <em>splici</em> index. The commands in the following code chunk are analogous to the <code class="docutils literal notranslate"><span class="pre">simpleaf</span> <span class="pre">index</span></code> command discussed above. This includes two steps:</p>
<ol class="arabic simple">
<li><p>Building the <em>splici</em> reference (<u>splic</u>ed transcripts + <u>i</u>ntrons) by calling <code class="docutils literal notranslate"><span class="pre">pyroe</span> <span class="pre">make-splici</span></code>, using the genome and gene annotation file</p></li>
<li><p>Indexing the <em>splici</em> reference by calling <code class="docutils literal notranslate"><span class="pre">salmon</span> <span class="pre">index</span></code></p></li>
</ol>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># make splici reference</span>
<span class="c1">## Usage: pyroe make-splici genome_file gtf_file read_length out_dir</span>
<span class="c1">## The read_lengh is the number of sequencing cycles performed by the sequencer. Ask your technician if you are not sure about it.</span>
<span class="c1">## Publicly available datasets usually have the read length in the description.</span>
pyroe<span class="w"> </span>make-splici<span class="w"> </span><span class="se">\</span>
<span class="si">${</span><span class="nv">ref_dir</span><span class="si">}</span>/fasta/genome.fa<span class="w"> </span><span class="se">\</span>
<span class="si">${</span><span class="nv">ref_dir</span><span class="si">}</span>/genes/genes.gtf<span class="w"> </span><span class="se">\</span>
<span class="m">90</span><span class="w"> </span><span class="se">\</span>
splici_rl90_ref

<span class="c1"># Index the reference</span>
<span class="c1">## Usage: salmon index -t extend_txome.fa -i idx_out_dir -p num_threads</span>
<span class="c1">## The $() expression runs the command inside and puts the output in place.</span>
<span class="c1">## Please ensure that only one file ends with &quot;.fa&quot; in the `splici_ref` folder.</span>
salmon<span class="w"> </span>index<span class="w"> </span><span class="se">\</span>
-t<span class="w"> </span><span class="k">$(</span>ls<span class="w"> </span>splici_rl90_ref/*<span class="se">\.</span>fa<span class="k">)</span><span class="w"> </span><span class="se">\</span>
-i<span class="w"> </span>salmon_index<span class="w"> </span><span class="se">\</span>
-p<span class="w"> </span><span class="m">8</span>
</pre></div>
</div>
<p>The <em>splici</em> index can be found in the <code class="docutils literal notranslate"><span class="pre">salmon_index</span></code> directory.</p>
</section>
<section id="mapping-and-quantification">
<span id="raw-proc-example-quant"></span><h4><span class="section-number">3.8.3.2. </span>Mapping and quantification<a class="headerlink" href="#mapping-and-quantification" title="Link to this heading">#</a></h4>
<p>Next, we will map the sequencing reads recorded against the <em>splici</em> index by calling <a class="reference external" href="https://salmon.readthedocs.io/en/latest/alevin.html"><code class="docutils literal notranslate"><span class="pre">salmon</span> <span class="pre">alevin</span></code></a>. This will produce an output folder called <code class="docutils literal notranslate"><span class="pre">salmon_alevin</span></code> that contains all the information we need to process the mapped reads using <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Collect FASTQ files</span>
<span class="c1">## The filenames are sorted and separated by space.</span>
<span class="nv">reads1</span><span class="o">=</span><span class="s2">&quot;</span><span class="k">$(</span>find<span class="w"> </span>-L<span class="w"> </span><span class="nv">$fastq_dir</span><span class="w"> </span>-name<span class="w"> </span><span class="s2">&quot;*</span><span class="nv">$reads1_pat</span><span class="s2">*&quot;</span><span class="w"> </span>-type<span class="w"> </span>f<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span><span class="s1">&#39;{$1=$1;print}&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>paste<span class="w"> </span>-sd<span class="s1">&#39; &#39;</span><span class="k">)</span><span class="s2">&quot;</span>
<span class="nv">reads2</span><span class="o">=</span><span class="s2">&quot;</span><span class="k">$(</span>find<span class="w"> </span>-L<span class="w"> </span><span class="nv">$fastq_dir</span><span class="w"> </span>-name<span class="w"> </span><span class="s2">&quot;*</span><span class="nv">$reads2_pat</span><span class="s2">*&quot;</span><span class="w"> </span>-type<span class="w"> </span>f<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span><span class="p">|</span><span class="w"> </span>awk<span class="w"> </span><span class="s1">&#39;{$1=$1;print}&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>paste<span class="w"> </span>-sd<span class="s1">&#39; &#39;</span><span class="k">)</span><span class="s2">&quot;</span>

<span class="c1"># Mapping</span>
<span class="c1">## Usage: salmon alevin -i index_dir -l library_type -1 reads1_files -2 reads2_files -p num_threads -o output_dir</span>
<span class="c1">## The variable reads1 and reads2 defined above are passed in using ${}.</span>
salmon<span class="w"> </span>alevin<span class="w"> </span><span class="se">\</span>
-i<span class="w"> </span>salmon_index<span class="w"> </span><span class="se">\</span>
-l<span class="w"> </span>ISR<span class="w"> </span><span class="se">\</span>
-1<span class="w"> </span><span class="si">${</span><span class="nv">reads1</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
-2<span class="w"> </span><span class="si">${</span><span class="nv">reads2</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
-p<span class="w"> </span><span class="m">8</span><span class="w"> </span><span class="se">\</span>
-o<span class="w"> </span>salmon_alevin<span class="w"> </span><span class="se">\</span>
--chromiumV3<span class="w"> </span><span class="se">\</span>
--sketch
</pre></div>
</div>
<p>Then, we execute the cell barcode correction and UMI resolution step using <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code>. This procedure involves three <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> commands:</p>
<ol class="arabic simple">
<li><p>The <a class="reference external" href="https://alevin-fry.readthedocs.io/en/latest/generate_permit_list.html"><code class="docutils literal notranslate"><span class="pre">generate-permit-list</span></code></a> command is used for cell barcode correction.</p></li>
<li><p>The <a class="reference external" href="https://alevin-fry.readthedocs.io/en/latest/collate.html"><code class="docutils literal notranslate"><span class="pre">collate</span></code></a> command filters out invalid mapping records, corrects cell barcodes and collates mapping records originating from the same corrected cell barcode.</p></li>
<li><p>The <a class="reference external" href="https://alevin-fry.readthedocs.io/en/latest/quant.html"><code class="docutils literal notranslate"><span class="pre">quant</span></code></a> command performs UMI resolution and quantification.</p></li>
</ol>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Cell barcode correction</span>
<span class="c1">## Usage: alevin-fry generate-permit-list -u CB_permit_list -d expected_orientation -o gpl_out_dir</span>
<span class="c1">## Here, the reads that map to the reverse complement strand of transcripts are filtered out by specifying `-d fw`.</span>
alevin-fry<span class="w"> </span>generate-permit-list<span class="w"> </span><span class="se">\</span>
-u<span class="w"> </span>3M-february-2018.txt<span class="w"> </span><span class="se">\</span>
-d<span class="w"> </span>fw<span class="w"> </span><span class="se">\</span>
-i<span class="w"> </span>salmon_alevin<span class="w"> </span><span class="se">\</span>
-o<span class="w"> </span>alevin_fry_gpl

<span class="c1"># Filter mapping information</span>
<span class="c1">## Usage: alevin-fry collate -i gpl_out_dir -r alevin_map_dir -t num_threads</span>
alevin-fry<span class="w"> </span>collate<span class="w"> </span><span class="se">\</span>
-i<span class="w"> </span>alevin_fry_gpl<span class="w"> </span><span class="se">\</span>
-r<span class="w"> </span>salmon_alevin<span class="w"> </span><span class="se">\</span>
-t<span class="w"> </span><span class="m">8</span>

<span class="c1"># UMI resolution + quantification</span>
<span class="c1">## Usage: alevin-fry quant -r resolution -m txp_to_gene_mapping -i gpl_out_dir -o quant_out_dir -t num_threads</span>
<span class="c1">## The file ends with `3col.tsv` in the splici_ref folder will be passed to the -m argument.</span>
<span class="c1">## Please ensure that there is only one such file in the `splici_ref` folder.</span>
alevin-fry<span class="w"> </span>quant<span class="w"> </span>-r<span class="w"> </span>cr-like<span class="w"> </span><span class="se">\</span>
-m<span class="w"> </span><span class="k">$(</span>ls<span class="w"> </span>splici_rl90_ref/*3col.tsv<span class="k">)</span><span class="w"> </span><span class="se">\</span>
-i<span class="w"> </span>alevin_fry_gpl<span class="w"> </span><span class="se">\</span>
-o<span class="w"> </span>alevin_fry_quant<span class="w"> </span><span class="se">\</span>
-t<span class="w"> </span><span class="m">8</span>
</pre></div>
</div>
<p>After running these commands, the resulting quantification information can be found in <code class="docutils literal notranslate"><span class="pre">alevin_fry_quant/alevin</span></code>. Other relevant information concerning the mapping, CB correction, and UMI resolution steps can be found in the <code class="docutils literal notranslate"><span class="pre">salmon_alevin</span></code>, <code class="docutils literal notranslate"><span class="pre">alevin_fry_gpl</span></code>, and <code class="docutils literal notranslate"><span class="pre">alevin_fry_quant</span></code> folders, respectively.</p>
<p>In the example given here, we demonstrate using <code class="docutils literal notranslate"><span class="pre">simpleaf</span></code> and <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> to process a 10x Chromium 3’ v3 dataset. <code class="docutils literal notranslate"><span class="pre">Alevin-fry</span></code> and <code class="docutils literal notranslate"><span class="pre">simpleaf</span></code> provide many other options for processing different single-cell protocols, including but not limited to Dropseq<span id="id82">[<a class="reference internal" href="#id150" title="Evan Z. Macosko, Anindita Basu, Rahul Satija, James Nemesh, Karthik Shekhar, Melissa Goldman, Itay Tirosh, Allison R. Bialas, Nolan Kamitaki, Emily M. Martersteck, John J. Trombetta, David A. Weitz, Joshua R. Sanes, Alex K. Shalek, Aviv Regev, and Steven A. McCarroll. Highly parallel genome-wide expression profiling of individual cells using nanoliter droplets. Cell, 161(5):1202-1214, May 2015. URL: https://doi.org/10.1016/j.cell.2015.05.002, doi:10.1016/j.cell.2015.05.002.">Macosko <em>et al.</em>, 2015</a>]</span>, sci-RNA-seq3<span id="id83">[<a class="reference internal" href="#id148" title="Junyue Cao, Malte Spielmann, Xiaojie Qiu, Xingfan Huang, Daniel M. Ibrahim, Andrew J. Hill, Fan Zhang, Stefan Mundlos, Lena Christiansen, Frank J. Steemers, Cole Trapnell, and Jay Shendure. The single-cell transcriptional landscape of mammalian organogenesis. Nature, 566(7745):496–502, February 2019. URL: https://doi.org/10.1038/s41586-019-0969-x, doi:10.1038/s41586-019-0969-x.">Cao <em>et al.</em>, 2019</a>]</span> and other 10x Chromium platforms. A more comprehensive list and description of available options for different stages of processing can be found in the <a class="reference external" href="https://alevin-fry.readthedocs.io/en/latest/"><code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code></a> and <a class="reference external" href="https://github.com/COMBINE-lab/simpleaf"><code class="docutils literal notranslate"><span class="pre">simpleaf</span></code></a> documentation. <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> also provides a <a class="reference external" href="https://www.nextflow.io/docs/latest/">nextflow</a>-based workflow, called <a class="reference external" href="https://github.com/COMBINE-lab/quantaf">quantaf</a>, for conveniently processing many samples from a simply-defined sample sheet.</p>
<p>Of course, similar resources exist for many of the other raw data processing tools referenced and described throughout this section, including <a class="reference external" href="https://github.com/sdparekh/zUMIs/wiki"><code class="docutils literal notranslate"><span class="pre">zUMIs</span></code></a><span id="id84">[<a class="reference internal" href="#id121" title="Swati Parekh, Christoph Ziegenhain, Beate Vieth, Wolfgang Enard, and Ines Hellmann. zUMIs - a fast and flexible pipeline to process term`RNA` sequencing data with UMIs. GigaScience, May 2018. URL: https://doi.org/10.1093/gigascience/giy059, doi:10.1093/gigascience/giy059.">Parekh <em>et al.</em>, 2018</a>]</span>, <a class="reference external" href="https://salmon.readthedocs.io/en/latest/alevin.html"><code class="docutils literal notranslate"><span class="pre">alevin</span></code></a><span id="id85">[<a class="reference internal" href="#id111" title="Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. Genome biology, 20(1):1–16, 2019.">Srivastava <em>et al.</em>, 2019</a>]</span>, <a class="reference external" href="https://www.kallistobus.tools/"><code class="docutils literal notranslate"><span class="pre">kallisto|bustools</span></code></a><span id="id86">[<a class="reference internal" href="#id116" title="Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. Nature Biotechnology, 39(7):813–818, April 2021. URL: https://doi.org/10.1038/s41587-021-00870-2, doi:10.1038/s41587-021-00870-2.">Melsted <em>et al.</em>, 2021</a>]</span>, <a class="reference external" href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md"><code class="docutils literal notranslate"><span class="pre">STARsolo</span></code></a><span id="id87">[<a class="reference internal" href="#id115" title="Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. bioRxiv, 2021.">Kaminow <em>et al.</em>, 2021</a>]</span> and <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger"><code class="docutils literal notranslate"><span class="pre">CellRanger</span></code></a>. The <a class="reference external" href="https://nf-co.re/scrnaseq"><code class="docutils literal notranslate"><span class="pre">scrnaseq</span></code></a> pipeline from <a class="reference external" href="https://nf-co.re/"><code class="docutils literal notranslate"><span class="pre">nf-core</span></code></a> also provides a nextflow-based pipeline for processing single-cell RNA-seq data generated using a range of different chemistries and integrates several of the tools described in this section.</p>
</section>
</section>
</section>
<section id="useful-links">
<span id="raw-proc-useful-links"></span><h2><span class="section-number">3.9. </span>Useful links<a class="headerlink" href="#useful-links" title="Link to this heading">#</a></h2>
<p><a class="reference external" href="https://combine-lab.github.io/alevin-fry-tutorials/">Alevin-fry tutorials</a> provide tutorials for processing different types of data.</p>
<p><a class="reference external" href="https://github.com/COMBINE-lab/pyroe"><code class="docutils literal notranslate"><span class="pre">Pyroe</span></code></a> in python and <a class="reference external" href="https://github.com/COMBINE-lab/roe"><code class="docutils literal notranslate"><span class="pre">roe</span></code></a> in R provide helper functions for processing <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> quantification information. They also provide an interface to the preprocessed datasets in <a class="reference external" href="https://combine-lab.github.io/quantaf"><code class="docutils literal notranslate"><span class="pre">quantaf</span></code></a>.</p>
<p><a class="reference external" href="https://github.com/COMBINE-lab/quantaf"><code class="docutils literal notranslate"><span class="pre">Quantaf</span></code></a> is a nextflow-based workflow of the <code class="docutils literal notranslate"><span class="pre">alevin-fry</span></code> pipeline for conveniently processing a large number of single-cell and single-nucleus data based on the input sheets. The preprocessed quantification information of publicly available single-cell datasets is available on its <a class="reference external" href="https://combine-lab.github.io/quantaf">webpage</a>.</p>
<p><a class="reference external" href="https://github.com/COMBINE-lab/simpleaf"><code class="docutils literal notranslate"><span class="pre">Simpleaf</span></code></a> is a wrapper of the alevin-fry workflow that allows executing the whole pipeline, from making <em>splici</em> reference to quantification as shown in the above example, using only two commands.</p>
<p>Tutorials for processing scRNA-seq raw data from <a class="reference external" href="https://galaxyproject.org/">the galaxy project</a> can be found at <a class="reference external" href="https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/scrna-preprocessing-tenx/tutorial.html">here</a> and <a class="reference external" href="https://training.galaxyproject.org/training-material/topics/transcriptomics/tutorials/scrna-preprocessing/tutorial.html">here</a>.</p>
<p>Tutorials for explaining and evaluating FastQC report are available from <a class="reference external" href="https://rtsf.natsci.msu.edu/genomics/tech-notes/fastqc-tutorial-and-faq/">MSU</a>, <a class="reference external" href="https://hbctraining.github.io/Intro-to-rnaseq-hpc-salmon/lessons/qc_fastqc_assessment.html">the HBC training program</a>, <a class="reference external" href="https://training.galaxyproject.org/training-material/topics/sequence-analysis/tutorials/quality-control/tutorial.html">Galaxy Training</a> and <a class="reference external" href="https://sequencing.qcfail.com/software/fastqc/">the QC Fail website</a>.</p>
</section>
<section id="references">
<span id="raw-proc-references"></span><h2><span class="section-number">3.10. </span>References<a class="headerlink" href="#references" title="Link to this heading">#</a></h2>
<div class="docutils container" id="id88">
<div role="list" class="citation-list">
<div class="citation" id="id127" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id66">rawARJ+20</a><span class="fn-bracket">]</span></span>
<p>Marcus Alvarez, Elior Rahmani, Brandon Jew, Kristina M. Garske, Zong Miao, Jihane N. Benhammou, Chun Jimmie Ye, Joseph R. Pisegna, Kirsi H. Pietiläinen, Eran Halperin, and Päivi Pajukanta. Enhancing droplet-based single-nucleus term`RNA`-seq resolution using the semi-supervised machine learning classifier DIEM. <em>Scientific Reports</em>, July 2020. URL: <a class="reference external" href="https://doi.org/10.1038/s41598-020-67513-5">https://doi.org/10.1038/s41598-020-67513-5</a>, <a class="reference external" href="https://doi.org/10.1038/s41598-020-67513-5">doi:10.1038/s41598-020-67513-5</a>.</p>
</div>
<div class="citation" id="id132" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawBK19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id39">1</a>,<a role="doc-backlink" href="#id69">2</a>)</span>
<p>Abha S Bais and Dennis Kostka. Scds: computational annotation of doublets in single-cell term`RNA` sequencing data. <em>Bioinformatics</em>, 36(4):1150–1158, September 2019. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btz698">https://doi.org/10.1093/bioinformatics/btz698</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btz698">doi:10.1093/bioinformatics/btz698</a>.</p>
</div>
<div class="citation" id="id145" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id69">rawBFL+20</a><span class="fn-bracket">]</span></span>
<p>Nicholas J. Bernstein, Nicole L. Fong, Irene Lam, Margaret A. Roy, David G. Hendrickson, and David R. Kelley. Solo: Doublet Identification in Single-Cell term`RNA`-Seq via Semi-Supervised Deep Learning. <em>Cell Systems</em>, 11(1):95–101.e5, July 2020. URL: <a class="reference external" href="https://doi.org/10.1016/j.cels.2020.05.010">https://doi.org/10.1016/j.cels.2020.05.010</a>, <a class="reference external" href="https://doi.org/10.1016/j.cels.2020.05.010">doi:10.1016/j.cels.2020.05.010</a>.</p>
</div>
<div class="citation" id="id114" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id48">rawBWC+15</a><span class="fn-bracket">]</span></span>
<p>Sayantan Bose, Zhenmao Wan, Ambrose Carr, Abbas H. Rizvi, Gregory Vieira, Dana Pe'er, and Peter A. Sims. Scalable microfluidics for single-cell term`RNA` printing and sequencing. <em>Genome Biology</em>, June 2015. URL: <a class="reference external" href="https://doi.org/10.1186/s13059-015-0684-3">https://doi.org/10.1186/s13059-015-0684-3</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-015-0684-3">doi:10.1186/s13059-015-0684-3</a>.</p>
</div>
<div class="citation" id="id135" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawBPMP16<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id22">1</a>,<a role="doc-backlink" href="#id72">2</a>)</span>
<p>Nicolas L Bray, Harold Pimentel, Páll Melsted, and Lior Pachter. Near-optimal probabilistic term`RNA`-seq quantification. <em>Nature biotechnology</em>, 34(5):525–527, 2016.</p>
</div>
<div class="citation" id="id141" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id31">rawBruningTS+22</a><span class="fn-bracket">]</span></span>
<p>Ralf Schulze Brüning, Lukas Tombor, Marcel H Schulz, Stefanie Dimmeler, and David John. Comparative analysis of common alignment tools for single-cell term`RNA` sequencing. <em>GigaScience</em>, 2022.</p>
</div>
<div class="citation" id="id103" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawBTS+22<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id75">1</a>,<a role="doc-backlink" href="#id77">2</a>,<a role="doc-backlink" href="#id79">3</a>)</span>
<p>Ralf Schulze Brüning, Lukas Tombor, Marcel H Schulz, Stefanie Dimmeler, and David John. Comparative analysis of common alignment tools for single-cell term`RNA` sequencing. <em>GigaScience</em>, 2022. URL: <a class="reference external" href="https://doi.org/10.1093%2Fgigascience%2Fgiac001">https://doi.org/10.1093%2Fgigascience%2Fgiac001</a>, <a class="reference external" href="https://doi.org/10.1093/gigascience/giac001">doi:10.1093/gigascience/giac001</a>.</p>
</div>
<div class="citation" id="id148" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id83">rawCSQ+19</a><span class="fn-bracket">]</span></span>
<p>Junyue Cao, Malte Spielmann, Xiaojie Qiu, Xingfan Huang, Daniel M. Ibrahim, Andrew J. Hill, Fan Zhang, Stefan Mundlos, Lena Christiansen, Frank J. Steemers, Cole Trapnell, and Jay Shendure. The single-cell transcriptional landscape of mammalian organogenesis. <em>Nature</em>, 566(7745):496–502, February 2019. URL: <a class="reference external" href="https://doi.org/10.1038/s41586-019-0969-x">https://doi.org/10.1038/s41586-019-0969-x</a>, <a class="reference external" href="https://doi.org/10.1038/s41586-019-0969-x">doi:10.1038/s41586-019-0969-x</a>.</p>
</div>
<div class="citation" id="id96" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id12">rawCPM92</a><span class="fn-bracket">]</span></span>
<p>Kun-Mao Chao, William R. Pearson, and Webb Miller. Aligning two sequences within a specified diagonal band. <em>Bioinformatics</em>, 8(5):481–487, 1992. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/8.5.481">https://doi.org/10.1093/bioinformatics/8.5.481</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/8.5.481">doi:10.1093/bioinformatics/8.5.481</a>.</p>
</div>
<div class="citation" id="id129" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawDSC+19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id39">1</a>,<a role="doc-backlink" href="#id69">2</a>)</span>
<p>Erica A.K. DePasquale, Daniel J. Schnell, Pieter-Jan Van Camp, Íñigo Valiente-Aland\'ı, Burns C. Blaxall, H. Leighton Grimes, Harinder Singh, and Nathan Salomonis. DoubletDecon: deconvoluting doublets from single-cell term`RNA`-sequencing data. <em>Cell Reports</em>, 29(6):1718–1727.e8, November 2019. URL: <a class="reference external" href="https://doi.org/10.1016/j.celrep.2019.09.082">https://doi.org/10.1016/j.celrep.2019.09.082</a>, <a class="reference external" href="https://doi.org/10.1016/j.celrep.2019.09.082">doi:10.1016/j.celrep.2019.09.082</a>.</p>
</div>
<div class="citation" id="id149" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawDDS+13<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id18">1</a>,<a role="doc-backlink" href="#id29">2</a>)</span>
<p>Alexander Dobin, Carrie A Davis, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha, Philippe Batut, Mark Chaisson, and Thomas R Gingeras. Star: ultrafast universal rna-seq aligner. <em>Bioinformatics</em>, 29(1):15–21, 2013.</p>
</div>
<div class="citation" id="id109" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawFDF+20<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id2">1</a>,<a role="doc-backlink" href="#id3">2</a>)</span>
<p>Rick Farouni, Haig Djambazian, Lorenzo E Ferri, Jiannis Ragoussis, and Hamed S Najafabadi. Model-based analysis of sample index hopping reveals its widespread artifacts in multiplexed single-cell term`RNA`-sequencing. <em>Nature communications</em>, 11(1):1–8, 2020.</p>
</div>
<div class="citation" id="id100" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id16">rawFar07</a><span class="fn-bracket">]</span></span>
<p>Michael Farrar. Striped Smith–Waterman speeds database searches six times over other SIMD implementations. <em>Bioinformatics</em>, 23(2):156–161, 2007.</p>
</div>
<div class="citation" id="id144" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id47">rawGP21</a><span class="fn-bracket">]</span></span>
<p>Gennady Gorin and Lior Pachter. Length Biases in Single-Cell term`RNA` Sequencing of pre-mRNA. <em>bioRxiv</em>, 2021. URL: <a class="reference external" href="https://www.biorxiv.org/content/early/2021/07/31/2021.07.30.454514">https://www.biorxiv.org/content/early/2021/07/31/2021.07.30.454514</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.biorxiv.org/content/early/2021/07/31/2021.07.30.454514.full.pdf">arXiv:https://www.biorxiv.org/content/early/2021/07/31/2021.07.30.454514.full.pdf</a>, <a class="reference external" href="https://doi.org/10.1101/2021.07.30.454514">doi:10.1101/2021.07.30.454514</a>.</p>
</div>
<div class="citation" id="id118" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawHZS+22<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id11">1</a>,<a role="doc-backlink" href="#id24">2</a>,<a role="doc-backlink" href="#id31">3</a>,<a role="doc-backlink" href="#id32">4</a>,<a role="doc-backlink" href="#id34">5</a>,<a role="doc-backlink" href="#id35">6</a>,<a role="doc-backlink" href="#id40">7</a>,<a role="doc-backlink" href="#id48">8</a>,<a role="doc-backlink" href="#id51">9</a>,<a role="doc-backlink" href="#id55">10</a>,<a role="doc-backlink" href="#id61">11</a>,<a role="doc-backlink" href="#id63">12</a>,<a role="doc-backlink" href="#id65">13</a>,<a role="doc-backlink" href="#id67">14</a>,<a role="doc-backlink" href="#id70">15</a>,<a role="doc-backlink" href="#id80">16</a>)</span>
<p>Dongze He, Mohsen Zakeri, Hirak Sarkar, Charlotte Soneson, Avi Srivastava, and Rob Patro. Alevin-fry unlocks rapid, accurate and memory-frugal quantification of single-cell RNA-seq data. <em>Nature Methods</em>, 19(3):316–322, 2022.</p>
</div>
<div class="citation" id="id124" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id66">rawHWC+21</a><span class="fn-bracket">]</span></span>
<p>Cody N. Heiser, Victoria M. Wang, Bob Chen, Jacob J. Hughey, and Ken S. Lau. Automated quality control and cell identification of droplet-based single-cell data using dropkick. <em>Genome Research</em>, 31(10):1742–1752, April 2021. URL: <a class="reference external" href="https://doi.org/10.1101/gr.271908.120">https://doi.org/10.1101/gr.271908.120</a>, <a class="reference external" href="https://doi.org/10.1101/gr.271908.120">doi:10.1101/gr.271908.120</a>.</p>
</div>
<div class="citation" id="id125" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id66">rawHFW+21</a><span class="fn-bracket">]</span></span>
<p>Ariel A Hippen, Matias M Falco, Lukas M Weber, Erdogan Pekcan Erkan, Kaiyang Zhang, Jennifer Anne Doherty, Anna Vähärautio, Casey S Greene, and Stephanie C Hicks. miQC: An adaptive probabilistic framework for quality control of single-cell term`RNA`-sequencing data. <em>PLoS computational biology</em>, 17(8):e1009290, 2021.</p>
</div>
<div class="citation" id="id113" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id48">rawIZJ+13</a><span class="fn-bracket">]</span></span>
<p>Saiful Islam, Amit Zeisel, Simon Joost, Gioele La Manno, Pawel Zajac, Maria Kasper, Peter Lönnerberg, and Sten Linnarsson. Quantitative single-cell term`RNA`-seq with unique molecular identifiers. <em>Nature Methods</em>, 11(2):163–166, December 2013. URL: <a class="reference external" href="https://doi.org/10.1038/nmeth.2772">https://doi.org/10.1038/nmeth.2772</a>, <a class="reference external" href="https://doi.org/10.1038/nmeth.2772">doi:10.1038/nmeth.2772</a>.</p>
</div>
<div class="citation" id="id133" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id72">rawJLW+17</a><span class="fn-bracket">]</span></span>
<p>Chelsea J-T Ju, Ruirui Li, Zhengliang Wu, Jyun-Yu Jiang, Zhao Yang, and Wei Wang. Fleximer: accurate quantification of term`RNA`-Seq via variable-length k-mers. In <em>Proceedings of the 8th ACM International Conference on Bioinformatics, Computational Biology, and Health Informatics</em>, 263–272. 2017.</p>
</div>
<div class="citation" id="id115" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawKYD21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id10">1</a>,<a role="doc-backlink" href="#id19">2</a>,<a role="doc-backlink" href="#id27">3</a>,<a role="doc-backlink" href="#id31">4</a>,<a role="doc-backlink" href="#id48">5</a>,<a role="doc-backlink" href="#id52">6</a>,<a role="doc-backlink" href="#id55">7</a>,<a role="doc-backlink" href="#id56">8</a>,<a role="doc-backlink" href="#id61">9</a>,<a role="doc-backlink" href="#id63">10</a>,<a role="doc-backlink" href="#id67">11</a>,<a role="doc-backlink" href="#id70">12</a>,<a role="doc-backlink" href="#id87">13</a>)</span>
<p>Benjamin Kaminow, Dinar Yunusov, and Alexander Dobin. STARsolo: accurate, fast and versatile mapping/quantification of single-cell and single-nucleus term`RNA`-seq data. <em>bioRxiv</em>, 2021.</p>
</div>
<div class="citation" id="id97" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id14">rawLi18</a><span class="fn-bracket">]</span></span>
<p>Heng Li. Minimap2: pairwise alignment for nucleotide sequences. <em>Bioinformatics</em>, 34(18):3094–3100, May 2018. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bty191">https://doi.org/10.1093/bioinformatics/bty191</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bty191">doi:10.1093/bioinformatics/bty191</a>.</p>
</div>
<div class="citation" id="id123" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawLRA+19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id38">1</a>,<a role="doc-backlink" href="#id40">2</a>,<a role="doc-backlink" href="#id66">3</a>,<a role="doc-backlink" href="#id68">4</a>)</span>
<p>Aaron TL Lun, Samantha Riesenfeld, Tallulah Andrews, Tomas Gomes, John C Marioni, and others. EmptyDrops: distinguishing cells from empty droplets in droplet-based single-cell term`RNA` sequencing data. <em>Genome biology</em>, 20(1):1–9, 2019.</p>
</div>
<div class="citation" id="id150" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawMBS+15<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id48">1</a>,<a role="doc-backlink" href="#id64">2</a>,<a role="doc-backlink" href="#id82">3</a>)</span>
<p>Evan Z. Macosko, Anindita Basu, Rahul Satija, James Nemesh, Karthik Shekhar, Melissa Goldman, Itay Tirosh, Allison R. Bialas, Nolan Kamitaki, Emily M. Martersteck, John J. Trombetta, David A. Weitz, Joshua R. Sanes, Alex K. Shalek, Aviv Regev, and Steven A. McCarroll. Highly parallel genome-wide expression profiling of individual cells using nanoliter droplets. <em>Cell</em>, 161(5):1202–1214, May 2015. URL: <a class="reference external" href="https://doi.org/10.1016/j.cell.2015.05.002">https://doi.org/10.1016/j.cell.2015.05.002</a>, <a class="reference external" href="https://doi.org/10.1016/j.cell.2015.05.002">doi:10.1016/j.cell.2015.05.002</a>.</p>
</div>
<div class="citation" id="id107" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id15">rawMSEG+22</a><span class="fn-bracket">]</span></span>
<p>Santiago Marco-Sola, Jordan M Eizenga, Andrea Guarracino, Benedict Paten, Erik Garrison, and Miquel Moreto. Optimal gap-affine alignment in o(s) space. <em>bioRxiv</em>, April 2022. URL: <a class="reference external" href="https://doi.org/10.1101/2022.04.14.488380">https://doi.org/10.1101/2022.04.14.488380</a>, <a class="reference external" href="https://doi.org/10.1101/2022.04.14.488380">doi:10.1101/2022.04.14.488380</a>.</p>
</div>
<div class="citation" id="id105" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id15">rawMSMME20</a><span class="fn-bracket">]</span></span>
<p>Santiago Marco-Sola, Juan Carlos Moure, Miquel Moreto, and Antonio Espinosa. Fast gap-affine pairwise alignment using the wavefront algorithm. <em>Bioinformatics</em>, September 2020. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btaa777">https://doi.org/10.1093/bioinformatics/btaa777</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btaa777">doi:10.1093/bioinformatics/btaa777</a>.</p>
</div>
<div class="citation" id="id130" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawMMG19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id39">1</a>,<a role="doc-backlink" href="#id69">2</a>)</span>
<p>Christopher S. McGinnis, Lyndsay M. Murrow, and Zev J. Gartner. DoubletFinder: doublet detection in single-cell term`RNA` sequencing data using artificial nearest neighbors. <em>Cell Systems</em>, 8(4):329–337.e4, April 2019. URL: <a class="reference external" href="https://doi.org/10.1016/j.cels.2019.03.003">https://doi.org/10.1016/j.cels.2019.03.003</a>, <a class="reference external" href="https://doi.org/10.1016/j.cels.2019.03.003">doi:10.1016/j.cels.2019.03.003</a>.</p>
</div>
<div class="citation" id="id116" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawMBL+21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id9">1</a>,<a role="doc-backlink" href="#id34">2</a>,<a role="doc-backlink" href="#id48">3</a>,<a role="doc-backlink" href="#id54">4</a>,<a role="doc-backlink" href="#id61">5</a>,<a role="doc-backlink" href="#id63">6</a>,<a role="doc-backlink" href="#id70">7</a>,<a role="doc-backlink" href="#id86">8</a>)</span>
<p>Páll Melsted, A. Sina Booeshaghi, Lauren Liu, Fan Gao, Lambda Lu, Kyung Hoi Min, Eduardo da Veiga Beltrame, Kristján Eldjárn Hjörleifsson, Jase Gehring, and Lior Pachter. Modular, efficient and constant-memory single-cell rna-seq preprocessing. <em>Nature Biotechnology</em>, 39(7):813–818, April 2021. URL: <a class="reference external" href="https://doi.org/10.1038/s41587-021-00870-2">https://doi.org/10.1038/s41587-021-00870-2</a>, <a class="reference external" href="https://doi.org/10.1038/s41587-021-00870-2">doi:10.1038/s41587-021-00870-2</a>.</p>
</div>
<div class="citation" id="id126" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawMP21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id38">1</a>,<a role="doc-backlink" href="#id66">2</a>)</span>
<p>Walter Muskovic and Joseph E Powell. DropletQC: improved identification of empty droplets and damaged cells in single-cell term`RNA`-seq data. <em>Genome Biology</em>, 22(1):1–9, 2021.</p>
</div>
<div class="citation" id="id140" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id72">rawNMMuandoiuZ11</a><span class="fn-bracket">]</span></span>
<p>Marius Nicolae, Serghei Mangul, Ion I Măndoiu, and Alex Zelikovsky. Estimation of alternative splicing isoform frequencies from term`RNA`-Seq data. <em>Algorithms for molecular biology</em>, 6(1):1–13, 2011.</p>
</div>
<div class="citation" id="id98" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id8">rawNMullerHS20</a><span class="fn-bracket">]</span></span>
<p>Stefan Niebler, André Müller, Thomas Hankeln, and Bertil Schmidt. RainDrop: Rapid activation matrix computation for droplet-based single-cell term`RNA`-seq reads. <em>BMC bioinformatics</em>, 21(1):1–14, 2020.</p>
</div>
<div class="citation" id="id147" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id73">rawNKZ+16</a><span class="fn-bracket">]</span></span>
<p>Vasilis Ntranos, Govinda M Kamath, Jesse M Zhang, Lior Pachter, and David N Tse. Fast and accurate single-cell term`RNA`-seq analysis by clustering of transcript-compatibility counts. <em>Genome biology</em>, 17(1):1–14, 2016.</p>
</div>
<div class="citation" id="id146" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id74">rawNYMP19</a><span class="fn-bracket">]</span></span>
<p>Vasilis Ntranos, Lynn Yi, Páll Melsted, and Lior Pachter. A discriminative learning approach to differential expression analysis for single-cell term`rna`-seq. <em>Nature methods</em>, 16(2):163–166, 2019.</p>
</div>
<div class="citation" id="id119" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id48">rawOEM+18</a><span class="fn-bracket">]</span></span>
<p>Baraa Orabi, Emre Erhan, Brian McConeghy, Stanislav V Volik, Stephane Le Bihan, Robert Bell, Colin C Collins, Cedric Chauve, and Faraz Hach. Alignment-free clustering of UMI tagged term`DNA` molecules. <em>Bioinformatics</em>, 35(11):1829–1836, October 2018. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bty888">https://doi.org/10.1093/bioinformatics/bty888</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bty888">doi:10.1093/bioinformatics/bty888</a>.</p>
</div>
<div class="citation" id="id121" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawPZV+18<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id6">1</a>,<a role="doc-backlink" href="#id25">2</a>,<a role="doc-backlink" href="#id48">3</a>,<a role="doc-backlink" href="#id84">4</a>)</span>
<p>Swati Parekh, Christoph Ziegenhain, Beate Vieth, Wolfgang Enard, and Ines Hellmann. zUMIs - a fast and flexible pipeline to process term`RNA` sequencing data with UMIs. <em>GigaScience</em>, May 2018. URL: <a class="reference external" href="https://doi.org/10.1093/gigascience/giy059">https://doi.org/10.1093/gigascience/giy059</a>, <a class="reference external" href="https://doi.org/10.1093/gigascience/giy059">doi:10.1093/gigascience/giy059</a>.</p>
</div>
<div class="citation" id="id102" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id78">rawPHJ+20</a><span class="fn-bracket">]</span></span>
<p>Ralph Patrick, David T. Humphreys, Vaibhao Janbandhu, Alicia Oshlack, Joshua W.K. Ho, Richard P. Harvey, and Kitty K. Lo. Sierra: discovery of differential transcript usage from polyA-captured single-cell term`RNA`-seq data. <em>Genome Biology</em>, jul 2020. URL: <a class="reference external" href="https://doi.org/10.1186%2Fs13059-020-02071-7">https://doi.org/10.1186%2Fs13059-020-02071-7</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-020-02071-7">doi:10.1186/s13059-020-02071-7</a>.</p>
</div>
<div class="citation" id="id137" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id72">rawPDL+17</a><span class="fn-bracket">]</span></span>
<p>Rob Patro, Geet Duggal, Michael I Love, Rafael A Irizarry, and Carl Kingsford. Salmon provides fast and bias-aware quantification of transcript expression. <em>Nature methods</em>, 14(4):417–419, 2017.</p>
</div>
<div class="citation" id="id136" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id72">rawPMK14</a><span class="fn-bracket">]</span></span>
<p>Rob Patro, Stephen M Mount, and Carl Kingsford. Sailfish enables alignment-free isoform quantification from term`RNA`-seq reads using lightweight algorithms. <em>Nature biotechnology</em>, 32(5):462–464, 2014.</p>
</div>
<div class="citation" id="id142" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawPPC+22<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id28">1</a>,<a role="doc-backlink" href="#id30">2</a>,<a role="doc-backlink" href="#id36">3</a>,<a role="doc-backlink" href="#id81">4</a>)</span>
<p>Allan-Hermann Pool, Helen Poldsam, Sisi Chen, Matt Thomson, and Yuki Oka. Enhanced recovery of single-cell term`RNA`-sequencing reads for missing gene expression data. <em>bioRxiv</em>, 2022. URL: <a class="reference external" href="https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449">https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449.full.pdf">arXiv:https://www.biorxiv.org/content/early/2022/04/27/2022.04.26.489449.full.pdf</a>, <a class="reference external" href="https://doi.org/10.1101/2022.04.26.489449">doi:10.1101/2022.04.26.489449</a>.</p>
</div>
<div class="citation" id="id101" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id16">rawRS00</a><span class="fn-bracket">]</span></span>
<p>Torbjørn Rognes and Erling Seeberg. Six-fold speed-up of Smith–Waterman sequence database searches using parallel processing on common microprocessors. <em>Bioinformatics</em>, 16(8):699–706, 2000.</p>
</div>
<div class="citation" id="id110" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawSHS17<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id4">1</a>,<a role="doc-backlink" href="#id40">2</a>,<a role="doc-backlink" href="#id41">3</a>,<a role="doc-backlink" href="#id43">4</a>,<a role="doc-backlink" href="#id48">5</a>,<a role="doc-backlink" href="#id49">6</a>,<a role="doc-backlink" href="#id50">7</a>,<a role="doc-backlink" href="#id56">8</a>,<a role="doc-backlink" href="#id57">9</a>,<a role="doc-backlink" href="#id58">10</a>,<a role="doc-backlink" href="#id65">11</a>)</span>
<p>Tom Smith, Andreas Heger, and Ian Sudbery. UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. <em>Genome research</em>, 27(3):491–499, 2017.</p>
</div>
<div class="citation" id="id117" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id33">rawSSPS21</a><span class="fn-bracket">]</span></span>
<p>Charlotte Soneson, Avi Srivastava, Rob Patro, and Michael B Stadler. Preprocessing choices affect RNA velocity results for droplet scRNA-seq data. <em>PLoS computational biology</em>, 17(1):e1008585, 2021.</p>
</div>
<div class="citation" id="id122" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id62">rawSMSP20</a><span class="fn-bracket">]</span></span>
<p>Avi Srivastava, Laraib Malik, Hirak Sarkar, and Rob Patro. A Bayesian framework for inter-cellular information sharing improves dscRNA-seq quantification. <em>Bioinformatics</em>, 36(Supplement_1):i292–i299, July 2020. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btaa450">https://doi.org/10.1093/bioinformatics/btaa450</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btaa450">doi:10.1093/bioinformatics/btaa450</a>.</p>
</div>
<div class="citation" id="id139" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id20">rawSMS+20</a><span class="fn-bracket">]</span></span>
<p>Avi Srivastava, Laraib Malik, Hirak Sarkar, Mohsen Zakeri, Fatemeh Almodaresi, Charlotte Soneson, Michael I Love, Carl Kingsford, and Rob Patro. Alignment and mapping methodology influence transcript abundance estimation. <em>Genome Biology</em>, 21(1):1–29, 2020.</p>
</div>
<div class="citation" id="id111" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawSMS+19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id7">1</a>,<a role="doc-backlink" href="#id21">2</a>,<a role="doc-backlink" href="#id34">3</a>,<a role="doc-backlink" href="#id42">4</a>,<a role="doc-backlink" href="#id44">5</a>,<a role="doc-backlink" href="#id45">6</a>,<a role="doc-backlink" href="#id48">7</a>,<a role="doc-backlink" href="#id51">8</a>,<a role="doc-backlink" href="#id55">9</a>,<a role="doc-backlink" href="#id56">10</a>,<a role="doc-backlink" href="#id57">11</a>,<a role="doc-backlink" href="#id59">12</a>,<a role="doc-backlink" href="#id60">13</a>,<a role="doc-backlink" href="#id70">14</a>,<a role="doc-backlink" href="#id71">15</a>,<a role="doc-backlink" href="#id85">16</a>)</span>
<p>Avi Srivastava, Laraib Malik, Tom Smith, Ian Sudbery, and Rob Patro. Alevin efficiently estimates accurate gene abundances from dscRNA-seq data. <em>Genome biology</em>, 20(1):1–16, 2019.</p>
</div>
<div class="citation" id="id134" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id23">rawSSGP16</a><span class="fn-bracket">]</span></span>
<p>Avi Srivastava, Hirak Sarkar, Nitish Gupta, and Rob Patro. Rapmap: a rapid, sensitive and accurate tool for mapping term`rna`-seq reads to transcriptomes. <em>Bioinformatics</em>, 32(12):i192–i200, 2016.</p>
</div>
<div class="citation" id="id106" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id17">rawSK18</a><span class="fn-bracket">]</span></span>
<p>Hajime Suzuki and Masahiro Kasahara. Introducing difference recurrence relations for faster semi-global alignment of long sequences. <em>BMC Bioinformatics</em>, February 2018. URL: <a class="reference external" href="https://doi.org/10.1186/s12859-018-2014-8">https://doi.org/10.1186/s12859-018-2014-8</a>, <a class="reference external" href="https://doi.org/10.1186/s12859-018-2014-8">doi:10.1186/s12859-018-2014-8</a>.</p>
</div>
<div class="citation" id="id120" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawTMP+21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id48">1</a>,<a role="doc-backlink" href="#id53">2</a>)</span>
<p>Maria Tsagiopoulou, Maria Christina Maniou, Nikolaos Pechlivanis, Anastasis Togkousidis, Michaela Kotrová, Tobias Hutzenlaub, Ilias Kappas, Anastasia Chatzidimitriou, and Fotis Psomopoulos. UMIc: a preprocessing method for UMI deduplication and reads correction. <em>Frontiers in Genetics</em>, May 2021. URL: <a class="reference external" href="https://doi.org/10.3389/fgene.2021.660366">https://doi.org/10.3389/fgene.2021.660366</a>, <a class="reference external" href="https://doi.org/10.3389/fgene.2021.660366">doi:10.3389/fgene.2021.660366</a>.</p>
</div>
<div class="citation" id="id138" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id72">rawTSGonccalves+11</a><span class="fn-bracket">]</span></span>
<p>Ernest Turro, Shu-Yi Su, Ângela Gonçalves, Lachlan JM Coin, Sylvia Richardson, and Alex Lewin. Haplotype and isoform specific expression estimation using multi-mapping term`RNA`-seq reads. <em>Genome biology</em>, 12(2):1–15, 2011.</p>
</div>
<div class="citation" id="id131" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawWLK19<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id39">1</a>,<a role="doc-backlink" href="#id69">2</a>)</span>
<p>Samuel L. Wolock, Romain Lopez, and Allon M. Klein. Scrublet: Computational Identification of Cell Doublets in Single-Cell Transcriptomic Data. <em>Cell Systems</em>, 8(4):281–291.e9, April 2019. URL: <a class="reference external" href="https://doi.org/10.1016/j.cels.2018.11.005">https://doi.org/10.1016/j.cels.2018.11.005</a>, <a class="reference external" href="https://doi.org/10.1016/j.cels.2018.11.005">doi:10.1016/j.cels.2018.11.005</a>.</p>
</div>
<div class="citation" id="id99" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id16">rawWoz97</a><span class="fn-bracket">]</span></span>
<p>Andrzej Wozniak. Using video-oriented instructions to speed up sequence comparison. <em>Bioinformatics</em>, 13(2):145–150, 1997.</p>
</div>
<div class="citation" id="id104" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawYTS+21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id75">1</a>,<a role="doc-backlink" href="#id76">2</a>)</span>
<p>Yue You, Luyi Tian, Shian Su, Xueyi Dong, Jafar S. Jabbari, Peter F. Hickey, and Matthew E. Ritchie. Benchmarking UMI-based single-cell term`RNA`-seq preprocessing workflows. <em>Genome Biology</em>, dec 2021. URL: <a class="reference external" href="https://doi.org/10.1186%2Fs13059-021-02552-3">https://doi.org/10.1186%2Fs13059-021-02552-3</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-021-02552-3">doi:10.1186/s13059-021-02552-3</a>.</p>
</div>
<div class="citation" id="id128" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawYB20<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id38">1</a>,<a role="doc-backlink" href="#id66">2</a>)</span>
<p>Matthew D Young and Sam Behjati. SoupX removes ambient term`RNA` contamination from droplet-based single-cell term`RNA` sequencing data. <em>GigaScience</em>, December 2020. URL: <a class="reference external" href="https://doi.org/10.1093/gigascience/giaa151">https://doi.org/10.1093/gigascience/giaa151</a>, <a class="reference external" href="https://doi.org/10.1093/gigascience/giaa151">doi:10.1093/gigascience/giaa151</a>.</p>
</div>
<div class="citation" id="id95" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">rawZT21</a><span class="fn-bracket">]</span></span>
<p>Luke Zappia and Fabian J. Theis. Over 1000 tools reveal trends in the single-cell rna-seq analysis landscape. <em>Genome Biology</em>, 22(1):301, Oct 2021. URL: <a class="reference external" href="https://doi.org/10.1186/s13059-021-02519-4">https://doi.org/10.1186/s13059-021-02519-4</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-021-02519-4">doi:10.1186/s13059-021-02519-4</a>.</p>
</div>
<div class="citation" id="id108" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id13">rawZSWM00</a><span class="fn-bracket">]</span></span>
<p>Zheng Zhang, Scott Schwartz, Lukas Wagner, and Webb Miller. A greedy algorithm for aligning term`DNA` sequences. <em>Journal of Computational Biology</em>, 7(1-2):203–214, February 2000. URL: <a class="reference external" href="https://doi.org/10.1089/10665270050081478">https://doi.org/10.1089/10665270050081478</a>, <a class="reference external" href="https://doi.org/10.1089/10665270050081478">doi:10.1089/10665270050081478</a>.</p>
</div>
<div class="citation" id="id151" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>rawZTB+17<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id5">1</a>,<a role="doc-backlink" href="#id26">2</a>,<a role="doc-backlink" href="#id52">3</a>,<a role="doc-backlink" href="#id63">4</a>)</span>
<p>Grace X. Y. Zheng, Jessica M. Terry, Phillip Belgrader, Paul Ryvkin, Zachary W. Bent, Ryan Wilson, Solongo B. Ziraldo, Tobias D. Wheeler, Geoff P. McDermott, Junjie Zhu, Mark T. Gregory, Joe Shuga, Luz Montesclaros, Jason G. Underwood, Donald A. Masquelier, Stefanie Y. Nishimura, Michael Schnall-Levin, Paul W. Wyatt, Christopher M. Hindson, Rajiv Bharadwaj, Alexander Wong, Kevin D. Ness, Lan W. Beppu, H. Joachim Deeg, Christopher McFarland, Keith R. Loeb, William J. Valente, Nolan G. Ericson, Emily A. Stevens, Jerald P. Radich, Tarjei S. Mikkelsen, Benjamin J. Hindson, and Jason H. Bielas. Massively parallel digital transcriptional profiling of single cells. <em>Nature Communications</em>, 8(1):14049, Jan 2017. URL: <a class="reference external" href="https://doi.org/10.1038/ncomms14049">https://doi.org/10.1038/ncomms14049</a>, <a class="reference external" href="https://doi.org/10.1038/ncomms14049">doi:10.1038/ncomms14049</a>.</p>
</div>
<div class="citation" id="id143" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id41">rawZHHJS22</a><span class="fn-bracket">]</span></span>
<p>Christoph Ziegenhain, Gert-Jan Hendriks, Michael Hagemann-Jensen, and Rickard Sandberg. Molecular spikes: a gold standard for single-cell term`RNA` counting. <em>Nature Methods</em>, 19(5):560–566, 2022.</p>
</div>
<div class="citation" id="id112" role="doc-biblioentry">
<span class="label"><span class="fn-bracket">[</span>raw10xGenomics21<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id30">1</a>,<a role="doc-backlink" href="#id36">2</a>,<a role="doc-backlink" href="#id37">3</a>,<a role="doc-backlink" href="#id46">4</a>,<a role="doc-backlink" href="#id47">5</a>)</span>
<p>10x Genomics. Technical note - interpreting intronic and antisense reads in 10x genomics single cell gene expression data. <a class="reference external" href="https://www.10xgenomics.com/support/single-cell-gene-expression/documentation/steps/sequencing/interpreting-intronic-and-antisense-reads-in-10-x-genomics-single-cell-gene-expression-data">https://www.10xgenomics.com/support/single-cell-gene-expression/documentation/steps/sequencing/interpreting-intronic-and-antisense-reads-in-10-x-genomics-single-cell-gene-expression-data</a>, August 2021.</p>
</div>
</div>
</div>
</section>
<section id="contributors">
<h2><span class="section-number">3.11. </span>Contributors<a class="headerlink" href="#contributors" title="Link to this heading">#</a></h2>
<p>We gratefully acknowledge the contributions of:</p>
<section id="authors">
<h3><span class="section-number">3.11.1. </span>Authors<a class="headerlink" href="#authors" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Dongze He</p></li>
<li><p>Avi Srivastava</p></li>
<li><p>Hirak Sarkar</p></li>
<li><p>Rob Patro</p></li>
</ul>
</section>
<section id="reviewers">
<h3><span class="section-number">3.11.2. </span>Reviewers<a class="headerlink" href="#reviewers" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p>Lukas Heumos</p></li>
</ul>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./introduction"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="scrna_seq.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title"><span class="section-number">2. </span>Single-cell RNA sequencing</p>
      </div>
    </a>
    <a class="right-next"
       href="analysis_tools.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">4. </span>Analysis frameworks and tools</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#raw-data-quality-control">3.1. Raw data quality control</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#alignment-and-mapping">3.2. Alignment and mapping</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#types-of-mapping">3.2.1. Types of mapping</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-against-different-reference-sequences">3.2.2. Mapping against different reference sequences</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-to-the-full-genome">3.2.2.1. Mapping to the full genome</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-to-the-spliced-transcriptome">3.2.2.2. Mapping to the spliced transcriptome</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-to-an-augmented-transcriptome">3.2.2.3. Mapping to an augmented transcriptome</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cell-barcode-correction">3.3. Cell barcode correction</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#type-of-errors-in-barcoding">3.3.1. Type of errors in barcoding</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#future-challenges">3.3.2. Future challenges</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#umi-resolution">3.4. UMI resolution</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-need-for-umi-resolution">3.4.1. The need for UMI resolution</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#graph-based-umi-resolution">3.4.2. Graph-based UMI resolution</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#defining-nodes">3.4.2.1. Defining nodes</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#defining-the-adjacency-relationship">3.4.2.2. Defining the adjacency relationship</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#defining-the-graph-resolution-approach">3.4.2.3. Defining the graph resolution approach</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#quantification">3.4.2.4. Quantification</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#count-matrix-quality-control">3.5. Count matrix quality control</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#empty-droplet-detection">3.5.1. Empty droplet detection</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#doublet-detection">3.5.2. Doublet detection</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#count-data-representation">3.6. Count data representation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#brief-discussion">3.7. Brief discussion</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#a-real-world-example">3.8. A real-world example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#preparation">3.8.1. Preparation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#simplified-raw-data-processing-pipeline">3.8.2. Simplified raw data processing pipeline</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#the-complete-alevin-fry-pipeline">3.8.3. The complete alevin-fry pipeline</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#building-the-index">3.8.3.1. Building the index</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mapping-and-quantification">3.8.3.2. Mapping and quantification</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#useful-links">3.9. Useful links</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#references">3.10. References</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#contributors">3.11. Contributors</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#authors">3.11.1. Authors</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reviewers">3.11.2. Reviewers</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Lukas Heumos, Anna Schaar, single-cell best practices consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2023.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>