
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Conversion of muon objects to Seurat objects to use in downstream analysis &#8212; Single-cell best practices</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/book.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Single-cell best practices</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Introduction
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/prior_art.html">
   1. Prior art
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/scrna_seq.html">
   2. Single-cell RNA sequencing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/raw_data_processing.html">
   3. Raw data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/analysis_tools.html">
   4. Analysis frameworks and tools
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../introduction/interoperability.html">
   5. Interoperability
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Preprocessing and visualization
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/quality_control.html">
   6. Quality Control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/normalization.html">
   7. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/feature_selection.html">
   8. Feature selection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/dimensionality_reduction.html">
   9. Dimensionality Reduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Identifying cellular structure
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/clustering.html">
   10. Clustering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/annotation.html">
   11. Annotation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/integration.html">
   12. Data integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Inferring trajectories
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/pseudotemporal.html">
   13. Pseudotemporal ordering
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/rna_velocity.html">
   14. RNA velocity
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../trajectories/lineage_tracing.html">
   15. Lineage tracing
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Dealing with conditions
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/differential_gene_expression.html">
   16. Differential gene expression analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/compositional.html">
   17. Compositional analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/gsea_pathway.html">
   18. Gene set enrichment and pathway analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/perturbation_modeling.html">
   19. Perturbation modeling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Modeling mechanisms
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/gene_regulatory_networks.html">
   20. Gene regulatory networks
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/cell_cell_communication.html">
   21. Cell-cell communication
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Deconvolution
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../deconvolution/bulk_deconvolution.html">
   22. Bulk deconvolution
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Chromatin Accessibility
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="introduction.html">
   23. Single-cell ATAC sequencing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="quality_control.html">
   24. Quality Control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="gene_regulatory_networks_atac.html">
   25. Gene regulatory networks using chromatin accessibility
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Spatial omics
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../spatial/introduction.html">
   26. Single-cell data resolved in space
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../spatial/neighborhood.html">
   27. Neighborhood analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../spatial/domains.html">
   28. Spatial domains
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../spatial/spatially_variable_genes.html">
   29. Spatially variable genes
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../spatial/deconvolution.html">
   30. Spatial deconvolution
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../spatial/imputation.html">
   31. Imputation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Surface protein
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/quality_control.html">
   32. Quality control
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/normalization.html">
   33. Normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/doublet_detection.html">
   34. Doublet detection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/dimensionality_reduction.html">
   35. Dimensionality Reduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/batch_correction.html">
   36. Batch correction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../surface_protein/annotation.html">
   37. Annotation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Adaptive immune receptor repertoire
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/ir_profiling.html">
   38. Immune Receptor Profiling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/clonotype.html">
   39. Clonotype analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/specificity.html">
   43. Specificity analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../air_repertoire/multimodal_integration.html">
   44. Integrating AIR and transcriptomics
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multimodal integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../multimodal_integration/paired_integration.html">
   45. Paired integration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../multimodal_integration/advanced_integration.html">
   46. Advanced integration
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reproducibility
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../reproducibility/introduction.html">
   47. Reproducibility
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Outlook
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../outlook.html">
   48. Outlook
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Acknowledgements
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../acknowledgements.html">
   49. Acknowledgements
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Glossary
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../glossary.html">
   50. Glossary
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/theislab/single-cell-best-practices/master?urlpath=tree/jupyter-book/chromatin_accessibility/muon_to_seurat.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="../_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/issues/new?title=Issue%20on%20page%20%2Fchromatin_accessibility/muon_to_seurat.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/theislab/single-cell-best-practices/edit/master/jupyter-book/chromatin_accessibility/muon_to_seurat.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Edit this page"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="headerbtn__text-container">suggest edit</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/chromatin_accessibility/muon_to_seurat.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#environment-setup">
   Environment setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#load-data">
   Load data
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#read-in-mudata-object">
     Read in mudata object
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#convert-seurat-assay-to-chromatin-assay">
     Convert Seurat assay to chromatin assay
    </a>
   </li>
  </ul>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Conversion of muon objects to Seurat objects to use in downstream analysis</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#environment-setup">
   Environment setup
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#load-data">
   Load data
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#read-in-mudata-object">
     Read in mudata object
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#convert-seurat-assay-to-chromatin-assay">
     Convert Seurat assay to chromatin assay
    </a>
   </li>
  </ul>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="conversion-of-muon-objects-to-seurat-objects-to-use-in-downstream-analysis">
<h1>Conversion of muon objects to Seurat objects to use in downstream analysis<a class="headerlink" href="#conversion-of-muon-objects-to-seurat-objects-to-use-in-downstream-analysis" title="Permalink to this headline">#</a></h1>
<p>As we saw in the introduction, there are some downstream analyses that are currently only available in R. In this notebook, we therfore want to showcase how you can convert a muon object into a Seurat object for downstream analysis using Signac.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">.libPaths</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;/home/icb/laura.martens/miniconda3/envs/best_practices/lib/R/library&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;/home/icb/laura.martens/miniconda3/envs/signac/lib/R/library&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<section id="environment-setup">
<h2>Environment setup<a class="headerlink" href="#environment-setup" title="Permalink to this headline">#</a></h2>
<p>We load the R libraries <code class="docutils literal notranslate"><span class="pre">Signac</span></code> and <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> to handle scATAC-seq data in R. Furthermore, we need to load the annotation for our reference genome. In this example, we load the annotation for human and the <code class="docutils literal notranslate"><span class="pre">hg38</span></code> reference genome.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">MuDataSeurat</span><span class="p">)</span><span class="c1"># to read in muon object and convert to Seurat</span>
<span class="nf">library</span><span class="p">(</span><span class="n">Signac</span><span class="p">)</span><span class="w"> </span>
<span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span>

<span class="nf">library</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v86</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Warning message:
&quot;package &#39;stringr&#39; was built under R version 4.2.2&quot;
</pre></div>
</div>
</div>
</div>
</section>
<section id="load-data">
<h2>Load data<a class="headerlink" href="#load-data" title="Permalink to this headline">#</a></h2>
<p>With Signac, we have the possibility to visualize the coverage of the features or perform other analysis like footprinting. For some of these downstream analysis, it is necessary to add the fragment file to the Seurat object which we specifiy here with <code class="docutils literal notranslate"><span class="pre">fragmen_file</span></code></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">fragment_file</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;/lustre/groups/ml01/datasets/projects/20220323_neurips21_bmmc_christopher.lance/multiome/aggr_donors/atac_fragments.tsv.gz&quot;</span>
<span class="n">adata_file</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;/lustre/groups/ml01/code/laura.martens/scATAC_pipelines/src/best_practices/multiome_donor_s4d8_pp_test.h5mu&quot;</span><span class="w"> </span><span class="c1">#&quot;/lustre/groups/ml01/workspace/christopher.lance/mm_best/src/scATAC/muon_data/neurips_qc_filtered_allsamp.h5ad&quot;</span>
</pre></div>
</div>
</div>
</div>
<section id="read-in-mudata-object">
<h3>Read in mudata object<a class="headerlink" href="#read-in-mudata-object" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">atac</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">MuDataSeurat</span><span class="o">::</span><span class="nf">ReadH5MU</span><span class="p">(</span><span class="n">adata_file</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Warning message in missing_on_read(&quot;/var&quot;, paste0(&quot;global variables metadata (&quot;, :
&quot;Missing on read: /var. Seurat does not support global variables metadata (gene_ids, feature_types, genome, interval).&quot;
Warning message in read_layers_to_assay(h5[[&quot;mod&quot;]][[mod]], mod):
&quot;The var_names from modality gene_activity have been renamed as feature names cannot contain &#39;_&#39;. E.g. gene_activity:MIR1302-2HG -&gt; gene-activity:MIR1302-2HG.&quot;
as(&lt;dsCMatrix&gt;, &quot;dgCMatrix&quot;) is deprecated since Matrix 1.5-0; do as(., &quot;generalMatrix&quot;) instead

Warning message in missing_on_read(paste0(&quot;/mod&quot;, mod, &quot;/varp&quot;), &quot;pairwise annotation of variables&quot;):
&quot;Missing on read: /modrna/varp. Seurat does not support pairwise annotation of variables.&quot;
Warning message in missing_on_read(paste0(&quot;/mod&quot;, mod, &quot;/varp&quot;), &quot;pairwise annotation of variables&quot;):
&quot;Missing on read: /modatac/varp. Seurat does not support pairwise annotation of variables.&quot;
Warning message in missing_on_read(paste0(&quot;/mod&quot;, mod, &quot;/varp&quot;), &quot;pairwise annotation of variables&quot;):
&quot;Missing on read: /modgene_activity/varp. Seurat does not support pairwise annotation of variables.&quot;
Warning message:
&quot;Keys should be one or more alphanumeric characters followed by an underscore, setting key from rna to rna_&quot;
Warning message:
&quot;Keys should be one or more alphanumeric characters followed by an underscore, setting key from gene_activity_ to geneactivity_&quot;
Warning message:
&quot;No columnames present in cell embeddings, setting to &#39;LSI_1:49&#39;&quot;
Warning message:
&quot;No columnames present in cell embeddings, setting to &#39;UMAP_1:2&#39;&quot;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">atac</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>An object of class Seurat 
220291 features across 14710 samples within 3 assays 
Active assay: rna (36601 features, 0 variable features)
 2 other assays present: atac, gene_activity
 2 dimensional reductions calculated: LSI, UMAP
</pre></div>
</div>
</div>
</div>
<p>We have succesfully loaded the muon object into R and it has been converted to a Seurat object with three assays <code class="docutils literal notranslate"><span class="pre">atac</span></code>, <code class="docutils literal notranslate"><span class="pre">rna</span></code> and <code class="docutils literal notranslate"><span class="pre">gene_activity</span></code>. The dimensionality reductions and metadata have also been transferred.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">head</span><span class="p">(</span><span class="n">atac</span><span class="p">[[]])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A data.frame: 6 × 27</caption>
<thead>
	<tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_rna</th><th scope=col>nFeature_rna</th><th scope=col>nCount_atac</th><th scope=col>nFeature_atac</th><th scope=col>nCount_gene_activity</th><th scope=col>nFeature_gene_activity</th><th scope=col>n_features_per_cell</th><th scope=col>total_fragment_counts</th><th scope=col>log_total_fragment_counts</th><th scope=col>⋯</th><th scope=col>n_features_per_cell.1</th><th scope=col>total_fragment_counts.1</th><th scope=col>log_total_fragment_counts.1</th><th scope=col>nucleosome_signal.1</th><th scope=col>nuc_signal_filter.1</th><th scope=col>tss_score.1</th><th scope=col>site.1</th><th scope=col>donor.1</th><th scope=col>sample.1</th><th scope=col>leiden.1</th></tr>
	<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th></tr>
</thead>
<tbody>
	<tr><th scope=row>AAACAGCCAAGCTTAT-14</th><td>SeuratProject</td><td>1707</td><td>1128</td><td>3588</td><td>1797</td><td>2738</td><td>1409</td><td>1799</td><td>3591</td><td>3.555215</td><td>⋯</td><td>1799</td><td>3591</td><td>3.555215</td><td>0.6643357</td><td>NS_PASS</td><td> 3.521329</td><td>site4</td><td>donor8</td><td>s4d8</td><td>4 </td></tr>
	<tr><th scope=row>AAACAGCCATGAAATG-14</th><td>SeuratProject</td><td>1409</td><td> 479</td><td>5056</td><td>2552</td><td>3965</td><td>1957</td><td>2556</td><td>5062</td><td>3.704322</td><td>⋯</td><td>2556</td><td>5062</td><td>3.704322</td><td>0.6542056</td><td>NS_PASS</td><td> 2.296477</td><td>site4</td><td>donor8</td><td>s4d8</td><td>10</td></tr>
	<tr><th scope=row>AAACAGCCATGTTTGG-14</th><td>SeuratProject</td><td>1666</td><td>1077</td><td>7110</td><td>3360</td><td>5105</td><td>2495</td><td>3361</td><td>7112</td><td>3.851992</td><td>⋯</td><td>3361</td><td>7112</td><td>3.851992</td><td>0.6789298</td><td>NS_PASS</td><td>12.890983</td><td>site4</td><td>donor8</td><td>s4d8</td><td>3 </td></tr>
	<tr><th scope=row>AAACATGCAACGTGCT-14</th><td>SeuratProject</td><td> 882</td><td> 641</td><td>7569</td><td>3586</td><td>5397</td><td>2565</td><td>3591</td><td>7578</td><td>3.879555</td><td>⋯</td><td>3591</td><td>7578</td><td>3.879555</td><td>1.0609319</td><td>NS_PASS</td><td> 4.689176</td><td>site4</td><td>donor8</td><td>s4d8</td><td>6 </td></tr>
	<tr><th scope=row>AAACATGCAATATAGG-14</th><td>SeuratProject</td><td> 325</td><td> 277</td><td>8550</td><td>4005</td><td>6500</td><td>2932</td><td>4006</td><td>8551</td><td>3.932017</td><td>⋯</td><td>4006</td><td>8551</td><td>3.932017</td><td>0.9504644</td><td>NS_PASS</td><td> 4.600777</td><td>site4</td><td>donor8</td><td>s4d8</td><td>3 </td></tr>
	<tr><th scope=row>AAACATGCAATTAACC-14</th><td>SeuratProject</td><td>1704</td><td>1083</td><td>6993</td><td>3268</td><td>5061</td><td>2450</td><td>3269</td><td>6995</td><td>3.844788</td><td>⋯</td><td>3269</td><td>6995</td><td>3.844788</td><td>0.7665505</td><td>NS_PASS</td><td> 6.894277</td><td>site4</td><td>donor8</td><td>s4d8</td><td>0 </td></tr>
</tbody>
</table>
</div></div>
</div>
<p>We verfiy that the correct embedding and clustering has been added using <code class="docutils literal notranslate"><span class="pre">DimPlot</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="o">=</span><span class="s">&#39;leiden&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/muon_to_seurat_15_0.png" src="../_images/muon_to_seurat_15_0.png" />
</div>
</div>
<p>Seurat by default groups its cells by <code class="docutils literal notranslate"><span class="pre">idents</span></code> which we can set to our cell types or clustering</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">Idents</span><span class="p">(</span><span class="n">atac</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">atac</span><span class="p">[[]]</span><span class="o">$</span><span class="n">leiden</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="convert-seurat-assay-to-chromatin-assay">
<h3>Convert Seurat assay to chromatin assay<a class="headerlink" href="#convert-seurat-assay-to-chromatin-assay" title="Permalink to this headline">#</a></h3>
<p>Loading using <code class="docutils literal notranslate"><span class="pre">MuDataSeurat</span></code> does not support featues with the same name in different assays (the same gene name for <code class="docutils literal notranslate"><span class="pre">rna</span></code> and <code class="docutils literal notranslate"><span class="pre">gene_activity</span></code>), which is why we had to add the assay name to the feature with <code class="docutils literal notranslate"><span class="pre">assay:feature</span></code>. Additionally, the <code class="docutils literal notranslate"><span class="pre">atac</span></code> assay is still a regular Seurat <code class="docutils literal notranslate"><span class="pre">Assay</span></code> and not a Signac <code class="docutils literal notranslate"><span class="pre">ChromatinAssay</span></code> which we need for chromatin specific downstream analyses.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">split_assay_names</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">function</span><span class="p">(</span><span class="n">assay</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s">&#39;:&#39;</span><span class="p">){</span>
<span class="w">    </span><span class="c1">#Function to split the assay name from the feature name</span>
<span class="w">    </span><span class="n">new_names</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">str_split</span><span class="p">(</span><span class="nf">rownames</span><span class="p">(</span><span class="n">assay</span><span class="p">),</span><span class="w"> </span><span class="n">sep</span><span class="p">,</span><span class="w"> </span><span class="n">simplify</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">n</span><span class="o">=</span><span class="m">2</span><span class="p">)[,</span><span class="w"> </span><span class="m">2</span><span class="p">]</span>
<span class="w">    </span><span class="nf">rownames</span><span class="p">(</span><span class="n">assay</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">new_names</span>
<span class="w">    </span><span class="nf">return</span><span class="p">(</span><span class="n">assay</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<p>We will add gene annotation to our <code class="docutils literal notranslate"><span class="pre">ChromatinAssay</span></code>. You have to make sure that the chromosome names in your annotation match the names in your fragment files (so either <code class="docutils literal notranslate"><span class="pre">chr1</span></code> or <code class="docutils literal notranslate"><span class="pre">1</span></code>). See <code class="docutils literal notranslate"><span class="pre">?seqlevelsStyle</span></code> to see the different options.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># extract gene annotations from EnsDb</span>
<span class="n">annotations</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">GetGRangesFromEnsDb</span><span class="p">(</span><span class="n">ensdb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb.Hsapiens.v86</span><span class="p">)</span>
<span class="nf">seqlevelsStyle</span><span class="p">(</span><span class="n">annotations</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&#39;UCSC&#39;</span>
<span class="nf">genome</span><span class="p">(</span><span class="n">annotations</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;hg38&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">split_assay_names</span><span class="p">(</span><span class="n">atac</span><span class="p">[[</span><span class="s">&#39;atac&#39;</span><span class="p">]]</span><span class="o">@</span><span class="n">data</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">chrom_assay</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">CreateChromatinAssay</span><span class="p">(</span>
<span class="w">  </span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span>
<span class="w">  </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;:&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;-&quot;</span><span class="p">),</span><span class="w"> </span><span class="c1">#seperator of adata$var_names</span>
<span class="w">  </span><span class="n">genome</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&#39;hg38&#39;</span><span class="p">,</span><span class="w"> </span><span class="c1">#used reference genome</span>
<span class="w">  </span><span class="n">fragments</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">fragment_file</span><span class="p">,</span><span class="w"> </span>
<span class="w">  </span><span class="n">annotation</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">annotations</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Computing hash
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">atac</span><span class="p">[[</span><span class="s">&#39;atac&#39;</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">chrom_assay</span>
</pre></div>
</div>
</div>
</div>
<p>Finally, we can check if we can plot all features in all assays</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">atac</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&#39;atac&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">cov_plot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">CoveragePlot</span><span class="p">(</span>
<span class="w">  </span><span class="n">object</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">atac</span><span class="p">,</span>
<span class="w">  </span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;PAX5&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="n">annotation</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span>
<span class="w">  </span><span class="n">peaks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span>
<span class="p">)</span>
<span class="n">cov_plot</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Warning message:
&quot;Removed 1 rows containing missing values (`geom_segment()`).&quot;
</pre></div>
</div>
<img alt="../_images/muon_to_seurat_28_1.png" src="../_images/muon_to_seurat_28_1.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">atac</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&#39;rna&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">FeaturePlot</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;rna:PAX5&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/muon_to_seurat_30_0.png" src="../_images/muon_to_seurat_30_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">atac</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&#39;gene_activity&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">FeaturePlot</span><span class="p">(</span><span class="n">atac</span><span class="p">,</span><span class="w"> </span><span class="s">&#39;gene-activity:PAX5&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/muon_to_seurat_32_0.png" src="../_images/muon_to_seurat_32_0.png" />
</div>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "r"
        },
        kernelOptions: {
            name: "ir",
            path: "./chromatin_accessibility"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'ir'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Lukas Heumos, Anna Schaar, single-cell best practices consortium<br/>
  
      &copy; Copyright 2022.<br/>
    <div class="extra_footer">
      <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>